{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "PG-tGRnlFLA3"
   },
   "source": [
    "# CNN + GlobalAvgPool\n",
    "Run this model on all lncRNA cell lines using some theshold. This notebook tests zero.\n",
    "\n",
    "* zero\n",
    "* mean RCI\n",
    "* Antilog threshold set by EM/GMM.\n",
    "\n",
    "Slight disconnect: transcripts are filtered by length, but RCI threshold is computed with the RCI from all genes.\n",
    "\n",
    "Ran on CoLab Pro. RAM 5.6/12.7 GB. GPU 1.9/15.0 GB."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "0RmwUsVLFLA6",
    "outputId": "3e128712-441e-44e7-9a3d-0190c6584b67"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2023-02-05 18:45:16.641776\n",
      "Python 3.8.10\n",
      "sklearn 1.0.2\n"
     ]
    }
   ],
   "source": [
    "from datetime import datetime\n",
    "print(datetime.now())\n",
    "from platform import python_version\n",
    "print('Python',python_version())\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pickle\n",
    "import time # sleep function\n",
    "from os.path import isfile\n",
    "from matplotlib import pyplot as plt \n",
    "import sklearn   # pip install --upgrade scikit-learn\n",
    "print('sklearn',sklearn.__version__)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "tUtGXPrcFLA8"
   },
   "source": [
    "We prevously used sklearn.model_selection.ShuffleSplit   \n",
    "Now we avoid it due to this note in the \n",
    "[documentation](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.ShuffleSplit.html):\n",
    "Note: contrary to other cross-validation strategies, random splits do not guarantee that all folds will be different, although this is still very likely for sizeable datasets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "id": "PRX-UEr8FLA8"
   },
   "outputs": [],
   "source": [
    "import tensorflow as tf\n",
    "dt='float32'\n",
    "tf.keras.backend.set_floatx('float32')\n",
    "tf.random.set_seed(42) \n",
    "\n",
    "from tensorflow import keras\n",
    "from sklearn.mixture import GaussianMixture\n",
    "from sklearn.model_selection import KFold\n",
    "from sklearn.metrics import accuracy_score\n",
    "from sklearn.metrics import precision_score\n",
    "from sklearn.metrics import recall_score\n",
    "from sklearn.metrics import f1_score\n",
    "from sklearn.metrics import precision_recall_curve\n",
    "from sklearn.metrics import roc_curve\n",
    "from sklearn.metrics import roc_auc_score\n",
    "from sklearn.metrics import auc\n",
    "\n",
    "from keras.models import Sequential\n",
    "from keras.layers import Masking\n",
    "from keras.layers import Embedding\n",
    "from keras.layers import Conv1D\n",
    "from keras.layers import GlobalAveragePooling1D\n",
    "from keras.layers import GlobalMaxPooling1D\n",
    "from keras.layers import AveragePooling1D\n",
    "from keras.layers import MaxPooling1D\n",
    "from keras.layers import Bidirectional\n",
    "from keras.layers import LSTM\n",
    "from keras.layers import TimeDistributed\n",
    "from keras.layers import Flatten\n",
    "from keras.layers import Dense\n",
    "from keras.layers import Dropout\n",
    "from keras.losses import BinaryCrossentropy\n",
    "\n",
    "K=4\n",
    "CFILTERS=64\n",
    "FILTERSIZE=16\n",
    "RCELLS=32\n",
    "DCELLS=16\n",
    "EPOCHS=3\n",
    "FOLDS=5      \n",
    "EMBED_DIMEN = 4 # arbitrary hyperparameter\n",
    "BREAK = False   # break after first fold\n",
    "MINLEN=200\n",
    "MAXLEN=5000   \n",
    "PRETTY_PICTURES = False\n",
    "RCI_THRESHOLD_MECHANISM = 'Zero'\n",
    "CACHING = False # Ran on CoLab with True but it wrote the cache"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "OlzN9OdsFWEU",
    "outputId": "2c6d32ef-0911-4c6f-ec78-a64a52079db5"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running on CoLab\n",
      "Mounted at /content/drive/\n",
      "/content/drive/My Drive/data/Localization/TrainTest/\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    from google.colab import drive\n",
    "    IN_COLAB = True\n",
    "    print('Running on CoLab')\n",
    "    PATH='/content/drive/'\n",
    "    drive.mount(PATH)\n",
    "    DATA_DIR=PATH+'My Drive/data/Localization/TrainTest/'  # must end in \"/\"\n",
    "except:\n",
    "    IN_COLAB = False\n",
    "    DATA_DIR = 'D:/Adjeroh/Localization/TrainTest/'   # Windows\n",
    "    DATA_DIR = '/Users/jasonmiller/WVU/Localization/TrainTest/'    # Mac\n",
    "print(DATA_DIR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "LnkpVKdMFLA-",
    "outputId": "7b79f7dd-2bf2-495f-d6a0-0e49448ad4a2"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: this run will write sequence to cache.\n"
     ]
    }
   ],
   "source": [
    "RCI_FILE =      'CNRCI_noncoding_train_RCI.gc42.csv'\n",
    "SEQUENCE_FILE = 'CNRCI_noncoding_train_transcripts.gc42.csv'\n",
    "CACHE_SEQUENCE = 'cache.seq.pickle'\n",
    "CACHE_IDS = 'cache.id.pickle'\n",
    "if CACHING:\n",
    "    cache_seq = DATA_DIR+CACHE_SEQUENCE \n",
    "    cache_ids = DATA_DIR+CACHE_IDS \n",
    "    if isfile(cache_seq) and isfile(cache_ids):\n",
    "        print('WARNING: this run will read sequence from cache.')\n",
    "    else:\n",
    "        print('INFO: this run will write sequence to cache.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "id": "e3p4QzQJFLA_"
   },
   "outputs": [],
   "source": [
    "def get_ordered_list():\n",
    "    ordered_list = \\\n",
    "    ['A549','H1.hESC','HeLa.S3','HepG2','HT1080',\\\n",
    "      'HUVEC','MCF.7','NCI.H460','NHEK','SK.MEL.5',\\\n",
    "      'SK.N.DZ','SK.N.SH','GM12878','K562','IMR.90']\n",
    "    return ordered_list\n",
    "all_cell_lines = get_ordered_list()\n",
    "\n",
    "EXCLUSIONS = [1,7]   # use these though they are clearly different"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "UtqdpJOxFLBA"
   },
   "source": [
    "## Data Load"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "id": "p35ehKV3Kq0z"
   },
   "outputs": [],
   "source": [
    "class DataLoader():\n",
    "    def __init__(self):\n",
    "        self.cache=dict() \n",
    "        self.vals = {'A':0, 'C':1, 'G':2, 'T':3}\n",
    "        \n",
    "    def load_gene_rci_values(self,filepath,cell_line):\n",
    "        '''\n",
    "        Load from RCI csv file.\n",
    "        Return dict with keys=gene:str and values=RCI:float.\n",
    "        '''\n",
    "        gene_to_rci = {}\n",
    "        overall_sum = 0\n",
    "        with open (filepath,'r') as handle:\n",
    "            header = None\n",
    "            for row in handle:\n",
    "                if header is None:\n",
    "                    header = row # skip file's header line\n",
    "                else:\n",
    "                    line = row.strip()\n",
    "                    fields = line.split(',')\n",
    "                    gene_id = fields.pop(0)\n",
    "                    rci_str = fields[cell_line]\n",
    "                    if rci_str != \"nan\":\n",
    "                        rci_val = float(rci_str)\n",
    "                        gene_to_rci[gene_id] = rci_val\n",
    "        print('Number of RCI values loaded',len(gene_to_rci.keys()))\n",
    "        return gene_to_rci\n",
    "\n",
    "    def get_threshold(self,cell_line,gene_to_rci):\n",
    "        return 0   # IN THIS EXPERIMENT, ALL POSITIVE RCI IS CYTO\n",
    "    \n",
    "    def seq_to_kmer_values(self,rna,K):\n",
    "        vec=[] # seq converted to list of K-mers \n",
    "        N_indicator = 0 # indicator value\n",
    "        length = len(rna)\n",
    "        for i in range(length-K+1):\n",
    "            kmer = rna[i:i+K]\n",
    "            if 'N' in kmer:\n",
    "                value = N_indicator\n",
    "            elif kmer in self.cache.keys():\n",
    "                value = self.cache[kmer]\n",
    "            else:\n",
    "                value = 0\n",
    "                for j in range(K):\n",
    "                    value *= 4   \n",
    "                    nextnuc = kmer[j] \n",
    "                    nucval = self.vals[nextnuc]\n",
    "                    value += nucval\n",
    "                value += 1   # NNN => 0, AAA => 1\n",
    "                self.cache[kmer] = value\n",
    "            vec.append(value)\n",
    "        return vec\n",
    "\n",
    "    def load_sequence(self,filepath):\n",
    "        allids=[]\n",
    "        allseq=[]\n",
    "        cache_seq = DATA_DIR+CACHE_SEQUENCE \n",
    "        cache_ids = DATA_DIR+CACHE_IDS \n",
    "        if CACHING and isfile(cache_seq) and isfile(cache_ids):\n",
    "            # Warning: the cache may represent the wrong K.\n",
    "            with open(cache_seq,'rb') as fin:\n",
    "                allseq = pickle.load(fin)\n",
    "            with open(cache_ids,'rb') as fin:\n",
    "                allids = pickle.load(fin)\n",
    "            return allids,allseq           \n",
    "        NREPEAT = str('N'*MAXLEN)\n",
    "        with open (filepath,'r') as handle:\n",
    "            header = None\n",
    "            for row in handle:\n",
    "                if header is None:\n",
    "                    header = row\n",
    "                else:\n",
    "                    line    = row.strip()\n",
    "                    fields  = line.split(',')\n",
    "                    tran_id = fields[0]  # with version number\n",
    "                    gene_id = fields[1]        # without version number\n",
    "                    seq_len = int(fields[3])\n",
    "                    seq_txt = fields[4]\n",
    "                    if seq_len>=MINLEN and seq_len<=MAXLEN and gene_id in gene_to_rci.keys():\n",
    "                        allids.append( (gene_id,tran_id) )\n",
    "                        if seq_len<MAXLEN:\n",
    "                            seq_txt = seq_txt + NREPEAT\n",
    "                            seq_txt = seq_txt[:MAXLEN]\n",
    "                        hot_vec = self.seq_to_kmer_values(seq_txt,K)\n",
    "                        allseq.append(hot_vec)\n",
    "        if CACHING:\n",
    "            with open(cache_seq,'wb') as fout:\n",
    "                pickle.dump(allseq,fout)\n",
    "            with open(cache_ids,'wb') as fout:\n",
    "                pickle.dump(allids,fout)\n",
    "        return allids,allseq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "VDZ6siB_Kq04"
   },
   "source": [
    "## Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "id": "AwMbRjm0FLBF"
   },
   "outputs": [],
   "source": [
    "def build_model():\n",
    "    ALPHABET=4**K+1  # NUMBER OF DISTINCT KMERS POSSIBLE, add one if N gets mask value\n",
    "    ADJUST_LENGTH = MAXLEN-K+1  # fixed length sequences\n",
    "    cnn = Sequential()\n",
    "    embed_layer = Embedding(ALPHABET,EMBED_DIMEN,input_length=ADJUST_LENGTH,mask_zero=True)   \n",
    "    cnn1_layer = Conv1D(CFILTERS, FILTERSIZE)\n",
    "    pool_layer = MaxPooling1D(pool_size=FILTERSIZE, strides=FILTERSIZE//2)\n",
    "    #pool_layer = GlobalAveragePooling1D()\n",
    "    flat_layer = Flatten()\n",
    "    dens_layer = Dense(DCELLS)\n",
    "    drop_layer = Dropout(0.5)\n",
    "    output_layer = Dense(1,activation='sigmoid',dtype=dt)\n",
    "\n",
    "    cnn.add(embed_layer)\n",
    "    cnn.add(cnn1_layer)\n",
    "    cnn.add(pool_layer)\n",
    "    cnn.add(flat_layer)\n",
    "    cnn.add(dens_layer)\n",
    "    cnn.add(drop_layer)\n",
    "    cnn.add(output_layer)\n",
    "\n",
    "    bc=BinaryCrossentropy(from_logits=False)\n",
    "    cnn.compile(loss=bc, optimizer=\"Adam\",metrics=[\"accuracy\"])\n",
    "    return cnn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "clj-wufgFLBF",
    "outputId": "2127141f-4275-49b0-9f8b-129f8fa10aa9"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2023-02-05 18:45:45.446046\n",
      "Model: \"sequential\"\n",
      "_________________________________________________________________\n",
      " Layer (type)                Output Shape              Param #   \n",
      "=================================================================\n",
      " embedding (Embedding)       (None, 4997, 4)           1028      \n",
      "                                                                 \n",
      " conv1d (Conv1D)             (None, 4982, 64)          4160      \n",
      "                                                                 \n",
      " max_pooling1d (MaxPooling1D  (None, 621, 64)          0         \n",
      " )                                                               \n",
      "                                                                 \n",
      " flatten (Flatten)           (None, 39744)             0         \n",
      "                                                                 \n",
      " dense (Dense)               (None, 16)                635920    \n",
      "                                                                 \n",
      " dropout (Dropout)           (None, 16)                0         \n",
      "                                                                 \n",
      " dense_1 (Dense)             (None, 1)                 17        \n",
      "                                                                 \n",
      "=================================================================\n",
      "Total params: 641,125\n",
      "Trainable params: 641,125\n",
      "Non-trainable params: 0\n",
      "_________________________________________________________________\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "print(datetime.now())\n",
    "model=build_model()\n",
    "print(model.summary())  # Print this only once\n",
    "model=None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "MgrC1alOKq07"
   },
   "source": [
    "## Training"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "id": "W9xiFzNbFLBE"
   },
   "outputs": [],
   "source": [
    "import gc\n",
    "class CrossValidator():\n",
    "    def __init__(self,epochs,folds,quick_test=False,score_threshold=0.5):\n",
    "        self.epochs = epochs\n",
    "        self.folds = folds\n",
    "        self.quick_test = quick_test\n",
    "        self.score_threshold = score_threshold\n",
    "        \n",
    "    def set_sequences(self, allids, allseq):\n",
    "        self.all_ids = allids\n",
    "        self.all_seqs = allseq\n",
    "        \n",
    "    def set_rci_map(self, gene_to_rci):\n",
    "        self.rci_map = gene_to_rci\n",
    "    \n",
    "    def get_gene_subset(self, all_genes, sub_index):\n",
    "        sub_genes = set()\n",
    "        for index in sub_index:\n",
    "            one_gene = all_genes[index]\n",
    "            sub_genes.add(one_gene)\n",
    "        return sub_genes\n",
    "    \n",
    "    def get_X_y(self, gene_set):\n",
    "        cnt = len(self.all_ids)\n",
    "        subsetX=[]\n",
    "        subsetY=[]\n",
    "        for i in range(cnt):\n",
    "            gene_id,tran_id = self.all_ids[i]\n",
    "            if gene_id in gene_set:\n",
    "                oneX = self.all_seqs[i]\n",
    "                oneY = self.rci_map[gene_id]\n",
    "                subsetX.append(oneX)\n",
    "                subsetY.append(oneY)\n",
    "        subsetX = np.array(subsetX)\n",
    "        subsetY = np.array(subsetY).reshape((-1,1))\n",
    "        return subsetX,subsetY\n",
    "    \n",
    "    def rcis_to_labels(self, array_of_rci, rci_threshold):\n",
    "        IS_CYTO = lambda rci: 1 if rci>rci_threshold else 0\n",
    "        return np.array(list(map(IS_CYTO, array_of_rci)))\n",
    "    \n",
    "    def choose_rci_threshold(self, rci_values):\n",
    "        #if RCI_THRESHOLD_MECHANISM == 'Zero':\n",
    "        return 0\n",
    "        \n",
    "    def do_cross_validation(self):\n",
    "        cv_accuracy=[]\n",
    "        cv_precision=[]\n",
    "        cv_recall=[]\n",
    "        cv_f1=[]\n",
    "        cv_auprc=[]\n",
    "        cv_auroc=[]\n",
    "        fold=0\n",
    "        print(datetime.now())\n",
    "        print('splitting')\n",
    "        # KFold shuffles once before making the partitions\n",
    "        splitter = KFold(n_splits=self.folds,shuffle=True,random_state=42)\n",
    "        splits = splitter.split(all_genes)\n",
    "        splitter = None\n",
    "        for train_index,valid_index in splits:\n",
    "            fold += 1\n",
    "            print('Fold',fold)\n",
    "            train_genes = self.get_gene_subset(all_genes,train_index)\n",
    "            X_train,y_train = self.get_X_y(train_genes)\n",
    "            train_based_rci_threshold = self.choose_rci_threshold(y_train)\n",
    "            y_train = self.rcis_to_labels(y_train,train_based_rci_threshold)\n",
    "            valid_genes = self.get_gene_subset(all_genes,valid_index)\n",
    "            X_valid,y_valid = self.get_X_y(valid_genes)\n",
    "            y_valid = self.rcis_to_labels(y_valid,train_based_rci_threshold) # yes, train_based\n",
    "\n",
    "            #print('Training example')\n",
    "            #print(X_train[0])\n",
    "            #print(\"BUILD MODEL\")\n",
    "            model=build_model()\n",
    "            print(\"FIT\")\n",
    "            print(datetime.now())\n",
    "            history=model.fit(X_train, y_train, # batch_size=10, default=32 works nicely\n",
    "              epochs=self.epochs, verbose=0,  # verbose=1 for ascii art, verbose=0 for none\n",
    "              validation_data=(X_valid,y_valid) )\n",
    "            print(\"PREDICT\")\n",
    "            print(datetime.now())\n",
    "            yhat_pred=model.predict(X_valid, verbose=0)             \n",
    "            yhat_classes=np.where(yhat_pred > self.score_threshold, 1, 0)\n",
    "\n",
    "            # accuracy: (tp + tn) / (p + n)\n",
    "            accuracy = accuracy_score(y_valid, yhat_classes)*100.\n",
    "            # precision tp / (tp + fp)\n",
    "            precision = precision_score(y_valid, yhat_classes)*100.\n",
    "            # recall: tp / (tp + fn)\n",
    "            recall = recall_score(y_valid, yhat_classes)*100.\n",
    "            # f1: 2 tp / (2 tp + fp + fn)\n",
    "            f1 = f1_score(y_valid, yhat_classes)*100.\n",
    "            # PRC\n",
    "            prc_Y, prc_X, prc_bins = precision_recall_curve(y_valid, yhat_pred)\n",
    "            auprc = auc(prc_X,prc_Y)*100.\n",
    "            auroc = roc_auc_score(y_valid, yhat_pred)*100.\n",
    "\n",
    "            if PRETTY_PICTURES:\n",
    "                pd.DataFrame(history.history).plot(figsize=(8,5))\n",
    "                plt.grid(True)\n",
    "                plt.gca().set_ylim(0,1)\n",
    "                plt.show()\n",
    "\n",
    "                print('Train set ones/size',\n",
    "                      np.count_nonzero(y_train),'/',len(y_train))\n",
    "                print(\"Compute valiation accuracy\")\n",
    "                print('Valid sizes',X_valid.shape,y_valid.shape)\n",
    "                print('Valid set ones/size',\n",
    "                      np.count_nonzero(y_valid),'/',len(y_valid))\n",
    "                print('Range of scores:',np.min(yhat_pred),'to',np.max(yhat_pred))\n",
    "                print('Score threshold',self.score_threshold)\n",
    "                print('Prediction set ones/size',\n",
    "                      np.count_nonzero(yhat_classes),'/',len(yhat_classes))\n",
    "            \n",
    "                count_ones= len(y_valid[y_valid==1])\n",
    "                count_zeros= len(y_valid[y_valid==0])\n",
    "                guess = max(count_ones,count_zeros) / len(y_valid)\n",
    "                plt.plot(prc_X, prc_Y, marker='.')\n",
    "                plt.plot([0, 1], [guess,guess], linestyle='--')\n",
    "                plt.xlabel('Recall')\n",
    "                plt.ylabel('Precision')\n",
    "                plt.show()\n",
    "                # ROC\n",
    "                fpr, tpr, roc_bins = roc_curve(y_valid, yhat_pred)\n",
    "                plt.plot(fpr, tpr, marker='.')\n",
    "                plt.xlabel('False Positive Rate')\n",
    "                plt.ylabel('True Positive Rate')\n",
    "                plt.show()\n",
    "            \n",
    "            print('Accuracy: %.2f%% Precision: %.2f%% Recall: %.2f%%' % (accuracy,precision,recall)) \n",
    "            print('F1: %.2f%% AUPRC: %.2f%% AUROC: %.2f%%' % (f1,auprc,auroc)) \n",
    "            cv_accuracy.append(accuracy)\n",
    "            cv_precision.append(precision)\n",
    "            cv_recall.append(recall)\n",
    "            cv_f1.append(f1)\n",
    "            cv_auprc.append(auprc)\n",
    "            cv_auroc.append(auroc)\n",
    "            \n",
    "            print(datetime.now())\n",
    "            if self.quick_test:   \n",
    "                print('Break -- this was for code testing only')\n",
    "                break\n",
    "\n",
    "            # There is a memory leak within the fit() command!\n",
    "            # Each successive call to fit() consumes more memory.\n",
    "            model = None\n",
    "            history = None\n",
    "            prc_Y = None\n",
    "            prc_X = None\n",
    "            prc_bins = None\n",
    "            yhat_classes = None\n",
    "            X_train = None\n",
    "            y_train = None\n",
    "            train_genes = None\n",
    "            X_valid = None\n",
    "            y_valid = None\n",
    "            valid_genes = None\n",
    "            tf.keras.backend.clear_session()\n",
    "            gc.collect()\n",
    "            time.sleep(1)  # hope gc kicks in\n",
    "        print()\n",
    "        return cv_accuracy, cv_precision, cv_recall, cv_f1, cv_auprc, cv_auroc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "XC9m0W-pFLBH",
    "outputId": "ea9141c0-bf22-4cbd-f79f-86a913f931e2",
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2023-02-05 18:45:48.851244\n",
      "Start cell Line 0 A549\n",
      "Number of RCI values loaded 1447\n",
      "Num RCI: 1447\n",
      "Example RCI: [('ENSG00000116652', -1.848), ('ENSG00000117242', -0.25673), ('ENSG00000125462', 0.13734)]\n",
      "RCI threshold 0\n",
      "Load sequence...\n",
      "Num IDs: 8687\n",
      "Cross validation...\n",
      "2023-02-05 18:46:05.344189\n",
      "splitting\n",
      "Fold 1\n",
      "FIT\n",
      "2023-02-05 18:46:07.710391\n",
      "PREDICT\n",
      "2023-02-05 18:46:23.662112\n",
      "Accuracy: 52.15% Precision: 51.87% Recall: 76.98%\n",
      "F1: 61.98% AUPRC: 55.90% AUROC: 54.78%\n",
      "2023-02-05 18:46:24.115196\n",
      "Fold 2\n",
      "FIT\n",
      "2023-02-05 18:46:28.098991\n",
      "PREDICT\n",
      "2023-02-05 18:46:36.123653\n",
      "Accuracy: 45.09% Precision: 46.29% Recall: 25.11%\n",
      "F1: 32.56% AUPRC: 49.51% AUROC: 48.22%\n",
      "2023-02-05 18:46:36.586624\n",
      "Fold 3\n",
      "FIT\n",
      "2023-02-05 18:46:40.305454\n",
      "PREDICT\n",
      "2023-02-05 18:46:51.501381\n",
      "Accuracy: 50.93% Precision: 60.93% Recall: 36.00%\n",
      "F1: 45.26% AUPRC: 58.65% AUROC: 54.68%\n",
      "2023-02-05 18:46:51.953052\n",
      "Fold 4\n",
      "FIT\n",
      "2023-02-05 18:46:56.514155\n",
      "PREDICT\n",
      "2023-02-05 18:47:04.541981\n",
      "Accuracy: 53.80% Precision: 47.21% Recall: 52.64%\n",
      "F1: 49.78% AUPRC: 48.86% AUROC: 57.33%\n",
      "2023-02-05 18:47:04.978834\n",
      "Fold 5\n",
      "FIT\n",
      "2023-02-05 18:47:09.119580\n",
      "PREDICT\n",
      "2023-02-05 18:47:20.390749\n",
      "Accuracy: 57.72% Precision: 64.55% Recall: 70.40%\n",
      "F1: 67.35% AUPRC: 65.28% AUROC: 54.35%\n",
      "2023-02-05 18:47:20.693470\n",
      "\n",
      "Completed cross validation 5 folds 3 epochs\n",
      " accuracy mean 51.94% +/- 4.12\n",
      " precision mean 54.17% +/- 7.34\n",
      " recall mean 52.23% +/- 19.71\n",
      " F1 mean 51.38% +/- 12.34\n",
      " AUPRC mean 55.64% +/- 6.09\n",
      " AUROC mean 53.87% +/- 3.02\n",
      "Finished cell Line 0 A549\n",
      "\n",
      "2023-02-05 18:47:22.061106\n",
      "Start cell Line 1 H1.hESC\n",
      "Number of RCI values loaded 3327\n",
      "Num RCI: 3327\n",
      "Example RCI: [('ENSG00000099869', 1.0), ('ENSG00000116652', 1.65208), ('ENSG00000117242', -0.793877)]\n",
      "RCI threshold 0\n",
      "Load sequence...\n",
      "Num IDs: 8687\n",
      "Cross validation...\n",
      "2023-02-05 18:47:23.779885\n",
      "splitting\n",
      "Fold 1\n",
      "FIT\n",
      "2023-02-05 18:47:26.543873\n",
      "PREDICT\n",
      "2023-02-05 18:47:37.587590\n",
      "Accuracy: 60.69% Precision: 60.00% Recall: 41.78%\n",
      "F1: 49.26% AUPRC: 59.39% AUROC: 64.95%\n",
      "2023-02-05 18:47:37.946428\n",
      "Fold 2\n",
      "FIT\n",
      "2023-02-05 18:47:41.457566\n",
      "PREDICT\n",
      "2023-02-05 18:47:52.562998\n",
      "Accuracy: 58.61% Precision: 38.79% Recall: 33.11%\n",
      "F1: 35.72% AUPRC: 38.68% AUROC: 55.45%\n",
      "2023-02-05 18:47:53.002821\n",
      "Fold 3\n",
      "FIT\n",
      "2023-02-05 18:47:56.459475\n",
      "PREDICT\n",
      "2023-02-05 18:48:07.408228\n",
      "Accuracy: 58.64% Precision: 30.11% Recall: 37.68%\n",
      "F1: 33.47% AUPRC: 33.72% AUROC: 54.96%\n",
      "2023-02-05 18:48:07.776524\n",
      "Fold 4\n",
      "FIT\n",
      "2023-02-05 18:48:11.204654\n",
      "PREDICT\n",
      "2023-02-05 18:48:22.177562\n",
      "Accuracy: 57.94% Precision: 31.18% Recall: 20.99%\n",
      "F1: 25.09% AUPRC: 33.44% AUROC: 50.96%\n",
      "2023-02-05 18:48:22.597761\n",
      "Fold 5\n",
      "FIT\n",
      "2023-02-05 18:48:26.065911\n",
      "PREDICT\n",
      "2023-02-05 18:48:33.450641\n",
      "Accuracy: 59.33% Precision: 62.84% Recall: 26.07%\n",
      "F1: 36.85% AUPRC: 58.80% AUROC: 64.42%\n",
      "2023-02-05 18:48:33.887467\n",
      "\n",
      "Completed cross validation 5 folds 3 epochs\n",
      " accuracy mean 59.04% +/- 0.93\n",
      " precision mean 44.58% +/- 14.10\n",
      " recall mean 31.92% +/- 7.56\n",
      " F1 mean 36.08% +/- 7.77\n",
      " AUPRC mean 44.81% +/- 11.82\n",
      " AUROC mean 58.15% +/- 5.56\n",
      "Finished cell Line 1 H1.hESC\n",
      "\n",
      "2023-02-05 18:48:35.253756\n",
      "Start cell Line 2 HeLa.S3\n",
      "Number of RCI values loaded 919\n",
      "Num RCI: 919\n",
      "Example RCI: [('ENSG00000117242', -0.0703893), ('ENSG00000125804', -2.58496), ('ENSG00000128254', -0.423808)]\n",
      "RCI threshold 0\n",
      "Load sequence...\n",
      "Num IDs: 8687\n",
      "Cross validation...\n",
      "2023-02-05 18:48:36.478614\n",
      "splitting\n",
      "Fold 1\n",
      "FIT\n",
      "2023-02-05 18:48:37.600280\n",
      "PREDICT\n",
      "2023-02-05 18:48:41.797956\n",
      "Accuracy: 70.19% Precision: 22.86% Recall: 9.36%\n",
      "F1: 13.28% AUPRC: 26.57% AUROC: 55.46%\n",
      "2023-02-05 18:48:42.084434\n",
      "Fold 2\n",
      "FIT\n",
      "2023-02-05 18:48:44.633848\n",
      "PREDICT\n",
      "2023-02-05 18:48:50.428979\n",
      "Accuracy: 77.97% Precision: 64.29% Recall: 18.95%\n",
      "F1: 29.27% AUPRC: 43.84% AUROC: 70.48%\n",
      "2023-02-05 18:48:50.674184\n",
      "Fold 3\n",
      "FIT\n",
      "2023-02-05 18:48:53.116299\n",
      "PREDICT\n",
      "2023-02-05 18:48:59.025785\n",
      "Accuracy: 59.09% Precision: 11.80% Recall: 23.12%\n",
      "F1: 15.62% AUPRC: 13.47% AUROC: 40.02%\n",
      "2023-02-05 18:48:59.447379\n",
      "Fold 4\n",
      "FIT\n",
      "2023-02-05 18:49:01.847589\n",
      "PREDICT\n",
      "2023-02-05 18:49:06.024550\n",
      "Accuracy: 89.73% Precision: 4.76% Recall: 2.63%\n",
      "F1: 3.39% AUPRC: 7.17% AUROC: 47.87%\n",
      "2023-02-05 18:49:06.268419\n",
      "Fold 5\n",
      "FIT\n",
      "2023-02-05 18:49:09.461090\n",
      "PREDICT\n",
      "2023-02-05 18:49:15.416957\n",
      "Accuracy: 83.29% Precision: 92.00% Recall: 16.67%\n",
      "F1: 28.22% AUPRC: 40.62% AUROC: 56.36%\n",
      "2023-02-05 18:49:15.614343\n",
      "\n",
      "Completed cross validation 5 folds 3 epochs\n",
      " accuracy mean 76.05% +/- 10.63\n",
      " precision mean 39.14% +/- 33.53\n",
      " recall mean 14.14% +/- 7.29\n",
      " F1 mean 17.96% +/- 9.72\n",
      " AUPRC mean 26.33% +/- 14.44\n",
      " AUROC mean 54.04% +/- 10.12\n",
      "Finished cell Line 2 HeLa.S3\n",
      "\n",
      "2023-02-05 18:49:16.979195\n",
      "Start cell Line 3 HepG2\n",
      "Number of RCI values loaded 1362\n",
      "Num RCI: 1362\n",
      "Example RCI: [('ENSG00000099869', 0.00846158), ('ENSG00000116652', -4.42626), ('ENSG00000117242', 0.464641)]\n",
      "RCI threshold 0\n",
      "Load sequence...\n",
      "Num IDs: 8687\n",
      "Cross validation...\n",
      "2023-02-05 18:49:18.130739\n",
      "splitting\n",
      "Fold 1\n",
      "FIT\n",
      "2023-02-05 18:49:19.745564\n",
      "PREDICT\n",
      "2023-02-05 18:49:25.881084\n",
      "Accuracy: 72.53% Precision: 22.41% Recall: 6.63%\n",
      "F1: 10.24% AUPRC: 23.88% AUROC: 52.05%\n",
      "2023-02-05 18:49:26.126998\n",
      "Fold 2\n",
      "FIT\n",
      "2023-02-05 18:49:28.993841\n",
      "PREDICT\n",
      "2023-02-05 18:49:34.625712\n",
      "Accuracy: 68.21% Precision: 34.88% Recall: 27.68%\n",
      "F1: 30.86% AUPRC: 31.82% AUROC: 57.36%\n",
      "2023-02-05 18:49:34.947229\n",
      "Fold 3\n",
      "FIT\n",
      "2023-02-05 18:49:38.040944\n",
      "PREDICT\n",
      "2023-02-05 18:49:43.817096\n",
      "Accuracy: 72.89% Precision: 39.65% Recall: 22.50%\n",
      "F1: 28.71% AUPRC: 31.86% AUROC: 52.89%\n",
      "2023-02-05 18:49:44.242382\n",
      "Fold 4\n",
      "FIT\n",
      "2023-02-05 18:49:47.458114\n",
      "PREDICT\n",
      "2023-02-05 18:49:53.510986\n",
      "Accuracy: 61.25% Precision: 35.14% Recall: 33.05%\n",
      "F1: 34.06% AUPRC: 36.00% AUROC: 54.50%\n",
      "2023-02-05 18:49:53.770144\n",
      "Fold 5\n",
      "FIT\n",
      "2023-02-05 18:49:56.649448\n",
      "PREDICT\n",
      "2023-02-05 18:50:07.586382\n",
      "Accuracy: 33.66% Precision: 11.73% Recall: 62.81%\n",
      "F1: 19.77% AUPRC: 13.13% AUROC: 47.09%\n",
      "2023-02-05 18:50:07.832915\n",
      "\n",
      "Completed cross validation 5 folds 3 epochs\n",
      " accuracy mean 61.71% +/- 14.64\n",
      " precision mean 28.76% +/- 10.27\n",
      " recall mean 30.53% +/- 18.40\n",
      " F1 mean 24.73% +/- 8.66\n",
      " AUPRC mean 27.34% +/- 8.11\n",
      " AUROC mean 52.78% +/- 3.37\n",
      "Finished cell Line 3 HepG2\n",
      "\n",
      "2023-02-05 18:50:09.187764\n",
      "Start cell Line 4 HT1080\n",
      "Number of RCI values loaded 936\n",
      "Num RCI: 936\n",
      "Example RCI: [('ENSG00000105501', -0.415038), ('ENSG00000117242', -0.0596429), ('ENSG00000153363', 2.00657)]\n",
      "RCI threshold 0\n",
      "Load sequence...\n",
      "Num IDs: 8687\n",
      "Cross validation...\n",
      "2023-02-05 18:50:10.349544\n",
      "splitting\n",
      "Fold 1\n",
      "FIT\n",
      "2023-02-05 18:50:11.780540\n",
      "PREDICT\n",
      "2023-02-05 18:50:17.331417\n",
      "Accuracy: 62.04% Precision: 37.96% Recall: 29.71%\n",
      "F1: 33.33% AUPRC: 36.24% AUROC: 55.90%\n",
      "2023-02-05 18:50:17.558740\n",
      "Fold 2\n",
      "FIT\n",
      "2023-02-05 18:50:20.284721\n",
      "PREDICT\n",
      "2023-02-05 18:50:26.065878\n",
      "Accuracy: 54.72% Precision: 68.58% Recall: 64.59%\n",
      "F1: 66.52% AUPRC: 71.12% AUROC: 50.55%\n",
      "2023-02-05 18:50:26.322545\n",
      "Fold 3\n",
      "FIT\n",
      "2023-02-05 18:50:29.971677\n",
      "PREDICT\n",
      "2023-02-05 18:50:35.745847\n",
      "Accuracy: 58.23% Precision: 39.48% Recall: 47.71%\n",
      "F1: 43.21% AUPRC: 39.76% AUROC: 55.86%\n",
      "2023-02-05 18:50:35.996668\n",
      "Fold 4\n",
      "FIT\n",
      "2023-02-05 18:50:38.725152\n",
      "PREDICT\n",
      "2023-02-05 18:50:44.372700\n",
      "Accuracy: 53.93% Precision: 39.23% Recall: 67.92%\n",
      "F1: 49.74% AUPRC: 45.65% AUROC: 62.69%\n",
      "2023-02-05 18:50:44.627828\n",
      "Fold 5\n",
      "FIT\n",
      "2023-02-05 18:50:47.329107\n",
      "PREDICT\n",
      "2023-02-05 18:50:52.350742\n",
      "Accuracy: 53.11% Precision: 48.12% Recall: 25.05%\n",
      "F1: 32.95% AUPRC: 48.20% AUROC: 51.38%\n",
      "2023-02-05 18:50:52.587450\n",
      "\n",
      "Completed cross validation 5 folds 3 epochs\n",
      " accuracy mean 56.41% +/- 3.32\n",
      " precision mean 46.67% +/- 11.53\n",
      " recall mean 47.00% +/- 17.48\n",
      " F1 mean 45.15% +/- 12.41\n",
      " AUPRC mean 48.19% +/- 12.21\n",
      " AUROC mean 55.28% +/- 4.32\n",
      "Finished cell Line 4 HT1080\n",
      "\n",
      "2023-02-05 18:50:53.957533\n",
      "Start cell Line 5 HUVEC\n",
      "Number of RCI values loaded 1488\n",
      "Num RCI: 1488\n",
      "Example RCI: [('ENSG00000117242', 0.787885), ('ENSG00000120664', -1.76017), ('ENSG00000128254', 0.459432)]\n",
      "RCI threshold 0\n",
      "Load sequence...\n",
      "Num IDs: 8687\n",
      "Cross validation...\n",
      "2023-02-05 18:50:55.718622\n",
      "splitting\n",
      "Fold 1\n",
      "FIT\n",
      "2023-02-05 18:50:57.441868\n",
      "PREDICT\n",
      "2023-02-05 18:51:03.214602\n",
      "Accuracy: 72.06% Precision: 47.73% Recall: 10.97%\n",
      "F1: 17.83% AUPRC: 40.53% AUROC: 65.86%\n",
      "2023-02-05 18:51:03.632413\n",
      "Fold 2\n",
      "FIT\n",
      "2023-02-05 18:51:07.044409\n",
      "PREDICT\n",
      "2023-02-05 18:51:18.232562\n",
      "Accuracy: 72.41% Precision: 56.34% Recall: 11.80%\n",
      "F1: 19.51% AUPRC: 42.82% AUROC: 65.76%\n",
      "2023-02-05 18:51:18.509118\n",
      "Fold 3\n",
      "FIT\n",
      "2023-02-05 18:51:22.725360\n",
      "PREDICT\n",
      "2023-02-05 18:51:28.577212\n",
      "Accuracy: 55.01% Precision: 20.60% Recall: 50.93%\n",
      "F1: 29.34% AUPRC: 20.59% AUROC: 54.11%\n",
      "2023-02-05 18:51:28.832173\n",
      "Fold 4\n",
      "FIT\n",
      "2023-02-05 18:51:31.766683\n",
      "PREDICT\n",
      "2023-02-05 18:51:42.704068\n",
      "Accuracy: 64.06% Precision: 29.69% Recall: 5.64%\n",
      "F1: 9.48% AUPRC: 31.43% AUROC: 47.05%\n",
      "2023-02-05 18:51:43.111480\n",
      "Fold 5\n",
      "FIT\n",
      "2023-02-05 18:51:46.075337\n",
      "PREDICT\n",
      "2023-02-05 18:51:57.320202\n",
      "Accuracy: 72.08% Precision: 62.50% Recall: 3.12%\n",
      "F1: 5.95% AUPRC: 32.37% AUROC: 52.00%\n",
      "2023-02-05 18:51:57.604388\n",
      "\n",
      "Completed cross validation 5 folds 3 epochs\n",
      " accuracy mean 67.12% +/- 6.83\n",
      " precision mean 43.37% +/- 15.87\n",
      " recall mean 16.49% +/- 17.52\n",
      " F1 mean 16.42% +/- 8.20\n",
      " AUPRC mean 33.55% +/- 7.85\n",
      " AUROC mean 56.95% +/- 7.59\n",
      "Finished cell Line 5 HUVEC\n",
      "\n",
      "2023-02-05 18:51:58.971462\n",
      "Start cell Line 6 MCF.7\n",
      "Number of RCI values loaded 2156\n",
      "Num RCI: 2156\n",
      "Example RCI: [('ENSG00000116652', -4.88753), ('ENSG00000117242', -0.892197), ('ENSG00000132832', -1.95042)]\n",
      "RCI threshold 0\n",
      "Load sequence...\n",
      "Num IDs: 8687\n",
      "Cross validation...\n",
      "2023-02-05 18:52:00.412062\n",
      "splitting\n",
      "Fold 1\n",
      "FIT\n",
      "2023-02-05 18:52:03.539299\n",
      "PREDICT\n",
      "2023-02-05 18:52:14.510602\n",
      "Accuracy: 71.96% Precision: 26.45% Recall: 30.48%\n",
      "F1: 28.32% AUPRC: 27.78% AUROC: 61.43%\n",
      "2023-02-05 18:52:14.978992\n",
      "Fold 2\n",
      "FIT\n",
      "2023-02-05 18:52:18.325843\n",
      "PREDICT\n",
      "2023-02-05 18:52:24.669393\n",
      "Accuracy: 65.13% Precision: 31.87% Recall: 34.26%\n",
      "F1: 33.02% AUPRC: 35.16% AUROC: 59.12%\n",
      "2023-02-05 18:52:25.087160\n",
      "Fold 3\n",
      "FIT\n",
      "2023-02-05 18:52:29.376251\n",
      "PREDICT\n",
      "2023-02-05 18:52:40.320153\n",
      "Accuracy: 67.79% Precision: 27.12% Recall: 2.88%\n",
      "F1: 5.20% AUPRC: 31.34% AUROC: 52.68%\n",
      "2023-02-05 18:52:40.788475\n",
      "Fold 4\n",
      "FIT\n",
      "2023-02-05 18:52:44.070511\n",
      "PREDICT\n",
      "2023-02-05 18:52:50.730001\n",
      "Accuracy: 67.33% Precision: 31.71% Recall: 3.78%\n",
      "F1: 6.75% AUPRC: 36.90% AUROC: 59.54%\n",
      "2023-02-05 18:52:50.979690\n",
      "Fold 5\n",
      "FIT\n",
      "2023-02-05 18:52:55.089141\n",
      "PREDICT\n",
      "2023-02-05 18:53:01.485793\n",
      "Accuracy: 67.71% Precision: 34.49% Recall: 42.67%\n",
      "F1: 38.15% AUPRC: 33.63% AUROC: 60.65%\n",
      "2023-02-05 18:53:01.740667\n",
      "\n",
      "Completed cross validation 5 folds 3 epochs\n",
      " accuracy mean 67.98% +/- 2.21\n",
      " precision mean 30.33% +/- 3.06\n",
      " recall mean 22.81% +/- 16.40\n",
      " F1 mean 22.29% +/- 13.69\n",
      " AUPRC mean 32.96% +/- 3.17\n",
      " AUROC mean 58.68% +/- 3.11\n",
      "Finished cell Line 6 MCF.7\n",
      "\n",
      "2023-02-05 18:53:03.116568\n",
      "Start cell Line 7 NCI.H460\n",
      "Number of RCI values loaded 610\n",
      "Num RCI: 610\n",
      "Example RCI: [('ENSG00000116652', -4.98868), ('ENSG00000117242', 0.229026), ('ENSG00000125462', -3.32593)]\n",
      "RCI threshold 0\n",
      "Load sequence...\n",
      "Num IDs: 8687\n",
      "Cross validation...\n",
      "2023-02-05 18:53:04.231709\n",
      "splitting\n",
      "Fold 1\n",
      "FIT\n",
      "2023-02-05 18:53:06.152080\n",
      "PREDICT\n",
      "2023-02-05 18:53:10.755224\n",
      "Accuracy: 85.08% Precision: 25.78% Recall: 30.28%\n",
      "F1: 27.85% AUPRC: 18.05% AUROC: 67.57%\n",
      "2023-02-05 18:53:11.003932\n",
      "Fold 2\n",
      "FIT\n",
      "2023-02-05 18:53:13.488753\n",
      "PREDICT\n",
      "2023-02-05 18:53:17.842645\n",
      "Accuracy: 67.30% Precision: 5.00% Recall: 2.98%\n",
      "F1: 3.73% AUPRC: 16.48% AUROC: 34.02%\n",
      "2023-02-05 18:53:18.096143\n",
      "Fold 3\n",
      "FIT\n",
      "2023-02-05 18:53:21.432475\n",
      "PREDICT\n",
      "2023-02-05 18:53:27.130075\n",
      "Accuracy: 90.94% Precision: 9.38% Recall: 6.38%\n",
      "F1: 7.59% AUPRC: 10.90% AUROC: 71.45%\n",
      "2023-02-05 18:53:27.386337\n",
      "Fold 4\n",
      "FIT\n",
      "2023-02-05 18:53:29.884270\n",
      "PREDICT\n",
      "2023-02-05 18:53:34.686831\n",
      "Accuracy: 59.83% Precision: 27.42% Recall: 6.44%\n",
      "F1: 10.43% AUPRC: 36.48% AUROC: 50.37%\n",
      "2023-02-05 18:53:34.934719\n",
      "Fold 5\n",
      "FIT\n",
      "2023-02-05 18:53:37.493818\n",
      "PREDICT\n",
      "2023-02-05 18:53:41.873782\n",
      "Accuracy: 75.59% Precision: 11.76% Recall: 4.08%\n",
      "F1: 6.06% AUPRC: 16.06% AUROC: 40.95%\n",
      "2023-02-05 18:53:42.066424\n",
      "\n",
      "Completed cross validation 5 folds 3 epochs\n",
      " accuracy mean 75.75% +/- 11.34\n",
      " precision mean 15.87% +/- 9.04\n",
      " recall mean 10.03% +/- 10.21\n",
      " F1 mean 11.13% +/- 8.64\n",
      " AUPRC mean 19.59% +/- 8.78\n",
      " AUROC mean 52.87% +/- 14.59\n",
      "Finished cell Line 7 NCI.H460\n",
      "\n",
      "2023-02-05 18:53:43.426681\n",
      "Start cell Line 8 NHEK\n",
      "Number of RCI values loaded 1080\n",
      "Num RCI: 1080\n",
      "Example RCI: [('ENSG00000116652', -2.97982), ('ENSG00000117242', -1.277), ('ENSG00000153363', 2.27892)]\n",
      "RCI threshold 0\n",
      "Load sequence...\n",
      "Num IDs: 8687\n",
      "Cross validation...\n",
      "2023-02-05 18:53:44.563404\n",
      "splitting\n",
      "Fold 1\n",
      "FIT\n",
      "2023-02-05 18:53:47.102936\n",
      "PREDICT\n",
      "2023-02-05 18:53:52.791418\n",
      "Accuracy: 68.63% Precision: 58.47% Recall: 16.87%\n",
      "F1: 26.19% AUPRC: 45.15% AUROC: 62.22%\n",
      "2023-02-05 18:53:53.050183\n",
      "Fold 2\n",
      "FIT\n",
      "2023-02-05 18:53:55.800985\n",
      "PREDICT\n",
      "2023-02-05 18:54:01.270043\n",
      "Accuracy: 63.97% Precision: 30.59% Recall: 38.66%\n",
      "F1: 34.15% AUPRC: 30.33% AUROC: 54.83%\n",
      "2023-02-05 18:54:01.723155\n",
      "Fold 3\n",
      "FIT\n",
      "2023-02-05 18:54:04.460151\n",
      "PREDICT\n",
      "2023-02-05 18:54:09.431747\n",
      "Accuracy: 62.50% Precision: 18.86% Recall: 12.13%\n",
      "F1: 14.77% AUPRC: 24.31% AUROC: 48.30%\n",
      "2023-02-05 18:54:09.680654\n",
      "Fold 4\n",
      "FIT\n",
      "2023-02-05 18:54:13.060428\n",
      "PREDICT\n",
      "2023-02-05 18:54:18.579356\n",
      "Accuracy: 64.64% Precision: 40.00% Recall: 0.68%\n",
      "F1: 1.33% AUPRC: 32.18% AUROC: 48.24%\n",
      "2023-02-05 18:54:18.779625\n",
      "Fold 5\n",
      "FIT\n",
      "2023-02-05 18:54:21.535437\n",
      "PREDICT\n",
      "2023-02-05 18:54:27.042539\n",
      "Accuracy: 52.91% Precision: 36.43% Recall: 22.60%\n",
      "F1: 27.89% AUPRC: 37.18% AUROC: 41.98%\n",
      "2023-02-05 18:54:27.488073\n",
      "\n",
      "Completed cross validation 5 folds 3 epochs\n",
      " accuracy mean 62.53% +/- 5.22\n",
      " precision mean 36.87% +/- 12.97\n",
      " recall mean 18.19% +/- 12.52\n",
      " F1 mean 20.87% +/- 11.60\n",
      " AUPRC mean 33.83% +/- 7.00\n",
      " AUROC mean 51.12% +/- 6.88\n",
      "Finished cell Line 8 NHEK\n",
      "\n",
      "2023-02-05 18:54:28.978245\n",
      "Start cell Line 9 SK.MEL.5\n",
      "Number of RCI values loaded 534\n",
      "Num RCI: 534\n",
      "Example RCI: [('ENSG00000136275', -4.19162), ('ENSG00000153363', 1.02805), ('ENSG00000163364', 1.47959)]\n",
      "RCI threshold 0\n",
      "Load sequence...\n",
      "Num IDs: 8687\n",
      "Cross validation...\n",
      "2023-02-05 18:54:30.079134\n",
      "splitting\n",
      "Fold 1\n",
      "FIT\n",
      "2023-02-05 18:54:31.263671\n",
      "PREDICT\n",
      "2023-02-05 18:54:37.056028\n",
      "Accuracy: 69.72% Precision: 0.00% Recall: 0.00%\n",
      "F1: 0.00% AUPRC: 40.83% AUROC: 53.24%\n",
      "2023-02-05 18:54:37.238483\n",
      "Fold 2\n",
      "FIT\n",
      "2023-02-05 18:54:40.585258\n",
      "PREDICT\n",
      "2023-02-05 18:54:46.602219\n",
      "Accuracy: 90.37% Precision: 33.33% Recall: 1.80%\n",
      "F1: 3.42% AUPRC: 13.55% AUROC: 58.82%\n",
      "2023-02-05 18:54:46.886502\n",
      "Fold 3\n",
      "FIT\n",
      "2023-02-05 18:54:49.356601\n",
      "PREDICT\n",
      "2023-02-05 18:54:53.684952\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, msg_start, len(result))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accuracy: 80.95% Precision: 0.00% Recall: 0.00%\n",
      "F1: 0.00% AUPRC: 13.75% AUROC: 33.70%\n",
      "2023-02-05 18:54:53.933416\n",
      "Fold 4\n",
      "FIT\n",
      "2023-02-05 18:54:56.503072\n",
      "PREDICT\n",
      "2023-02-05 18:55:00.561002\n",
      "Accuracy: 65.41% Precision: 12.70% Recall: 3.38%\n",
      "F1: 5.33% AUPRC: 26.05% AUROC: 48.66%\n",
      "2023-02-05 18:55:00.757608\n",
      "Fold 5\n",
      "FIT\n",
      "2023-02-05 18:55:03.193690\n",
      "PREDICT\n",
      "2023-02-05 18:55:07.773922\n",
      "Accuracy: 63.06% Precision: 24.00% Recall: 38.40%\n",
      "F1: 29.54% AUPRC: 21.90% AUROC: 55.18%\n",
      "2023-02-05 18:55:08.057199\n",
      "\n",
      "Completed cross validation 5 folds 3 epochs\n",
      " accuracy mean 73.90% +/- 10.28\n",
      " precision mean 14.01% +/- 13.17\n",
      " recall mean 8.72% +/- 14.90\n",
      " F1 mean 7.66% +/- 11.13\n",
      " AUPRC mean 23.22% +/- 10.03\n",
      " AUROC mean 49.92% +/- 8.75\n",
      "Finished cell Line 9 SK.MEL.5\n",
      "\n",
      "2023-02-05 18:55:09.423759\n",
      "Start cell Line 10 SK.N.DZ\n",
      "Number of RCI values loaded 603\n",
      "Num RCI: 603\n",
      "Example RCI: [('ENSG00000117242', -0.147342), ('ENSG00000125462', -1.18067), ('ENSG00000162913', -3.77259)]\n",
      "RCI threshold 0\n",
      "Load sequence...\n",
      "Num IDs: 8687\n",
      "Cross validation...\n",
      "2023-02-05 18:55:10.531798\n",
      "splitting\n",
      "Fold 1\n",
      "FIT\n",
      "2023-02-05 18:55:11.689103\n",
      "PREDICT\n",
      "2023-02-05 18:55:15.656372\n",
      "Accuracy: 41.33% Precision: 37.18% Recall: 35.46%\n",
      "F1: 36.30% AUPRC: 37.71% AUROC: 37.89%\n",
      "2023-02-05 18:55:15.901966\n",
      "Fold 2\n",
      "FIT\n",
      "2023-02-05 18:55:18.535655\n",
      "PREDICT\n",
      "2023-02-05 18:55:23.226024\n",
      "Accuracy: 47.59% Precision: 39.88% Recall: 43.54%\n",
      "F1: 41.63% AUPRC: 42.18% AUROC: 48.71%\n",
      "2023-02-05 18:55:23.465030\n",
      "Fold 3\n",
      "FIT\n",
      "2023-02-05 18:55:25.892669\n",
      "PREDICT\n",
      "2023-02-05 18:55:29.911415\n",
      "Accuracy: 66.67% Precision: 28.85% Recall: 37.88%\n",
      "F1: 32.75% AUPRC: 25.99% AUROC: 60.49%\n",
      "2023-02-05 18:55:30.161579\n",
      "Fold 4\n",
      "FIT\n",
      "2023-02-05 18:55:33.488096\n",
      "PREDICT\n",
      "2023-02-05 18:55:38.000324\n",
      "Accuracy: 50.56% Precision: 34.92% Recall: 51.23%\n",
      "F1: 41.53% AUPRC: 32.69% AUROC: 49.39%\n",
      "2023-02-05 18:55:38.190856\n",
      "Fold 5\n",
      "FIT\n",
      "2023-02-05 18:55:40.640840\n",
      "PREDICT\n",
      "2023-02-05 18:55:46.408065\n",
      "Accuracy: 61.30% Precision: 36.65% Recall: 23.79%\n",
      "F1: 28.85% AUPRC: 35.17% AUROC: 53.80%\n",
      "2023-02-05 18:55:46.699180\n",
      "\n",
      "Completed cross validation 5 folds 3 epochs\n",
      " accuracy mean 53.49% +/- 9.23\n",
      " precision mean 35.49% +/- 3.69\n",
      " recall mean 38.38% +/- 9.09\n",
      " F1 mean 36.21% +/- 4.97\n",
      " AUPRC mean 34.75% +/- 5.39\n",
      " AUROC mean 50.06% +/- 7.39\n",
      "Finished cell Line 10 SK.N.DZ\n",
      "\n",
      "2023-02-05 18:55:48.182232\n",
      "Start cell Line 11 SK.N.SH\n",
      "Number of RCI values loaded 1647\n",
      "Num RCI: 1647\n",
      "Example RCI: [('ENSG00000117242', 0.0675743), ('ENSG00000125804', -1.90182), ('ENSG00000136275', 1.03562)]\n",
      "RCI threshold 0\n",
      "Load sequence...\n",
      "Num IDs: 8687\n",
      "Cross validation...\n",
      "2023-02-05 18:55:49.334460\n",
      "splitting\n",
      "Fold 1\n",
      "FIT\n",
      "2023-02-05 18:55:51.115548\n",
      "PREDICT\n",
      "2023-02-05 18:56:02.037915\n",
      "Accuracy: 72.06% Precision: 45.16% Recall: 4.40%\n",
      "F1: 8.02% AUPRC: 30.94% AUROC: 53.20%\n",
      "2023-02-05 18:56:02.301823\n",
      "Fold 2\n",
      "FIT\n",
      "2023-02-05 18:56:05.338417\n",
      "PREDICT\n",
      "2023-02-05 18:56:16.257775\n",
      "Accuracy: 74.30% Precision: 31.44% Recall: 19.35%\n",
      "F1: 23.96% AUPRC: 26.91% AUROC: 57.59%\n",
      "2023-02-05 18:56:16.622537\n",
      "Fold 3\n",
      "FIT\n",
      "2023-02-05 18:56:19.680112\n",
      "PREDICT\n",
      "2023-02-05 18:56:30.598921\n",
      "Accuracy: 78.60% Precision: 31.51% Recall: 9.87%\n",
      "F1: 15.03% AUPRC: 24.02% AUROC: 58.18%\n",
      "2023-02-05 18:56:30.879944\n",
      "Fold 4\n",
      "FIT\n",
      "2023-02-05 18:56:33.946759\n",
      "PREDICT\n",
      "2023-02-05 18:56:44.883317\n",
      "Accuracy: 63.38% Precision: 44.02% Recall: 20.44%\n",
      "F1: 27.92% AUPRC: 43.69% AUROC: 58.68%\n",
      "2023-02-05 18:56:45.302221\n",
      "Fold 5\n",
      "FIT\n",
      "2023-02-05 18:56:48.349335\n",
      "PREDICT\n",
      "2023-02-05 18:56:59.291054\n",
      "Accuracy: 63.60% Precision: 46.53% Recall: 19.14%\n",
      "F1: 27.13% AUPRC: 42.29% AUROC: 61.12%\n",
      "2023-02-05 18:56:59.539534\n",
      "\n",
      "Completed cross validation 5 folds 3 epochs\n",
      " accuracy mean 70.39% +/- 6.01\n",
      " precision mean 39.73% +/- 6.79\n",
      " recall mean 14.64% +/- 6.38\n",
      " F1 mean 20.41% +/- 7.70\n",
      " AUPRC mean 33.57% +/- 8.01\n",
      " AUROC mean 57.76% +/- 2.57\n",
      "Finished cell Line 11 SK.N.SH\n",
      "\n",
      "2023-02-05 18:57:00.892204\n",
      "Start cell Line 12 GM12878\n",
      "Number of RCI values loaded 1698\n",
      "Num RCI: 1698\n",
      "Example RCI: [('ENSG00000105501', 0.363348), ('ENSG00000116652', -3.42824), ('ENSG00000117242', 0.82852)]\n",
      "RCI threshold 0\n",
      "Load sequence...\n",
      "Num IDs: 8687\n",
      "Cross validation...\n",
      "2023-02-05 18:57:01.999039\n",
      "splitting\n",
      "Fold 1\n",
      "FIT\n",
      "2023-02-05 18:57:04.245473\n",
      "PREDICT\n",
      "2023-02-05 18:57:10.082651\n",
      "Accuracy: 72.88% Precision: 38.81% Recall: 18.12%\n",
      "F1: 24.70% AUPRC: 32.77% AUROC: 55.39%\n",
      "2023-02-05 18:57:10.332186\n",
      "Fold 2\n",
      "FIT\n",
      "2023-02-05 18:57:13.295277\n",
      "PREDICT\n",
      "2023-02-05 18:57:19.543430\n",
      "Accuracy: 61.98% Precision: 30.70% Recall: 23.03%\n",
      "F1: 26.32% AUPRC: 27.75% AUROC: 47.14%\n",
      "2023-02-05 18:57:19.833151\n",
      "Fold 3\n",
      "FIT\n",
      "2023-02-05 18:57:22.746738\n",
      "PREDICT\n",
      "2023-02-05 18:57:28.311375\n",
      "Accuracy: 61.77% Precision: 51.82% Recall: 22.11%\n",
      "F1: 30.99% AUPRC: 46.44% AUROC: 60.58%\n",
      "2023-02-05 18:57:28.601532\n",
      "Fold 4\n",
      "FIT\n",
      "2023-02-05 18:57:32.872578\n",
      "PREDICT\n",
      "2023-02-05 18:57:38.823444\n",
      "Accuracy: 58.94% Precision: 54.46% Recall: 40.22%\n",
      "F1: 46.27% AUPRC: 53.36% AUROC: 59.87%\n",
      "2023-02-05 18:57:39.073951\n",
      "Fold 5\n",
      "FIT\n",
      "2023-02-05 18:57:42.064533\n",
      "PREDICT\n",
      "2023-02-05 18:57:52.990529\n",
      "Accuracy: 57.84% Precision: 36.11% Recall: 36.67%\n",
      "F1: 36.39% AUPRC: 33.88% AUROC: 54.61%\n",
      "2023-02-05 18:57:53.398407\n",
      "\n",
      "Completed cross validation 5 folds 3 epochs\n",
      " accuracy mean 62.68% +/- 5.34\n",
      " precision mean 42.38% +/- 9.20\n",
      " recall mean 28.03% +/- 8.74\n",
      " F1 mean 32.93% +/- 7.81\n",
      " AUPRC mean 38.84% +/- 9.52\n",
      " AUROC mean 55.52% +/- 4.81\n",
      "Finished cell Line 12 GM12878\n",
      "\n",
      "2023-02-05 18:57:54.768012\n",
      "Start cell Line 13 K562\n",
      "Number of RCI values loaded 958\n",
      "Num RCI: 958\n",
      "Example RCI: [('ENSG00000117242', 0.316474), ('ENSG00000131484', -4.19438), ('ENSG00000151303', -4.16993)]\n",
      "RCI threshold 0\n",
      "Load sequence...\n",
      "Num IDs: 8687\n",
      "Cross validation...\n",
      "2023-02-05 18:57:55.902689\n",
      "splitting\n",
      "Fold 1\n",
      "FIT\n",
      "2023-02-05 18:57:57.509758\n",
      "PREDICT\n",
      "2023-02-05 18:58:02.461125\n",
      "Accuracy: 57.21% Precision: 31.58% Recall: 24.66%\n",
      "F1: 27.69% AUPRC: 30.52% AUROC: 45.30%\n",
      "2023-02-05 18:58:02.708541\n",
      "Fold 2\n",
      "FIT\n",
      "2023-02-05 18:58:05.243316\n",
      "PREDICT\n",
      "2023-02-05 18:58:09.466946\n",
      "Accuracy: 75.00% Precision: 48.87% Recall: 43.20%\n",
      "F1: 45.86% AUPRC: 42.86% AUROC: 71.83%\n",
      "2023-02-05 18:58:09.726685\n",
      "Fold 3\n",
      "FIT\n",
      "2023-02-05 18:58:13.204054\n",
      "PREDICT\n",
      "2023-02-05 18:58:17.750446\n",
      "Accuracy: 66.93% Precision: 47.42% Recall: 25.56%\n",
      "F1: 33.21% AUPRC: 38.57% AUROC: 57.50%\n",
      "2023-02-05 18:58:18.003479\n",
      "Fold 4\n",
      "FIT\n",
      "2023-02-05 18:58:20.533779\n",
      "PREDICT\n",
      "2023-02-05 18:58:26.346892\n",
      "Accuracy: 54.08% Precision: 15.38% Recall: 3.95%\n",
      "F1: 6.29% AUPRC: 35.01% AUROC: 47.37%\n",
      "2023-02-05 18:58:26.591012\n",
      "Fold 5\n",
      "FIT\n",
      "2023-02-05 18:58:29.146422\n",
      "PREDICT\n",
      "2023-02-05 18:58:33.461326\n",
      "Accuracy: 67.24% Precision: 30.15% Recall: 41.21%\n",
      "F1: 34.82% AUPRC: 35.34% AUROC: 63.89%\n",
      "2023-02-05 18:58:33.711865\n",
      "\n",
      "Completed cross validation 5 folds 3 epochs\n",
      " accuracy mean 64.09% +/- 7.54\n",
      " precision mean 34.68% +/- 12.38\n",
      " recall mean 27.71% +/- 14.15\n",
      " F1 mean 29.57% +/- 13.05\n",
      " AUPRC mean 36.46% +/- 4.10\n",
      " AUROC mean 57.18% +/- 9.97\n",
      "Finished cell Line 13 K562\n",
      "\n",
      "2023-02-05 18:58:35.064125\n",
      "Start cell Line 14 IMR.90\n",
      "Number of RCI values loaded 390\n",
      "Num RCI: 390\n",
      "Example RCI: [('ENSG00000117242', 0.245683), ('ENSG00000170919', -4.93667), ('ENSG00000171889', 0.0331669)]\n",
      "RCI threshold 0\n",
      "Load sequence...\n",
      "Num IDs: 8687\n",
      "Cross validation...\n",
      "2023-02-05 18:58:36.242121\n",
      "splitting\n",
      "Fold 1\n",
      "FIT\n",
      "2023-02-05 18:58:37.384562\n",
      "PREDICT\n",
      "2023-02-05 18:58:40.538162\n",
      "Accuracy: 51.09% Precision: 62.75% Recall: 12.36%\n",
      "F1: 20.65% AUPRC: 51.31% AUROC: 47.03%\n",
      "2023-02-05 18:58:40.697001\n",
      "Fold 2\n",
      "FIT\n",
      "2023-02-05 18:58:42.649557\n",
      "PREDICT\n",
      "2023-02-05 18:58:45.840238\n",
      "Accuracy: 56.21% Precision: 24.68% Recall: 52.00%\n",
      "F1: 33.48% AUPRC: 30.00% AUROC: 56.11%\n",
      "2023-02-05 18:58:45.976576\n",
      "Fold 3\n",
      "FIT\n",
      "2023-02-05 18:58:47.928759\n",
      "PREDICT\n",
      "2023-02-05 18:58:51.089612\n",
      "Accuracy: 71.26% Precision: 65.52% Recall: 23.17%\n",
      "F1: 34.23% AUPRC: 53.91% AUROC: 69.88%\n",
      "2023-02-05 18:58:51.386474\n",
      "Fold 4\n",
      "FIT\n",
      "2023-02-05 18:58:53.478323\n",
      "PREDICT\n",
      "2023-02-05 18:58:56.663170\n",
      "Accuracy: 72.45% Precision: 41.67% Recall: 15.96%\n",
      "F1: 23.08% AUPRC: 37.53% AUROC: 61.61%\n",
      "2023-02-05 18:58:56.873042\n",
      "Fold 5\n",
      "FIT\n",
      "2023-02-05 18:58:59.405048\n",
      "PREDICT\n",
      "2023-02-05 18:59:01.831211\n",
      "Accuracy: 51.48% Precision: 32.99% Recall: 21.77%\n",
      "F1: 26.23% AUPRC: 38.87% AUROC: 44.79%\n",
      "2023-02-05 18:59:01.979245\n",
      "\n",
      "Completed cross validation 5 folds 3 epochs\n",
      " accuracy mean 60.50% +/- 9.45\n",
      " precision mean 45.52% +/- 16.14\n",
      " recall mean 25.05% +/- 14.03\n",
      " F1 mean 27.53% +/- 5.46\n",
      " AUPRC mean 42.32% +/- 8.97\n",
      " AUROC mean 55.88% +/- 9.28\n",
      "Finished cell Line 14 IMR.90\n",
      "\n",
      "2023-02-05 18:59:03.356569\n"
     ]
    }
   ],
   "source": [
    "for CELL_LINE in range(15):\n",
    "    print(datetime.now())\n",
    "    print('Start cell Line',CELL_LINE,all_cell_lines[CELL_LINE])\n",
    "    loader = DataLoader()\n",
    "    filepath = DATA_DIR+RCI_FILE\n",
    "    gene_to_rci = loader.load_gene_rci_values(filepath,CELL_LINE)\n",
    "    print('Num RCI:', len(gene_to_rci.keys()))\n",
    "    print('Example RCI:', list(gene_to_rci.items())[:3])\n",
    "    all_genes = list(gene_to_rci.keys())\n",
    "    RCI_THRESHOLD = loader.get_threshold(CELL_LINE,gene_to_rci)\n",
    "    print('RCI threshold',RCI_THRESHOLD)\n",
    "    print('Load sequence...')\n",
    "    filepath = DATA_DIR+SEQUENCE_FILE\n",
    "    allids,allseq = loader.load_sequence(filepath)\n",
    "    print('Num IDs:',len(allids))\n",
    "    loader = None  # drop K-mer cache to save RAM\n",
    "\n",
    "    print(\"Cross validation...\")\n",
    "    cvdo = CrossValidator(EPOCHS,FOLDS,BREAK)\n",
    "    cvdo.set_sequences(allids,allseq)\n",
    "    cvdo.set_rci_map(gene_to_rci)\n",
    "    cv_accuracy, cv_precision, cv_recall, cv_f1, cv_auprc, cv_auroc = cvdo.do_cross_validation()   \n",
    "    cvdo = None\n",
    "    print(\"Completed cross validation %d folds %d epochs\" % (FOLDS,EPOCHS)) \n",
    "    print(\" accuracy mean %.2f%% +/- %.2f\" % (np.mean(cv_accuracy), np.std(cv_accuracy)))\n",
    "    print(\" precision mean %.2f%% +/- %.2f\" % (np.mean(cv_precision), np.std(cv_precision)))\n",
    "    print(\" recall mean %.2f%% +/- %.2f\" % (np.mean(cv_recall), np.std(cv_recall)))\n",
    "    print(\" F1 mean %.2f%% +/- %.2f\" % (np.mean(cv_f1), np.std(cv_f1)))\n",
    "    print(\" AUPRC mean %.2f%% +/- %.2f\" % (np.mean(cv_auprc), np.std(cv_auprc)))\n",
    "    print(\" AUROC mean %.2f%% +/- %.2f\" % (np.mean(cv_auroc), np.std(cv_auroc)))\n",
    "    print('Finished cell Line',CELL_LINE,all_cell_lines[CELL_LINE])\n",
    "    print()\n",
    "print(datetime.now())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "id": "HkCeDg_HdQ36"
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "accelerator": "GPU",
  "colab": {
   "provenance": []
  },
  "gpuClass": "standard",
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
