{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "PG-tGRnlFLA3"
   },
   "source": [
    "# Logistic Regression\n",
    "This model predicted nearly the same score (0.47) on every RNA. What's wrong?\n",
    "\n",
    "canonical lncRNA, -1 threshold, cross-valiation, all the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "0RmwUsVLFLA6",
    "outputId": "64cc44e8-c4cd-4eee-f276-47d1983f985c"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2023-04-18 08:45:19.518576\n"
     ]
    }
   ],
   "source": [
    "from datetime import datetime\n",
    "print(datetime.now())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "OlzN9OdsFWEU",
    "outputId": "20173346-1246-44f4-eaf7-2a5f2c032df7"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-04-18 08:45:19.587874: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GPU device not found\n",
      "/Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-04-18 08:45:35.289861: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\n"
     ]
    }
   ],
   "source": [
    "import tensorflow as tf\n",
    "dt='float32'\n",
    "tf.keras.backend.set_floatx('float32')\n",
    "# tf.random.set_seed(42) # supposedly leads to reproducible results\n",
    "\n",
    "device_name = tf.test.gpu_device_name()\n",
    "if device_name != '/device:GPU:0':\n",
    "    print('GPU device not found')\n",
    "else:\n",
    "    print('Found GPU at: {}'.format(device_name))\n",
    "\n",
    "try:\n",
    "    from google.colab import drive\n",
    "    IN_COLAB = True\n",
    "    print('Running on CoLab')\n",
    "    PATH='/content/drive/'\n",
    "    drive.mount(PATH)\n",
    "    DATA_DIR=PATH+'My Drive/data/Localization/TrainTest/TrainTest_ver43/'  # must end in \"/\"\n",
    "    MODEL_DIR=PATH+'My Drive/data/Localization/Models/'  # must end in \"/\"\n",
    "except:\n",
    "    IN_COLAB = False\n",
    "    DATA_DIR = 'D:/Adjeroh/Localization/TrainTest/'   # Windows\n",
    "    DATA_DIR = '/Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/'    # Mac\n",
    "    MODEL_DIR = '/Users/jasonmiller/WVU/Localization/Models/'    # Mac\n",
    "print(DATA_DIR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "PRX-UEr8FLA8",
    "outputId": "39890dac-1ca8-4ba2-9229-2b581f26b315"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Python 3.10.0\n",
      "sklearn 1.1.2\n"
     ]
    }
   ],
   "source": [
    "from platform import python_version\n",
    "print('Python',python_version())\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy.stats as ss\n",
    "import pickle\n",
    "import time # sleep function\n",
    "from os.path import isfile\n",
    "from matplotlib import pyplot as plt \n",
    "import sklearn   # pip install --upgrade scikit-learn\n",
    "print('sklearn',sklearn.__version__)\n",
    "\n",
    "from tensorflow import keras\n",
    "from sklearn.mixture import GaussianMixture\n",
    "from sklearn.model_selection import KFold\n",
    "from sklearn.metrics import accuracy_score\n",
    "from sklearn.metrics import precision_score\n",
    "from sklearn.metrics import recall_score\n",
    "from sklearn.metrics import f1_score\n",
    "from sklearn.metrics import precision_recall_curve\n",
    "from sklearn.metrics import roc_curve\n",
    "from sklearn.metrics import roc_auc_score\n",
    "from sklearn.metrics import auc\n",
    "from sklearn.metrics import matthews_corrcoef\n",
    "from sklearn.utils import class_weight\n",
    "from sklearn.metrics import confusion_matrix\n",
    "from sklearn.linear_model import LogisticRegression as LogReg\n",
    "\n",
    "from keras.models import Sequential\n",
    "from keras.layers import Masking\n",
    "from keras.layers import Embedding\n",
    "from keras.layers import Conv1D\n",
    "from keras.layers import GlobalAveragePooling1D\n",
    "from keras.layers import GlobalMaxPooling1D\n",
    "from keras.layers import AveragePooling1D\n",
    "from keras.layers import MaxPooling1D\n",
    "from keras.layers import Bidirectional\n",
    "from keras.layers import LSTM\n",
    "from keras.layers import TimeDistributed\n",
    "from keras.layers import Flatten\n",
    "from keras.layers import Dense\n",
    "from keras.layers import Dropout\n",
    "from keras.losses import BinaryCrossentropy\n",
    "#from keras.losses import Hinge\n",
    "\n",
    "K=5\n",
    "ALPHABET=4**K + 1\n",
    "CFILTERS=64 \n",
    "FILTERSIZE=8\n",
    "RCELLS=32\n",
    "DCELLS=16\n",
    "EPOCHS=150 \n",
    "EMBED_DIMEN = 4 # arbitrary hyperparameter\n",
    "# MINLEN=200   # NO LONGER USED\n",
    "# MAXLEN=5000\n",
    "RCI_THRESHOLD_MECHANISM = 'ZERO'  # 'RCI_GMM' 'ZERO' 'THE_MEAN'\n",
    "BREAK = False   # break after first fold\n",
    "EXCLUSIONS = [1]   # possibly exclude cell line 1 = H1.hESC\n",
    "FILTER_TRAIN        = False\n",
    "FILTER_TAILS_TRAIN  = False\n",
    "FILTER_MIDDLE_TRAIN = False\n",
    "FILTER_TEST         = False\n",
    "FILTER_TAILS_TEST   = False\n",
    "FILTER_MIDDLE_TEST  = False\n",
    "REPEATS = 2\n",
    "FOLDS = 5\n",
    "\n",
    "SAVE_MODEL_FILENAME = None # 'RF_101'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "id": "LnkpVKdMFLA-"
   },
   "outputs": [],
   "source": [
    "RCI_FILE_TRAIN = 'train.lncRNA_RCI.csv'\n",
    "RCI_FILE_TEST  = None #'test.lncRNA_RCI.csv'\n",
    "\n",
    "SEQ_FILE_TRAIN = 'train.canon_lncRNA_transcripts.csv'\n",
    "SEQ_FILE_TEST  = None #'test.canon_lncRNA_transcripts.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "id": "e3p4QzQJFLA_"
   },
   "outputs": [],
   "source": [
    "def get_ordered_list():\n",
    "    ordered_list = \\\n",
    "    ['A549','H1.hESC','HeLa.S3','HepG2','HT1080',\\\n",
    "      'HUVEC','MCF.7','NCI.H460','NHEK','SK.MEL.5',\\\n",
    "      'SK.N.DZ','SK.N.SH','GM12878','K562','IMR.90']\n",
    "    return ordered_list\n",
    "all_cell_lines = get_ordered_list()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "UtqdpJOxFLBA"
   },
   "source": [
    "## Data Load"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "id": "p35ehKV3Kq0z"
   },
   "outputs": [],
   "source": [
    "class DataLoader():\n",
    "    def __init__(self):\n",
    "        self.cache=dict() \n",
    "        self.vals = {'A':0, 'C':1, 'G':2, 'T':3}\n",
    "        self.gene2rci = dict()\n",
    "        \n",
    "    def load_gene_rci_values(self,filepath,exclusions):\n",
    "        '''\n",
    "        Load all the genes from the given RCI csv file.\n",
    "        The given file usually contains train or test, not both.\n",
    "        Return dict with keys=gene:str and values=RCI:float.\n",
    "        '''\n",
    "        self.gene2rci = {}\n",
    "        overall_sum = 0\n",
    "        with open (filepath,'r') as handle:\n",
    "            header = None\n",
    "            for row in handle:\n",
    "                if header is None:\n",
    "                    header = row # skip file's header line\n",
    "                else:\n",
    "                    line = row.strip()\n",
    "                    fields = line.split(',')\n",
    "                    gene_id = fields.pop(0)\n",
    "                    log_mean = float(fields.pop(0))\n",
    "                    self.gene2rci[gene_id] = log_mean\n",
    "        print('Number of RCI values loaded',len(self.gene2rci.keys()))\n",
    "        return self.gene2rci\n",
    "\n",
    "    def _seq_to_kmer_values(self,rna,K):\n",
    "        vec=[] # seq converted to list of K-mers \n",
    "        N_indicator = 0 # indicator value\n",
    "        length = len(rna)\n",
    "        for i in range(length-K+1):\n",
    "            kmer = rna[i:i+K]\n",
    "            if 'N' in kmer:\n",
    "                value = N_indicator\n",
    "            elif kmer in self.cache.keys():\n",
    "                value = self.cache[kmer]\n",
    "            else:\n",
    "                value = 0\n",
    "                for j in range(K):\n",
    "                    value *= 4   \n",
    "                    nextnuc = kmer[j] \n",
    "                    nucval = self.vals[nextnuc]\n",
    "                    value += nucval\n",
    "                value += 1   # NNN => 0, AAA => 1\n",
    "                self.cache[kmer] = value\n",
    "            vec.append(value)\n",
    "        return vec\n",
    "\n",
    "    def load_sequence(self,filepath):\n",
    "        '''\n",
    "        Load all the sequences from the given file. \n",
    "        Load our version of GenCode -- csv, easier to parse than fasta.\n",
    "        Each line has IDs plus sequence.\n",
    "        The IDs in the file do not include dot-version numbers.\n",
    "        The file may already be filtered e.g. canonical transcripts.\n",
    "        '''\n",
    "        allids=[]\n",
    "        allseq=[]\n",
    "        #NREPEAT = str('N'*MAXLEN)   # not used for MLP\n",
    "        with open (filepath,'r') as handle:\n",
    "            header = None\n",
    "            for row in handle:\n",
    "                if header is None:\n",
    "                    header = row\n",
    "                else:\n",
    "                    line    = row.strip()\n",
    "                    fields  = line.split(',')\n",
    "                    tran_id = fields[0]  # without version number\n",
    "                    gene_id = fields[1]  # without version number\n",
    "                    seq_len = int(fields[3])\n",
    "                    seq_txt = fields[4]\n",
    "                    # Keep only transcripts having numeric RCI given the cell lines in use.\n",
    "                    # We have validated this by spot checking.\n",
    "                    # TO DO: validate this programmatically.\n",
    "                    if gene_id in self.gene2rci.keys():\n",
    "                        # no MAXLEN for MLP\n",
    "                        #if seq_len<=MAXLEN:\n",
    "                        #    seq_txt = seq_txt + NREPEAT\n",
    "                        #seq_txt = seq_txt[:MAXLEN]\n",
    "                        allids.append( (gene_id,tran_id) )\n",
    "                        hot_vec = self._seq_to_kmer_values(seq_txt,K)\n",
    "                        allseq.append(hot_vec)\n",
    "        self.cache=dict() # save RAM\n",
    "        return allids,allseq\n",
    "\n",
    "    def load_spectra(self,filepath):\n",
    "        '''\n",
    "        Load all (variable-length) sequences as lists of kmers.\n",
    "        Then convert each sequence to (fixed-length) kmer histograms.\n",
    "        '''\n",
    "        allids,allseq = self.load_sequence(filepath)\n",
    "        allspectra = []\n",
    "        for seq in allseq:\n",
    "            spectrum = np.zeros(ALPHABET)\n",
    "            for kmer in seq:\n",
    "                spectrum[kmer] += 1\n",
    "            spectrum /= len(seq)\n",
    "            allspectra.append(spectrum)\n",
    "        return allids,allspectra        \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "VDZ6siB_Kq04"
   },
   "source": [
    "## Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "id": "AwMbRjm0FLBF"
   },
   "outputs": [],
   "source": [
    "def build_model():\n",
    "    logreg = LogReg()\n",
    "    return logreg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "clj-wufgFLBF",
    "outputId": "a17cf7a6-3755-48aa-d96b-dd0383c6f2f2"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2023-04-18 08:45:39.109068\n",
      "LogisticRegression()\n"
     ]
    }
   ],
   "source": [
    "print(datetime.now())\n",
    "model=build_model()\n",
    "print(model)  # Print this only once\n",
    "model=None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "MgrC1alOKq07"
   },
   "source": [
    "## Training"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "id": "W9xiFzNbFLBE"
   },
   "outputs": [],
   "source": [
    "import gc\n",
    "class CrossValidator():\n",
    "    def __init__(self,epochs,score_threshold=0.5):\n",
    "        self.epochs = epochs\n",
    "        self.score_threshold = score_threshold\n",
    "        self.mechanism = 'ZERO'\n",
    "        self.discriminator = -1 # or zero\n",
    "        self.flip = False\n",
    "        self.reset_statistics()\n",
    "        \n",
    "    def reset_statistics(self):\n",
    "        self.cv_accuracy=[]\n",
    "        self.cv_precision=[]\n",
    "        self.cv_recall=[]\n",
    "        self.cv_f1=[]\n",
    "        self.cv_auprc=[]\n",
    "        self.cv_auroc=[]\n",
    "        self.cv_mcc=[]\n",
    "        \n",
    "    def _get_X_y(self, all_ids, all_seqs, rci_map): \n",
    "        # Prepare X and y for training or testing.\n",
    "        subsetX=[]\n",
    "        subsetY=[]\n",
    "        for t in range(len(all_ids)):\n",
    "            gene_id,tran_id = all_ids[t]\n",
    "            oneX            = all_seqs[t]\n",
    "            oneY            = rci_map[gene_id]\n",
    "            subsetX.append(oneX)\n",
    "            subsetY.append(oneY)\n",
    "        subsetX = np.array(subsetX)\n",
    "        subsetY = np.array(subsetY).reshape((-1,1))\n",
    "        return subsetX,subsetY\n",
    "    \n",
    "    def set_threshold_mechanism(self, mechanism):\n",
    "        if mechanism not in ['RCI_GMM','THE_MEAN','ZERO']:\n",
    "            raise Exception('Unrecognized mechansm:',mechanism)\n",
    "        self.mechanism = mechanism\n",
    "    \n",
    "    def _apply_threshold(self, array_of_rci):\n",
    "        # Takes list of float, returns list of labels [0,1].\n",
    "        if self.mechanism == 'RCI_GMM':\n",
    "            labels = self.discriminator.predict(array_of_rci)\n",
    "            if self.flip:\n",
    "                IS_CYTO = lambda label: 1 if label==0 else 0\n",
    "                labels = np.array(list(map(IS_CYTO, labels)))\n",
    "        else:  # 'THE_MEAN' or 'ZERO'\n",
    "            rci_threshold = self.discriminator\n",
    "            IS_CYTO = lambda rci: 1 if rci>rci_threshold else 0\n",
    "            labels = np.array(list(map(IS_CYTO, array_of_rci)))\n",
    "        return labels\n",
    "    \n",
    "    def _prepare_threshold(self, rci_values, create=True):\n",
    "        if self.mechanism == 'RCI_GMM':\n",
    "            if create:  # during training, create a new GMM\n",
    "                gmm = GaussianMixture(n_components=2, verbose=0, \n",
    "                  covariance_type='spherical', n_init=100) # random_state=42) \n",
    "                gmm.fit(rci_values)\n",
    "            else:   # during testing, use existing GMM\n",
    "                gmm=self.discriminator\n",
    "            self.flip = False\n",
    "            # The GMM labels are arbitrary.\n",
    "            if gmm.means_[0][0] > gmm.means_[1][0]:\n",
    "                self.flip = True\n",
    "            self.discriminator = gmm   # redundant but consistent\n",
    "        elif self.mechanism == 'THE_MEAN':\n",
    "            self.discriminator = np.mean(rci_values)\n",
    "        elif self.mechanism == 'ZERO':\n",
    "            self.discriminator = -1   # 0 usually, -1 is as in Yuan et al.\n",
    "        else: # not expected\n",
    "            self.discriminator = 0\n",
    "    \n",
    "    def _explain_threshold(self):\n",
    "        if self.mechanism == 'RCI_GMM':\n",
    "            gmm=self.discriminator\n",
    "            print('Discriminator is GMM')\n",
    "            print('Means',[gmm.means_[0][0],gmm.means_[1][0]])\n",
    "            print('Variances',gmm.covariances_)\n",
    "            print('Priors',gmm.weights_)\n",
    "            test_rcis=[-5,-4,-3.5,-3,-2.5,-2,-1.5,-1,-0.5,0,0.5,1,1.5,2,3]\n",
    "            print(test_rcis)\n",
    "            print(self._apply_threshold(np.array(test_rcis).reshape((-1,1))))\n",
    "        else:\n",
    "            print('Discriminator',self.mechanism,self.discriminator)\n",
    "    \n",
    "    def _show_sizes(self,label,values):\n",
    "        a = np.count_nonzero(values==1)\n",
    "        b = np.count_nonzero(values==0)\n",
    "        print('%s 1:0 %d:%d %5.2f%%'%(label,a,b,100*a/(a+b)))\n",
    "        \n",
    "    def save_model(self,filename):\n",
    "        if self.model is not None:\n",
    "            filepath = MODEL_DIR + filename\n",
    "            #self.model.save(filepath)\n",
    "            print('? Saved model to',filepath)\n",
    "        \n",
    "    def load_model(self,filename):\n",
    "        filepath = MODEL_DIR + filename\n",
    "        #self.model = keras.models.load_model(filepath)\n",
    "        print('? Loaded model from',filepath)\n",
    "        \n",
    "    def train_new_model(self,train_ids,train_seq,train_rci,\n",
    "            valid_ids=None,valid_seq=None,valid_rci=None):\n",
    "        print(datetime.now())\n",
    "        X_train,y_rci = self._get_X_y(train_ids,train_seq,train_rci) \n",
    "        self._prepare_threshold(y_rci,True)  \n",
    "        self._explain_threshold()\n",
    "        y_train = self._apply_threshold(y_rci)\n",
    "        self._show_sizes('Train',y_train)\n",
    "        #cw = class_weight.compute_class_weight('balanced', classes=[0,1], y=y_train)\n",
    "        #cwd = {0: cw[0], 1: cw[1]}\n",
    "        #print('Computed class weights:',cwd)\n",
    "        if valid_ids is not None:\n",
    "            X_valid,y_rci = self._get_X_y(valid_ids,valid_seq,valid_rci) \n",
    "            y_valid = self._apply_threshold(y_rci)\n",
    "            self._show_sizes('Valid',y_valid)\n",
    "        y_rci = None\n",
    "\n",
    "        self.model=build_model()\n",
    "        \n",
    "        print(\"FIT\")\n",
    "        print(datetime.now())\n",
    "        self.model.fit(X_train, y_train) # sample weight\n",
    "\n",
    "    def test_without_training(self,test_ids,test_seq,test_rci):\n",
    "        # For final test, do no train.\n",
    "        # Assume set_sequences() set the test set.\n",
    "        print(datetime.now())\n",
    "        X_test,y_rci = self._get_X_y(test_ids,test_seq,test_rci) \n",
    "        y_test = self._apply_threshold(y_rci)\n",
    "        y_rci = None\n",
    "        \n",
    "        print(\"PREDICT\")\n",
    "        print(datetime.now())        \n",
    "        yhat_pairs=self.model.predict_proba(X_test)  # [ prob of 0, prob of 1 ]\n",
    "        yhat_pred=[pair[1] for pair in yhat_pairs]\n",
    "        yhat_classes=self.model.predict(X_test)  # 0 or 1\n",
    "        \n",
    "        print('debug pred',yhat_pred[:3])\n",
    "        print('debug class',yhat_classes[:3])\n",
    "\n",
    "        self._show_sizes('Test',y_test)\n",
    "        self._show_sizes('Predict',yhat_classes)\n",
    "        print('Test sizes',X_test.shape,y_test.shape)\n",
    "        print('Distrib of scores:',np.mean(yhat_pred),'mean',np.std(yhat_pred),'std')\n",
    "        print('Range of scores:',np.min(yhat_pred),'to',np.max(yhat_pred))\n",
    "        print('Score threshold',self.score_threshold)\n",
    "        cm1 = confusion_matrix(y_test,yhat_classes)\n",
    "        print('Confusion matrix\\n',cm1)\n",
    "        cm2 = confusion_matrix(y_test,yhat_classes,normalize='all')\n",
    "        print('Normalized matrix\\n',cm2)\n",
    "\n",
    "        accuracy = accuracy_score(y_test, yhat_classes)*100.\n",
    "        precision = precision_score(y_test, yhat_classes)*100.\n",
    "        recall = recall_score(y_test, yhat_classes)*100.\n",
    "        f1 = f1_score(y_test, yhat_classes)*100.\n",
    "        prc_Y, prc_X, prc_bins = precision_recall_curve(y_test, yhat_pred)\n",
    "        auprc = auc(prc_X,prc_Y)*100.\n",
    "        auroc = roc_auc_score(y_test, yhat_pred)*100.\n",
    "        mcc = matthews_corrcoef(y_test, yhat_classes)\n",
    "\n",
    "        correct_pos = 0\n",
    "        correct_neg = 0\n",
    "        wrong_pos = 0\n",
    "        wrong_neg = 0\n",
    "        for i in range(len(y_test)):\n",
    "            if yhat_pred[i]>=0.65:\n",
    "                if y_test[i]==1:\n",
    "                    correct_pos += 1\n",
    "                else:\n",
    "                    wrong_pos += 1\n",
    "            elif yhat_pred[i]<=0.35:\n",
    "                if y_test[i]==0:\n",
    "                    correct_neg += 1\n",
    "                else:\n",
    "                    wrong_neg += 1\n",
    "        print('Extreme scores correct, pos:neg',correct_pos,correct_neg)  \n",
    "        print('Extreme scores incorrect pos:neg',wrong_pos,wrong_neg)  \n",
    "\n",
    "        #self._pretty_pictures(y_test,yhat_pred,prc_X,prc_Y)\n",
    "        \n",
    "        print('Accuracy: %.2f%% Precision: %.2f%% Recall: %.2f%%' % (accuracy,precision,recall)) \n",
    "        print('F1: %.2f%% MCC: %.4f' % (f1,mcc)) \n",
    "        print('AUPRC: %.2f%% AUROC: %.2f%%' % (auprc,auroc)) \n",
    "\n",
    "        self.cv_accuracy.append(accuracy)\n",
    "        self.cv_precision.append(precision)\n",
    "        self.cv_recall.append(recall)\n",
    "        self.cv_f1.append(f1)\n",
    "        self.cv_mcc.append(mcc)\n",
    "        self.cv_auprc.append(auprc)\n",
    "        self.cv_auroc.append(auroc)\n",
    "\n",
    "    def _pretty_pictures(self,y_valid,yhat_pred,prc_X,prc_Y):\n",
    "        count_ones= len(y_valid[y_valid==1])\n",
    "        count_zeros= len(y_valid[y_valid==0])\n",
    "        guess = max(count_ones,count_zeros) / len(y_valid)\n",
    "        # PRC\n",
    "        plt.plot(prc_X, prc_Y, marker='.')\n",
    "        plt.plot([0, 1], [guess,guess], linestyle='--')\n",
    "        plt.xlabel('Recall')\n",
    "        plt.ylabel('Precision')\n",
    "        plt.show()\n",
    "        # ROC\n",
    "        fpr, tpr, roc_bins = roc_curve(y_valid, yhat_pred)\n",
    "        plt.plot(fpr, tpr, marker='.')\n",
    "        plt.xlabel('False Positive Rate')\n",
    "        plt.ylabel('True Positive Rate')\n",
    "        plt.show()\n",
    "        \n",
    "    def get_statistics(self):\n",
    "        return \\\n",
    "        self.cv_accuracy,\\\n",
    "        self.cv_precision,\\\n",
    "        self.cv_recall,\\\n",
    "        self.cv_f1,\\\n",
    "        self.cv_mcc,\\\n",
    "        self.cv_auprc,\\\n",
    "        self.cv_auroc\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "Td97uyyj5qDq"
   },
   "source": [
    "## Train"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "id": "poe3rwAF4Ir7"
   },
   "outputs": [],
   "source": [
    "class Separator():\n",
    "    def __init__(self):\n",
    "        self.train_ids = []\n",
    "        self.train_seq = []\n",
    "        self.train_rci = dict()\n",
    "        self.val_ids = []\n",
    "        self.val_seq = []\n",
    "        self.val_rci = dict()\n",
    "    def load(self,data_dir,rep,fold):\n",
    "        filename='cv.{}.{}.validation_genes.txt'.format(rep,fold)\n",
    "        filename = data_dir + filename\n",
    "        self.val_genes = set()\n",
    "        print('Opening file',filename)\n",
    "        with open(filename,'r') as fin:\n",
    "            for line in fin:\n",
    "                gene_id = line.strip()\n",
    "                self.val_genes.add(gene_id)\n",
    "    def process(self,allids,allseq,gene_to_rci):\n",
    "        size = len(allids)\n",
    "        for t in range(size):\n",
    "            gene_id,tran_id = allids[t]\n",
    "            oneX            = allseq[t]\n",
    "            oneY            = gene_to_rci[gene_id]\n",
    "            in_middle = gene_to_rci[gene_id] >= -2 and gene_to_rci[gene_id] <= 0\n",
    "            in_tails = gene_to_rci[gene_id] < -2 or gene_to_rci[gene_id] > 0\n",
    "            if gene_id in self.val_genes:\n",
    "                if FILTER_TEST and (\\\n",
    "                    (FILTER_TAILS_TEST and in_tails) or \\\n",
    "                    (FILTER_MIDDLE_TEST and in_middle)):\n",
    "                    pass\n",
    "                else:\n",
    "                    self.val_ids.append(allids[t])\n",
    "                    self.val_seq.append(allseq[t])\n",
    "                    self.val_rci[gene_id]=gene_to_rci[gene_id]\n",
    "            else:\n",
    "                if FILTER_TRAIN and (\\\n",
    "                    (FILTER_TAILS_TRAIN and in_tails) or \\\n",
    "                    (FILTER_MIDDLE_TRAIN and in_middle)):\n",
    "                    pass\n",
    "                else:\n",
    "                    self.train_ids.append(allids[t])\n",
    "                    self.train_seq.append(allseq[t])\n",
    "                    self.train_rci[gene_id]=gene_to_rci[gene_id]\n",
    "    def get_ids(self):\n",
    "        return self.train_ids,self.val_ids\n",
    "    def get_seq(self):\n",
    "        return self.train_seq,self.val_seq\n",
    "    def get_rci(self):\n",
    "        return self.train_rci,self.val_rci"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 1000
    },
    "id": "XC9m0W-pFLBH",
    "outputId": "f1051adc-9a79-4671-dd2b-2d9617da1129",
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2023-04-18 08:45:39.223804\n",
      "Load RCI from /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/train.lncRNA_RCI.csv\n",
      "Number of RCI values loaded 4371\n",
      "Num RCI: 4371\n",
      "Load sequence from /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/train.canon_lncRNA_transcripts.csv\n",
      "Loaded 4371 sequences.\n",
      "\n",
      "Training # 1 1\n",
      "2023-04-18 08:45:45.275455\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.1.1.validation_genes.txt\n",
      "2023-04-18 08:45:45.283224\n",
      "Discriminator ZERO -1\n",
      "Train 1:0 1659:1837 47.45%\n",
      "Valid 1:0 402:473 45.94%\n",
      "FIT\n",
      "2023-04-18 08:45:45.322849\n",
      "\n",
      "Testing # 1 1\n",
      "2023-04-18 08:45:45.419213\n",
      "2023-04-18 08:45:45.419237\n",
      "PREDICT\n",
      "2023-04-18 08:45:45.424109\n",
      "debug pred [0.48109857075714224, 0.4759745505671915, 0.477379054510677]\n",
      "debug class [0 0 0]\n",
      "Test 1:0 402:473 45.94%\n",
      "Predict 1:0 0:875  0.00%\n",
      "Test sizes (875, 1025) (875,)\n",
      "Distrib of scores: 0.474860400376961 mean 0.0033338810042552566 std\n",
      "Range of scores: 0.46262125252084335 to 0.4862546078827394\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[473   0]\n",
      " [402   0]]\n",
      "Normalized matrix\n",
      " [[0.54057143 0.        ]\n",
      " [0.45942857 0.        ]]\n",
      "Extreme scores correct, pos:neg 0 0\n",
      "Extreme scores incorrect pos:neg 0 0\n",
      "Accuracy: 54.06% Precision: 0.00% Recall: 0.00%\n",
      "F1: 0.00% MCC: 0.0000\n",
      "AUPRC: 55.80% AUROC: 61.29%\n",
      " accuracy [54.05714285714286]\n",
      " precision [0.0]\n",
      " recall [0.0]\n",
      " F1 [0.0]\n",
      " MCC [0.0]\n",
      " AUPRC [55.79731981219367]\n",
      " AUROC [61.291323509303375]\n",
      "\n",
      "Training # 1 2\n",
      "2023-04-18 08:45:45.448824\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.1.2.validation_genes.txt\n",
      "2023-04-18 08:45:45.455450\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1334: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, msg_start, len(result))\n",
      "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1334: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, msg_start, len(result))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Discriminator ZERO -1\n",
      "Train 1:0 1643:1853 47.00%\n",
      "Valid 1:0 418:457 47.77%\n",
      "FIT\n",
      "2023-04-18 08:45:45.518804\n",
      "\n",
      "Testing # 1 2\n",
      "2023-04-18 08:45:45.621433\n",
      "2023-04-18 08:45:45.621464\n",
      "PREDICT\n",
      "2023-04-18 08:45:45.627133\n",
      "debug pred [0.4729146779131491, 0.46845969594234443, 0.47229504857564275]\n",
      "debug class [0 0 0]\n",
      "Test 1:0 418:457 47.77%\n",
      "Predict 1:0 0:875  0.00%\n",
      "Test sizes (875, 1025) (875,)\n",
      "Distrib of scores: 0.4700707405865916 mean 0.0029725128252553463 std\n",
      "Range of scores: 0.46016039560799316 to 0.48054026173303116\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[457   0]\n",
      " [418   0]]\n",
      "Normalized matrix\n",
      " [[0.52228571 0.        ]\n",
      " [0.47771429 0.        ]]\n",
      "Extreme scores correct, pos:neg 0 0\n",
      "Extreme scores incorrect pos:neg 0 0\n",
      "Accuracy: 52.23% Precision: 0.00% Recall: 0.00%\n",
      "F1: 0.00% MCC: 0.0000\n",
      "AUPRC: 56.86% AUROC: 63.15%\n",
      " accuracy [52.22857142857143]\n",
      " precision [0.0]\n",
      " recall [0.0]\n",
      " F1 [0.0]\n",
      " MCC [0.0]\n",
      " AUPRC [56.8583036142403]\n",
      " AUROC [63.152659847350634]\n",
      "\n",
      "Training # 1 3\n",
      "2023-04-18 08:45:45.638942\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.1.3.validation_genes.txt\n",
      "2023-04-18 08:45:45.646035\n",
      "Discriminator ZERO -1\n",
      "Train 1:0 1671:1826 47.78%\n",
      "Valid 1:0 390:484 44.62%\n",
      "FIT\n",
      "2023-04-18 08:45:45.707080\n",
      "\n",
      "Testing # 1 3\n",
      "2023-04-18 08:45:45.811223\n",
      "2023-04-18 08:45:45.811247\n",
      "PREDICT\n",
      "2023-04-18 08:45:45.815644\n",
      "debug pred [0.4762916458038255, 0.47750354334359985, 0.4773919563518359]\n",
      "debug class [0 0 0]\n",
      "Test 1:0 390:484 44.62%\n",
      "Predict 1:0 0:874  0.00%\n",
      "Test sizes (874, 1025) (874,)\n",
      "Distrib of scores: 0.4776982503799589 mean 0.0028548691882356782 std\n",
      "Range of scores: 0.46867657315634065 to 0.4912579775746316\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[484   0]\n",
      " [390   0]]\n",
      "Normalized matrix\n",
      " [[0.55377574 0.        ]\n",
      " [0.44622426 0.        ]]\n",
      "Extreme scores correct, pos:neg 0 0\n",
      "Extreme scores incorrect pos:neg 0 0\n",
      "Accuracy: 55.38% Precision: 0.00% Recall: 0.00%\n",
      "F1: 0.00% MCC: 0.0000\n",
      "AUPRC: 54.32% AUROC: 62.64%\n",
      " accuracy [55.37757437070938]\n",
      " precision [0.0]\n",
      " recall [0.0]\n",
      " F1 [0.0]\n",
      " MCC [0.0]\n",
      " AUPRC [54.32138711160819]\n",
      " AUROC [62.636681500317856]\n",
      "\n",
      "Training # 1 4\n",
      "2023-04-18 08:45:45.828263\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.1.4.validation_genes.txt\n",
      "2023-04-18 08:45:45.834884\n",
      "Discriminator ZERO -1\n",
      "Train 1:0 1637:1860 46.81%\n",
      "Valid 1:0 424:450 48.51%\n",
      "FIT\n",
      "2023-04-18 08:45:45.873720\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1334: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, msg_start, len(result))\n",
      "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1334: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, msg_start, len(result))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Testing # 1 4\n",
      "2023-04-18 08:45:45.953917\n",
      "2023-04-18 08:45:45.953944\n",
      "PREDICT\n",
      "2023-04-18 08:45:45.958704\n",
      "debug pred [0.46659954097016804, 0.47115778366844197, 0.47147000926601773]\n",
      "debug class [0 0 0]\n",
      "Test 1:0 424:450 48.51%\n",
      "Predict 1:0 0:874  0.00%\n",
      "Test sizes (874, 1025) (874,)\n",
      "Distrib of scores: 0.4679154617912653 mean 0.0034562727478281378 std\n",
      "Range of scores: 0.4581664465389035 to 0.48194950031311423\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[450   0]\n",
      " [424   0]]\n",
      "Normalized matrix\n",
      " [[0.51487414 0.        ]\n",
      " [0.48512586 0.        ]]\n",
      "Extreme scores correct, pos:neg 0 0\n",
      "Extreme scores incorrect pos:neg 0 0\n",
      "Accuracy: 51.49% Precision: 0.00% Recall: 0.00%\n",
      "F1: 0.00% MCC: 0.0000\n",
      "AUPRC: 55.45% AUROC: 59.74%\n",
      " accuracy [51.48741418764302]\n",
      " precision [0.0]\n",
      " recall [0.0]\n",
      " F1 [0.0]\n",
      " MCC [0.0]\n",
      " AUPRC [55.44689246719463]\n",
      " AUROC [59.74266247379454]\n",
      "\n",
      "Training # 1 5\n",
      "2023-04-18 08:45:45.970679\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.1.5.validation_genes.txt\n",
      "2023-04-18 08:45:45.977177\n",
      "Discriminator ZERO -1\n",
      "Train 1:0 1634:1864 46.71%\n",
      "Valid 1:0 427:446 48.91%\n",
      "FIT\n",
      "2023-04-18 08:45:46.018224\n",
      "\n",
      "Testing # 1 5\n",
      "2023-04-18 08:45:46.119669\n",
      "2023-04-18 08:45:46.119693\n",
      "PREDICT\n",
      "2023-04-18 08:45:46.124494\n",
      "debug pred [0.4726037550122321, 0.465446026671233, 0.46739527667057384]\n",
      "debug class [0 0 0]\n",
      "Test 1:0 427:446 48.91%\n",
      "Predict 1:0 0:873  0.00%\n",
      "Test sizes (873, 1025) (873,)\n",
      "Distrib of scores: 0.4670215498486897 mean 0.003238986830231763 std\n",
      "Range of scores: 0.456651625541797 to 0.4816445900974007\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[446   0]\n",
      " [427   0]]\n",
      "Normalized matrix\n",
      " [[0.51088202 0.        ]\n",
      " [0.48911798 0.        ]]\n",
      "Extreme scores correct, pos:neg 0 0\n",
      "Extreme scores incorrect pos:neg 0 0\n",
      "Accuracy: 51.09% Precision: 0.00% Recall: 0.00%\n",
      "F1: 0.00% MCC: 0.0000\n",
      "AUPRC: 58.66% AUROC: 62.85%\n",
      " accuracy [51.08820160366552]\n",
      " precision [0.0]\n",
      " recall [0.0]\n",
      " F1 [0.0]\n",
      " MCC [0.0]\n",
      " AUPRC [58.65706491700489]\n",
      " AUROC [62.85483244242342]\n",
      "\n",
      "Training # 2 1\n",
      "2023-04-18 08:45:46.136574\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.2.1.validation_genes.txt\n",
      "2023-04-18 08:45:46.143330\n",
      "Discriminator ZERO -1\n",
      "Train 1:0 1655:1841 47.34%\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1334: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, msg_start, len(result))\n",
      "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1334: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, msg_start, len(result))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Valid 1:0 406:469 46.40%\n",
      "FIT\n",
      "2023-04-18 08:45:46.169281\n",
      "\n",
      "Testing # 2 1\n",
      "2023-04-18 08:45:46.266469\n",
      "2023-04-18 08:45:46.266494\n",
      "PREDICT\n",
      "2023-04-18 08:45:46.271828\n",
      "debug pred [0.4798889901997015, 0.4712280893090313, 0.47382011564669535]\n",
      "debug class [0 0 0]\n",
      "Test 1:0 406:469 46.40%\n",
      "Predict 1:0 0:875  0.00%\n",
      "Test sizes (875, 1025) (875,)\n",
      "Distrib of scores: 0.473221320544831 mean 0.0033781988338725377 std\n",
      "Range of scores: 0.463140427839327 to 0.4863787830651503\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[469   0]\n",
      " [406   0]]\n",
      "Normalized matrix\n",
      " [[0.536 0.   ]\n",
      " [0.464 0.   ]]\n",
      "Extreme scores correct, pos:neg 0 0\n",
      "Extreme scores incorrect pos:neg 0 0\n",
      "Accuracy: 53.60% Precision: 0.00% Recall: 0.00%\n",
      "F1: 0.00% MCC: 0.0000\n",
      "AUPRC: 52.73% AUROC: 59.79%\n",
      " accuracy [53.6]\n",
      " precision [0.0]\n",
      " recall [0.0]\n",
      " F1 [0.0]\n",
      " MCC [0.0]\n",
      " AUPRC [52.72616331510953]\n",
      " AUROC [59.78867100108185]\n",
      "\n",
      "Training # 2 2\n",
      "2023-04-18 08:45:46.284953\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.2.2.validation_genes.txt\n",
      "2023-04-18 08:45:46.292112\n",
      "Discriminator ZERO -1\n",
      "Train 1:0 1645:1851 47.05%\n",
      "Valid 1:0 416:459 47.54%\n",
      "FIT\n",
      "2023-04-18 08:45:46.317616\n",
      "\n",
      "Testing # 2 2\n",
      "2023-04-18 08:45:46.445097\n",
      "2023-04-18 08:45:46.445125\n",
      "PREDICT\n",
      "2023-04-18 08:45:46.449888\n",
      "debug pred [0.4687786885662563, 0.46912579151771094, 0.4693234973112632]\n",
      "debug class [0 0 0]\n",
      "Test 1:0 416:459 47.54%\n",
      "Predict 1:0 0:875  0.00%\n",
      "Test sizes (875, 1025) (875,)\n",
      "Distrib of scores: 0.4707248530782465 mean 0.003116018909605544 std\n",
      "Range of scores: 0.4594644902255568 to 0.4838878932164928\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[459   0]\n",
      " [416   0]]\n",
      "Normalized matrix\n",
      " [[0.52457143 0.        ]\n",
      " [0.47542857 0.        ]]\n",
      "Extreme scores correct, pos:neg 0 0\n",
      "Extreme scores incorrect pos:neg 0 0\n",
      "Accuracy: 52.46% Precision: 0.00% Recall: 0.00%\n",
      "F1: 0.00% MCC: 0.0000\n",
      "AUPRC: 57.79% AUROC: 62.61%\n",
      " accuracy [52.457142857142856]\n",
      " precision [0.0]\n",
      " recall [0.0]\n",
      " F1 [0.0]\n",
      " MCC [0.0]\n",
      " AUPRC [57.78659925643406]\n",
      " AUROC [62.612598458186696]\n",
      "\n",
      "Training # 2 3\n",
      "2023-04-18 08:45:46.462289\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.2.3.validation_genes.txt\n",
      "2023-04-18 08:45:46.468963\n",
      "Discriminator ZERO -1\n",
      "Train 1:0 1634:1863 46.73%\n",
      "Valid 1:0 427:447 48.86%\n",
      "FIT\n",
      "2023-04-18 08:45:46.491517\n",
      "\n",
      "Testing # 2 3\n",
      "2023-04-18 08:45:46.574474\n",
      "2023-04-18 08:45:46.574493\n",
      "PREDICT\n",
      "2023-04-18 08:45:46.578868\n",
      "debug pred [0.4714396652206534, 0.46751384749230873, 0.470741888446535]\n",
      "debug class [0 0 0]\n",
      "Test 1:0 427:447 48.86%\n",
      "Predict 1:0 0:874  0.00%\n",
      "Test sizes (874, 1025) (874,)\n",
      "Distrib of scores: 0.46734450595279 mean 0.0033199097982535156 std\n",
      "Range of scores: 0.45666271589964824 to 0.4788561153572807\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[447   0]\n",
      " [427   0]]\n",
      "Normalized matrix\n",
      " [[0.51144165 0.        ]\n",
      " [0.48855835 0.        ]]\n",
      "Extreme scores correct, pos:neg 0 0\n",
      "Extreme scores incorrect pos:neg 0 0\n",
      "Accuracy: 51.14% Precision: 0.00% Recall: 0.00%\n",
      "F1: 0.00% MCC: 0.0000\n",
      "AUPRC: 56.98% AUROC: 60.54%\n",
      " accuracy [51.1441647597254]\n",
      " precision [0.0]\n",
      " recall [0.0]\n",
      " F1 [0.0]\n",
      " MCC [0.0]\n",
      " AUPRC [56.97916341282434]\n",
      " AUROC [60.535760128674646]\n",
      "\n",
      "Training # 2 4\n",
      "2023-04-18 08:45:46.590529\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.2.4.validation_genes.txt\n",
      "2023-04-18 08:45:46.603062\n",
      "Discriminator ZERO -1\n",
      "Train 1:0 1669:1829 47.71%\n",
      "Valid 1:0 392:481 44.90%\n",
      "FIT\n",
      "2023-04-18 08:45:46.632470\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1334: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, msg_start, len(result))\n",
      "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1334: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, msg_start, len(result))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Testing # 2 4\n",
      "2023-04-18 08:45:46.786501\n",
      "2023-04-18 08:45:46.786547\n",
      "PREDICT\n",
      "2023-04-18 08:45:46.791846\n",
      "debug pred [0.47575004326992154, 0.48566763087198045, 0.4792647061858054]\n",
      "debug class [0 0 0]\n",
      "Test 1:0 392:481 44.90%\n",
      "Predict 1:0 0:873  0.00%\n",
      "Test sizes (873, 1025) (873,)\n",
      "Distrib of scores: 0.4770209531399626 mean 0.003036323854655217 std\n",
      "Range of scores: 0.46674871875546575 to 0.4901512095271487\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[481   0]\n",
      " [392   0]]\n",
      "Normalized matrix\n",
      " [[0.55097365 0.        ]\n",
      " [0.44902635 0.        ]]\n",
      "Extreme scores correct, pos:neg 0 0\n",
      "Extreme scores incorrect pos:neg 0 0\n",
      "Accuracy: 55.10% Precision: 0.00% Recall: 0.00%\n",
      "F1: 0.00% MCC: 0.0000\n",
      "AUPRC: 54.84% AUROC: 62.26%\n",
      " accuracy [55.09736540664376]\n",
      " precision [0.0]\n",
      " recall [0.0]\n",
      " F1 [0.0]\n",
      " MCC [0.0]\n",
      " AUPRC [54.83532054737381]\n",
      " AUROC [62.25921761636046]\n",
      "\n",
      "Training # 2 5\n",
      "2023-04-18 08:45:46.807453\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.2.5.validation_genes.txt\n",
      "2023-04-18 08:45:46.815110\n",
      "Discriminator ZERO -1\n",
      "Train 1:0 1641:1856 46.93%\n",
      "Valid 1:0 420:454 48.05%\n",
      "FIT\n",
      "2023-04-18 08:45:46.845627\n",
      "\n",
      "Testing # 2 5\n",
      "2023-04-18 08:45:46.957922\n",
      "2023-04-18 08:45:46.957946\n",
      "PREDICT\n",
      "2023-04-18 08:45:46.962315\n",
      "debug pred [0.4703434707757107, 0.4707715191700156, 0.4688799594860608]\n",
      "debug class [0 0 0]\n",
      "Test 1:0 420:454 48.05%\n",
      "Predict 1:0 0:874  0.00%\n",
      "Test sizes (874, 1025) (874,)\n",
      "Distrib of scores: 0.4692564442863025 mean 0.002954761619112505 std\n",
      "Range of scores: 0.46036913320172684 to 0.4800902594019356\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[454   0]\n",
      " [420   0]]\n",
      "Normalized matrix\n",
      " [[0.5194508 0.       ]\n",
      " [0.4805492 0.       ]]\n",
      "Extreme scores correct, pos:neg 0 0\n",
      "Extreme scores incorrect pos:neg 0 0\n",
      "Accuracy: 51.95% Precision: 0.00% Recall: 0.00%\n",
      "F1: 0.00% MCC: 0.0000\n",
      "AUPRC: 58.66% AUROC: 64.10%\n",
      " accuracy [51.94508009153318]\n",
      " precision [0.0]\n",
      " recall [0.0]\n",
      " F1 [0.0]\n",
      " MCC [0.0]\n",
      " AUPRC [58.65745745273713]\n",
      " AUROC [64.10425844346548]\n",
      "2023-04-18 08:45:46.977751\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1334: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, msg_start, len(result))\n",
      "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1334: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, msg_start, len(result))\n"
     ]
    }
   ],
   "source": [
    "print(datetime.now())\n",
    "\n",
    "accuracy=[]\n",
    "precision=[]\n",
    "recall=[]\n",
    "f1=[]\n",
    "mcc=[]\n",
    "auprc=[]\n",
    "auroc=[]\n",
    "\n",
    "loader = DataLoader()\n",
    "filepath = DATA_DIR+RCI_FILE_TRAIN\n",
    "print(\"Load RCI from\",filepath)\n",
    "gene_to_rci = loader.load_gene_rci_values(filepath,EXCLUSIONS)\n",
    "print('Num RCI:', len(gene_to_rci.keys()))\n",
    "filepath = DATA_DIR+SEQ_FILE_TRAIN\n",
    "print('Load sequence from',filepath)\n",
    "allids,allseq = loader.load_spectra(filepath)  # for MLP (load_sequence() for CNN)\n",
    "print('Loaded',len(allseq),'sequences.')\n",
    "test_gene_to_rci = None\n",
    "test_allids = None\n",
    "test_allseq = None\n",
    "if SEQ_FILE_TEST is not None:\n",
    "    # Train on the entire train set (no cross-validation).\n",
    "    # Evaluate with the test files.\n",
    "    test_loader = DataLoader()\n",
    "    filepath = DATA_DIR+RCI_FILE_TEST\n",
    "    print(\"Load RCI from\",filepath)\n",
    "    test_gene_to_rci = test_loader.load_gene_rci_values(filepath,EXCLUSIONS)\n",
    "    print('Num RCI:', len(test_gene_to_rci.keys()))\n",
    "    filepath = DATA_DIR+SEQ_FILE_TEST\n",
    "    print('Load sequence from',filepath)\n",
    "    test_allids,test_allseq = test_loader.load_spectra(filepath)\n",
    "    print('Loaded',len(test_allseq),'sequences.')\n",
    "\n",
    "for repeat in range(REPEATS):\n",
    "    for fold in range(FOLDS):\n",
    "        show_r = repeat+1  # display one-based counting\n",
    "        show_f = fold+1    # display one-based counting\n",
    "\n",
    "        print()\n",
    "        print(\"Training #\",show_r,show_f)\n",
    "        print(datetime.now())\n",
    "        cvdo = CrossValidator(EPOCHS)\n",
    "        cvdo.set_threshold_mechanism(RCI_THRESHOLD_MECHANISM)\n",
    "        if SEQ_FILE_TEST is None:\n",
    "            # Train on 80% and evaluate on 20%.\n",
    "            separator = Separator()\n",
    "            separator.load(DATA_DIR,show_r,show_f)\n",
    "            separator.process(allids,allseq,gene_to_rci)\n",
    "            train_allids,test_allids = separator.get_ids()\n",
    "            train_allseq,test_allseq = separator.get_seq()\n",
    "            train_gene_to_rci,test_gene_to_rci = separator.get_rci()\n",
    "            cvdo.train_new_model(\n",
    "                train_allids,train_allseq,train_gene_to_rci,\n",
    "                test_allids,test_allseq,test_gene_to_rci)\n",
    "            if SAVE_MODEL_FILENAME is not None:\n",
    "                filename = f\"{SAVE_MODEL_FILENAME}.{show_r}.{show_f}.model\"\n",
    "                cvdo.save_model(filename)\n",
    "        else:\n",
    "            # Train on the entire train set (no cross-validation).\n",
    "            # Evaluate with the test files.\n",
    "            train_allids = allids\n",
    "            train_allseq = allseq\n",
    "            train_gene_to_rci = gene_to_rci\n",
    "            BREAK = True\n",
    "            cvdo.train_new_model(\n",
    "                train_allids,train_allseq,train_gene_to_rci,\n",
    "                None,None,None)\n",
    "\n",
    "        print()\n",
    "        print(\"Testing #\",show_r,show_f)\n",
    "        print(datetime.now())\n",
    "        cvdo.reset_statistics()\n",
    "        cvdo.test_without_training(\n",
    "            test_allids,test_allseq,test_gene_to_rci)\n",
    "        cv_accuracy,cv_precision,cv_recall,cv_f1,cv_mcc,cv_auprc,cv_auroc=\\\n",
    "            cvdo.get_statistics()\n",
    "\n",
    "        print(\" accuracy\" ,  cv_accuracy)\n",
    "        print(\" precision\" , cv_precision)\n",
    "        print(\" recall\" ,    cv_recall)\n",
    "        print(\" F1\" ,        cv_f1)\n",
    "        print(\" MCC\" ,       cv_mcc)\n",
    "        print(\" AUPRC\" ,     cv_auprc)\n",
    "        print(\" AUROC\" ,     cv_auroc)\n",
    "\n",
    "        accuracy.append(cv_accuracy)\n",
    "        precision.append(cv_precision)\n",
    "        recall.append(cv_recall)\n",
    "        f1.append(cv_f1)\n",
    "        mcc.append(cv_mcc)\n",
    "        auprc.append(cv_auprc)\n",
    "        auroc.append(cv_auroc)\n",
    "        if BREAK: break\n",
    "    if BREAK: break\n",
    "print(datetime.now())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "HkCeDg_HdQ36",
    "outputId": "b8486985-4447-4c8b-dde8-bbda95f9cc6d"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " accuracy mean 52.85% std 1.59\n",
      " precision mean 0.00% std 0.00\n",
      " recall mean 0.00% std 0.00\n",
      " F1 mean 0.00% std 0.00\n",
      " MCC mean 0.0000 std 0.0000\n",
      " AUPRC mean 56.21% std 1.94\n",
      " AUROC mean 61.90% std 1.49\n",
      " accuracy [[54.05714285714286], [52.22857142857143], [55.37757437070938], [51.48741418764302], [51.08820160366552], [53.6], [52.457142857142856], [51.1441647597254], [55.09736540664376], [51.94508009153318]]\n",
      " precision [[0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0]]\n",
      " recall [[0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0]]\n",
      " F1 [[0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0]]\n",
      " MCC [[0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0]]\n",
      " AUPRC [[55.79731981219367], [56.8583036142403], [54.32138711160819], [55.44689246719463], [58.65706491700489], [52.72616331510953], [57.78659925643406], [56.97916341282434], [54.83532054737381], [58.65745745273713]]\n",
      " AUROC [[61.291323509303375], [63.152659847350634], [62.636681500317856], [59.74266247379454], [62.85483244242342], [59.78867100108185], [62.612598458186696], [60.535760128674646], [62.25921761636046], [64.10425844346548]]\n"
     ]
    }
   ],
   "source": [
    "def STD (values):\n",
    "    # ddof=1 reduces bias when extrapolating from sample to population\n",
    "    return np.std(values,ddof=1)\n",
    "\n",
    "print(\" accuracy mean %.2f%% std %.2f\" %  (np.mean(accuracy),  STD(accuracy)))\n",
    "print(\" precision mean %.2f%% std %.2f\" % (np.mean(precision), STD(precision)))\n",
    "print(\" recall mean %.2f%% std %.2f\" %    (np.mean(recall),    STD(recall)))\n",
    "print(\" F1 mean %.2f%% std %.2f\" %        (np.mean(f1),        STD(f1)))\n",
    "print(\" MCC mean %.4f std %.4f\" %       (np.mean(mcc),       STD(mcc)))\n",
    "print(\" AUPRC mean %.2f%% std %.2f\" %     (np.mean(auprc),     STD(auprc)))\n",
    "print(\" AUROC mean %.2f%% std %.2f\" %     (np.mean(auroc),     STD(auroc)))\n",
    "\n",
    "print(\" accuracy\"  , accuracy)\n",
    "print(\" precision\" , precision)\n",
    "print(\" recall\"    , recall)\n",
    "print(\" F1\"        , f1)\n",
    "print(\" MCC\"       , mcc)\n",
    "print(\" AUPRC\"     , auprc)\n",
    "print(\" AUROC\"     , auroc)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "QjSVa72v4IsA"
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "gpuClass": "standard",
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
