{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "PG-tGRnlFLA3"
   },
   "source": [
    "# Random Forest\n",
    "canonical lncRNA, -1 threshold, cross-valiation, all the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "0RmwUsVLFLA6",
    "outputId": "64cc44e8-c4cd-4eee-f276-47d1983f985c"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2023-04-17 18:37:56.053939\n"
     ]
    }
   ],
   "source": [
    "from datetime import datetime\n",
    "print(datetime.now())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "OlzN9OdsFWEU",
    "outputId": "20173346-1246-44f4-eaf7-2a5f2c032df7"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-04-17 18:37:56.110444: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GPU device not found\n",
      "/Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-04-17 18:38:05.282374: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\n"
     ]
    }
   ],
   "source": [
    "import tensorflow as tf\n",
    "dt='float32'\n",
    "tf.keras.backend.set_floatx('float32')\n",
    "# tf.random.set_seed(42) # supposedly leads to reproducible results\n",
    "\n",
    "device_name = tf.test.gpu_device_name()\n",
    "if device_name != '/device:GPU:0':\n",
    "    print('GPU device not found')\n",
    "else:\n",
    "    print('Found GPU at: {}'.format(device_name))\n",
    "\n",
    "try:\n",
    "    from google.colab import drive\n",
    "    IN_COLAB = True\n",
    "    print('Running on CoLab')\n",
    "    PATH='/content/drive/'\n",
    "    drive.mount(PATH)\n",
    "    DATA_DIR=PATH+'My Drive/data/Localization/TrainTest/TrainTest_ver43/'  # must end in \"/\"\n",
    "    MODEL_DIR=PATH+'My Drive/data/Localization/Models/'  # must end in \"/\"\n",
    "except:\n",
    "    IN_COLAB = False\n",
    "    DATA_DIR = 'D:/Adjeroh/Localization/TrainTest/'   # Windows\n",
    "    DATA_DIR = '/Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/'    # Mac\n",
    "    MODEL_DIR = '/Users/jasonmiller/WVU/Localization/Models/'    # Mac\n",
    "print(DATA_DIR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "PRX-UEr8FLA8",
    "outputId": "39890dac-1ca8-4ba2-9229-2b581f26b315"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Python 3.10.0\n",
      "sklearn 1.1.2\n"
     ]
    }
   ],
   "source": [
    "from platform import python_version\n",
    "print('Python',python_version())\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy.stats as ss\n",
    "import pickle\n",
    "import time # sleep function\n",
    "from os.path import isfile\n",
    "from matplotlib import pyplot as plt \n",
    "import sklearn   # pip install --upgrade scikit-learn\n",
    "print('sklearn',sklearn.__version__)\n",
    "\n",
    "from tensorflow import keras\n",
    "from sklearn.mixture import GaussianMixture\n",
    "from sklearn.model_selection import KFold\n",
    "from sklearn.metrics import accuracy_score\n",
    "from sklearn.metrics import precision_score\n",
    "from sklearn.metrics import recall_score\n",
    "from sklearn.metrics import f1_score\n",
    "from sklearn.metrics import precision_recall_curve\n",
    "from sklearn.metrics import roc_curve\n",
    "from sklearn.metrics import roc_auc_score\n",
    "from sklearn.metrics import auc\n",
    "from sklearn.metrics import matthews_corrcoef\n",
    "from sklearn.utils import class_weight\n",
    "from sklearn.metrics import confusion_matrix\n",
    "from sklearn.ensemble import RandomForestClassifier as RFC\n",
    "\n",
    "from keras.models import Sequential\n",
    "from keras.layers import Masking\n",
    "from keras.layers import Embedding\n",
    "from keras.layers import Conv1D\n",
    "from keras.layers import GlobalAveragePooling1D\n",
    "from keras.layers import GlobalMaxPooling1D\n",
    "from keras.layers import AveragePooling1D\n",
    "from keras.layers import MaxPooling1D\n",
    "from keras.layers import Bidirectional\n",
    "from keras.layers import LSTM\n",
    "from keras.layers import TimeDistributed\n",
    "from keras.layers import Flatten\n",
    "from keras.layers import Dense\n",
    "from keras.layers import Dropout\n",
    "from keras.losses import BinaryCrossentropy\n",
    "#from keras.losses import Hinge\n",
    "\n",
    "K=5\n",
    "ALPHABET=4**K + 1\n",
    "CFILTERS=64 \n",
    "FILTERSIZE=8\n",
    "RCELLS=32\n",
    "DCELLS=16\n",
    "EPOCHS=150 \n",
    "EMBED_DIMEN = 4 # arbitrary hyperparameter\n",
    "# MINLEN=200   # NO LONGER USED\n",
    "# MAXLEN=5000\n",
    "RCI_THRESHOLD_MECHANISM = 'ZERO'  # 'RCI_GMM' 'ZERO' 'THE_MEAN'\n",
    "BREAK = False   # break after first fold\n",
    "EXCLUSIONS = [1]   # possibly exclude cell line 1 = H1.hESC\n",
    "FILTER_TRAIN        = False\n",
    "FILTER_TAILS_TRAIN  = False\n",
    "FILTER_MIDDLE_TRAIN = False\n",
    "FILTER_TEST         = False\n",
    "FILTER_TAILS_TEST   = False\n",
    "FILTER_MIDDLE_TEST  = False\n",
    "REPEATS = 2\n",
    "FOLDS = 5\n",
    "\n",
    "SAVE_MODEL_FILENAME = None # 'RF_101'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "id": "LnkpVKdMFLA-"
   },
   "outputs": [],
   "source": [
    "RCI_FILE_TRAIN = 'train.lncRNA_RCI.csv'\n",
    "RCI_FILE_TEST  = None #'test.lncRNA_RCI.csv'\n",
    "\n",
    "SEQ_FILE_TRAIN = 'train.canon_lncRNA_transcripts.csv'\n",
    "SEQ_FILE_TEST  = None #'test.canon_lncRNA_transcripts.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "id": "e3p4QzQJFLA_"
   },
   "outputs": [],
   "source": [
    "def get_ordered_list():\n",
    "    ordered_list = \\\n",
    "    ['A549','H1.hESC','HeLa.S3','HepG2','HT1080',\\\n",
    "      'HUVEC','MCF.7','NCI.H460','NHEK','SK.MEL.5',\\\n",
    "      'SK.N.DZ','SK.N.SH','GM12878','K562','IMR.90']\n",
    "    return ordered_list\n",
    "all_cell_lines = get_ordered_list()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "UtqdpJOxFLBA"
   },
   "source": [
    "## Data Load"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "id": "p35ehKV3Kq0z"
   },
   "outputs": [],
   "source": [
    "class DataLoader():\n",
    "    def __init__(self):\n",
    "        self.cache=dict() \n",
    "        self.vals = {'A':0, 'C':1, 'G':2, 'T':3}\n",
    "        self.gene2rci = dict()\n",
    "        \n",
    "    def load_gene_rci_values(self,filepath,exclusions):\n",
    "        '''\n",
    "        Load all the genes from the given RCI csv file.\n",
    "        The given file usually contains train or test, not both.\n",
    "        Return dict with keys=gene:str and values=RCI:float.\n",
    "        '''\n",
    "        self.gene2rci = {}\n",
    "        overall_sum = 0\n",
    "        with open (filepath,'r') as handle:\n",
    "            header = None\n",
    "            for row in handle:\n",
    "                if header is None:\n",
    "                    header = row # skip file's header line\n",
    "                else:\n",
    "                    line = row.strip()\n",
    "                    fields = line.split(',')\n",
    "                    gene_id = fields.pop(0)\n",
    "                    log_mean = float(fields.pop(0))\n",
    "                    self.gene2rci[gene_id] = log_mean\n",
    "        print('Number of RCI values loaded',len(self.gene2rci.keys()))\n",
    "        return self.gene2rci\n",
    "\n",
    "    def _seq_to_kmer_values(self,rna,K):\n",
    "        vec=[] # seq converted to list of K-mers \n",
    "        N_indicator = 0 # indicator value\n",
    "        length = len(rna)\n",
    "        for i in range(length-K+1):\n",
    "            kmer = rna[i:i+K]\n",
    "            if 'N' in kmer:\n",
    "                value = N_indicator\n",
    "            elif kmer in self.cache.keys():\n",
    "                value = self.cache[kmer]\n",
    "            else:\n",
    "                value = 0\n",
    "                for j in range(K):\n",
    "                    value *= 4   \n",
    "                    nextnuc = kmer[j] \n",
    "                    nucval = self.vals[nextnuc]\n",
    "                    value += nucval\n",
    "                value += 1   # NNN => 0, AAA => 1\n",
    "                self.cache[kmer] = value\n",
    "            vec.append(value)\n",
    "        return vec\n",
    "\n",
    "    def load_sequence(self,filepath):\n",
    "        '''\n",
    "        Load all the sequences from the given file. \n",
    "        Load our version of GenCode -- csv, easier to parse than fasta.\n",
    "        Each line has IDs plus sequence.\n",
    "        The IDs in the file do not include dot-version numbers.\n",
    "        The file may already be filtered e.g. canonical transcripts.\n",
    "        '''\n",
    "        allids=[]\n",
    "        allseq=[]\n",
    "        #NREPEAT = str('N'*MAXLEN)   # not used for MLP\n",
    "        with open (filepath,'r') as handle:\n",
    "            header = None\n",
    "            for row in handle:\n",
    "                if header is None:\n",
    "                    header = row\n",
    "                else:\n",
    "                    line    = row.strip()\n",
    "                    fields  = line.split(',')\n",
    "                    tran_id = fields[0]  # without version number\n",
    "                    gene_id = fields[1]  # without version number\n",
    "                    seq_len = int(fields[3])\n",
    "                    seq_txt = fields[4]\n",
    "                    # Keep only transcripts having numeric RCI given the cell lines in use.\n",
    "                    # We have validated this by spot checking.\n",
    "                    # TO DO: validate this programmatically.\n",
    "                    if gene_id in self.gene2rci.keys():\n",
    "                        # no MAXLEN for MLP\n",
    "                        #if seq_len<=MAXLEN:\n",
    "                        #    seq_txt = seq_txt + NREPEAT\n",
    "                        #seq_txt = seq_txt[:MAXLEN]\n",
    "                        allids.append( (gene_id,tran_id) )\n",
    "                        hot_vec = self._seq_to_kmer_values(seq_txt,K)\n",
    "                        allseq.append(hot_vec)\n",
    "        self.cache=dict() # save RAM\n",
    "        return allids,allseq\n",
    "\n",
    "    def load_spectra(self,filepath):\n",
    "        '''\n",
    "        Load all (variable-length) sequences as lists of kmers.\n",
    "        Then convert each sequence to (fixed-length) kmer histograms.\n",
    "        '''\n",
    "        allids,allseq = self.load_sequence(filepath)\n",
    "        allspectra = []\n",
    "        for seq in allseq:\n",
    "            spectrum = np.zeros(ALPHABET)\n",
    "            for kmer in seq:\n",
    "                spectrum[kmer] += 1\n",
    "            spectrum /= len(seq)\n",
    "            allspectra.append(spectrum)\n",
    "        return allids,allspectra        \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "VDZ6siB_Kq04"
   },
   "source": [
    "## Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "id": "AwMbRjm0FLBF"
   },
   "outputs": [],
   "source": [
    "def build_model():\n",
    "    rfc = RFC()\n",
    "    return rfc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "clj-wufgFLBF",
    "outputId": "a17cf7a6-3755-48aa-d96b-dd0383c6f2f2"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2023-04-17 18:38:07.971513\n",
      "RandomForestClassifier()\n"
     ]
    }
   ],
   "source": [
    "print(datetime.now())\n",
    "model=build_model()\n",
    "print(model)  # Print this only once\n",
    "model=None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "MgrC1alOKq07"
   },
   "source": [
    "## Training"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "id": "W9xiFzNbFLBE"
   },
   "outputs": [],
   "source": [
    "import gc\n",
    "class CrossValidator():\n",
    "    def __init__(self,epochs,score_threshold=0.5):\n",
    "        self.epochs = epochs\n",
    "        self.score_threshold = score_threshold\n",
    "        self.mechanism = 'ZERO'\n",
    "        self.discriminator = -1 # or zero\n",
    "        self.flip = False\n",
    "        self.reset_statistics()\n",
    "        \n",
    "    def reset_statistics(self):\n",
    "        self.cv_accuracy=[]\n",
    "        self.cv_precision=[]\n",
    "        self.cv_recall=[]\n",
    "        self.cv_f1=[]\n",
    "        self.cv_auprc=[]\n",
    "        self.cv_auroc=[]\n",
    "        self.cv_mcc=[]\n",
    "        \n",
    "    def _get_X_y(self, all_ids, all_seqs, rci_map): \n",
    "        # Prepare X and y for training or testing.\n",
    "        subsetX=[]\n",
    "        subsetY=[]\n",
    "        for t in range(len(all_ids)):\n",
    "            gene_id,tran_id = all_ids[t]\n",
    "            oneX            = all_seqs[t]\n",
    "            oneY            = rci_map[gene_id]\n",
    "            subsetX.append(oneX)\n",
    "            subsetY.append(oneY)\n",
    "        subsetX = np.array(subsetX)\n",
    "        subsetY = np.array(subsetY).reshape((-1,1))\n",
    "        return subsetX,subsetY\n",
    "    \n",
    "    def set_threshold_mechanism(self, mechanism):\n",
    "        if mechanism not in ['RCI_GMM','THE_MEAN','ZERO']:\n",
    "            raise Exception('Unrecognized mechansm:',mechanism)\n",
    "        self.mechanism = mechanism\n",
    "    \n",
    "    def _apply_threshold(self, array_of_rci):\n",
    "        # Takes list of float, returns list of labels [0,1].\n",
    "        if self.mechanism == 'RCI_GMM':\n",
    "            labels = self.discriminator.predict(array_of_rci)\n",
    "            if self.flip:\n",
    "                IS_CYTO = lambda label: 1 if label==0 else 0\n",
    "                labels = np.array(list(map(IS_CYTO, labels)))\n",
    "        else:  # 'THE_MEAN' or 'ZERO'\n",
    "            rci_threshold = self.discriminator\n",
    "            IS_CYTO = lambda rci: 1 if rci>rci_threshold else 0\n",
    "            labels = np.array(list(map(IS_CYTO, array_of_rci)))\n",
    "        return labels\n",
    "    \n",
    "    def _prepare_threshold(self, rci_values, create=True):\n",
    "        if self.mechanism == 'RCI_GMM':\n",
    "            if create:  # during training, create a new GMM\n",
    "                gmm = GaussianMixture(n_components=2, verbose=0, \n",
    "                  covariance_type='spherical', n_init=100) # random_state=42) \n",
    "                gmm.fit(rci_values)\n",
    "            else:   # during testing, use existing GMM\n",
    "                gmm=self.discriminator\n",
    "            self.flip = False\n",
    "            # The GMM labels are arbitrary.\n",
    "            if gmm.means_[0][0] > gmm.means_[1][0]:\n",
    "                self.flip = True\n",
    "            self.discriminator = gmm   # redundant but consistent\n",
    "        elif self.mechanism == 'THE_MEAN':\n",
    "            self.discriminator = np.mean(rci_values)\n",
    "        elif self.mechanism == 'ZERO':\n",
    "            self.discriminator = -1   # 0 usually, -1 is as in Yuan et al.\n",
    "        else: # not expected\n",
    "            self.discriminator = 0\n",
    "    \n",
    "    def _explain_threshold(self):\n",
    "        if self.mechanism == 'RCI_GMM':\n",
    "            gmm=self.discriminator\n",
    "            print('Discriminator is GMM')\n",
    "            print('Means',[gmm.means_[0][0],gmm.means_[1][0]])\n",
    "            print('Variances',gmm.covariances_)\n",
    "            print('Priors',gmm.weights_)\n",
    "            test_rcis=[-5,-4,-3.5,-3,-2.5,-2,-1.5,-1,-0.5,0,0.5,1,1.5,2,3]\n",
    "            print(test_rcis)\n",
    "            print(self._apply_threshold(np.array(test_rcis).reshape((-1,1))))\n",
    "        else:\n",
    "            print('Discriminator',self.mechanism,self.discriminator)\n",
    "    \n",
    "    def _show_sizes(self,label,values):\n",
    "        a = np.count_nonzero(values==1)\n",
    "        b = np.count_nonzero(values==0)\n",
    "        print('%s 1:0 %d:%d %5.2f%%'%(label,a,b,100*a/(a+b)))\n",
    "        \n",
    "    def save_model(self,filename):\n",
    "        if self.model is not None:\n",
    "            filepath = MODEL_DIR + filename\n",
    "            #self.model.save(filepath)\n",
    "            print('? Saved model to',filepath)\n",
    "        \n",
    "    def load_model(self,filename):\n",
    "        filepath = MODEL_DIR + filename\n",
    "        #self.model = keras.models.load_model(filepath)\n",
    "        print('? Loaded model from',filepath)\n",
    "        \n",
    "    def train_new_model(self,train_ids,train_seq,train_rci,\n",
    "            valid_ids=None,valid_seq=None,valid_rci=None):\n",
    "        print(datetime.now())\n",
    "        X_train,y_rci = self._get_X_y(train_ids,train_seq,train_rci) \n",
    "        self._prepare_threshold(y_rci,True)  \n",
    "        self._explain_threshold()\n",
    "        y_train = self._apply_threshold(y_rci)\n",
    "        self._show_sizes('Train',y_train)\n",
    "        #cw = class_weight.compute_class_weight('balanced', classes=[0,1], y=y_train)\n",
    "        #cwd = {0: cw[0], 1: cw[1]}\n",
    "        #print('Computed class weights:',cwd)\n",
    "        if valid_ids is not None:\n",
    "            X_valid,y_rci = self._get_X_y(valid_ids,valid_seq,valid_rci) \n",
    "            y_valid = self._apply_threshold(y_rci)\n",
    "            self._show_sizes('Valid',y_valid)\n",
    "        y_rci = None\n",
    "\n",
    "        self.model=build_model()\n",
    "        \n",
    "        print(\"FIT\")\n",
    "        print(datetime.now())\n",
    "        self.model.fit(X_train, y_train) # sample weight\n",
    "\n",
    "    def test_without_training(self,test_ids,test_seq,test_rci):\n",
    "        # For final test, do no train.\n",
    "        # Assume set_sequences() set the test set.\n",
    "        print(datetime.now())\n",
    "        X_test,y_rci = self._get_X_y(test_ids,test_seq,test_rci) \n",
    "        y_test = self._apply_threshold(y_rci)\n",
    "        y_rci = None\n",
    "        \n",
    "        print(\"PREDICT\")\n",
    "        print(datetime.now())        \n",
    "        yhat_pairs=self.model.predict_proba(X_test)  # [ prob of 0, prob of 1 ]\n",
    "        yhat_pred=[pair[1] for pair in yhat_pairs]\n",
    "        yhat_classes=self.model.predict(X_test)  # 0 or 1\n",
    "        \n",
    "        print('debug pred',yhat_pred[:3])\n",
    "        print('debug class',yhat_classes[:3])\n",
    "\n",
    "        self._show_sizes('Test',y_test)\n",
    "        self._show_sizes('Predict',yhat_classes)\n",
    "        print('Test sizes',X_test.shape,y_test.shape)\n",
    "        print('Distrib of scores:',np.mean(yhat_pred),'mean',np.std(yhat_pred),'std')\n",
    "        print('Range of scores:',np.min(yhat_pred),'to',np.max(yhat_pred))\n",
    "        print('Score threshold',self.score_threshold)\n",
    "        cm1 = confusion_matrix(y_test,yhat_classes)\n",
    "        print('Confusion matrix\\n',cm1)\n",
    "        cm2 = confusion_matrix(y_test,yhat_classes,normalize='all')\n",
    "        print('Normalized matrix\\n',cm2)\n",
    "\n",
    "        accuracy = accuracy_score(y_test, yhat_classes)*100.\n",
    "        precision = precision_score(y_test, yhat_classes)*100.\n",
    "        recall = recall_score(y_test, yhat_classes)*100.\n",
    "        f1 = f1_score(y_test, yhat_classes)*100.\n",
    "        prc_Y, prc_X, prc_bins = precision_recall_curve(y_test, yhat_pred)\n",
    "        auprc = auc(prc_X,prc_Y)*100.\n",
    "        auroc = roc_auc_score(y_test, yhat_pred)*100.\n",
    "        mcc = matthews_corrcoef(y_test, yhat_classes)\n",
    "\n",
    "        correct_pos = 0\n",
    "        correct_neg = 0\n",
    "        wrong_pos = 0\n",
    "        wrong_neg = 0\n",
    "        for i in range(len(y_test)):\n",
    "            if yhat_pred[i]>=0.65:\n",
    "                if y_test[i]==1:\n",
    "                    correct_pos += 1\n",
    "                else:\n",
    "                    wrong_pos += 1\n",
    "            elif yhat_pred[i]<=0.35:\n",
    "                if y_test[i]==0:\n",
    "                    correct_neg += 1\n",
    "                else:\n",
    "                    wrong_neg += 1\n",
    "        print('Extreme scores correct, pos:neg',correct_pos,correct_neg)  \n",
    "        print('Extreme scores incorrect pos:neg',wrong_pos,wrong_neg)  \n",
    "\n",
    "        #self._pretty_pictures(y_test,yhat_pred,prc_X,prc_Y)\n",
    "        \n",
    "        print('Accuracy: %.2f%% Precision: %.2f%% Recall: %.2f%%' % (accuracy,precision,recall)) \n",
    "        print('F1: %.2f%% MCC: %.4f' % (f1,mcc)) \n",
    "        print('AUPRC: %.2f%% AUROC: %.2f%%' % (auprc,auroc)) \n",
    "\n",
    "        self.cv_accuracy.append(accuracy)\n",
    "        self.cv_precision.append(precision)\n",
    "        self.cv_recall.append(recall)\n",
    "        self.cv_f1.append(f1)\n",
    "        self.cv_mcc.append(mcc)\n",
    "        self.cv_auprc.append(auprc)\n",
    "        self.cv_auroc.append(auroc)\n",
    "\n",
    "    def _pretty_pictures(self,y_valid,yhat_pred,prc_X,prc_Y):\n",
    "        count_ones= len(y_valid[y_valid==1])\n",
    "        count_zeros= len(y_valid[y_valid==0])\n",
    "        guess = max(count_ones,count_zeros) / len(y_valid)\n",
    "        # PRC\n",
    "        plt.plot(prc_X, prc_Y, marker='.')\n",
    "        plt.plot([0, 1], [guess,guess], linestyle='--')\n",
    "        plt.xlabel('Recall')\n",
    "        plt.ylabel('Precision')\n",
    "        plt.show()\n",
    "        # ROC\n",
    "        fpr, tpr, roc_bins = roc_curve(y_valid, yhat_pred)\n",
    "        plt.plot(fpr, tpr, marker='.')\n",
    "        plt.xlabel('False Positive Rate')\n",
    "        plt.ylabel('True Positive Rate')\n",
    "        plt.show()\n",
    "        \n",
    "    def get_statistics(self):\n",
    "        return \\\n",
    "        self.cv_accuracy,\\\n",
    "        self.cv_precision,\\\n",
    "        self.cv_recall,\\\n",
    "        self.cv_f1,\\\n",
    "        self.cv_mcc,\\\n",
    "        self.cv_auprc,\\\n",
    "        self.cv_auroc\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "Td97uyyj5qDq"
   },
   "source": [
    "## Train"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "id": "poe3rwAF4Ir7"
   },
   "outputs": [],
   "source": [
    "class Separator():\n",
    "    def __init__(self):\n",
    "        self.train_ids = []\n",
    "        self.train_seq = []\n",
    "        self.train_rci = dict()\n",
    "        self.val_ids = []\n",
    "        self.val_seq = []\n",
    "        self.val_rci = dict()\n",
    "    def load(self,data_dir,rep,fold):\n",
    "        filename='cv.{}.{}.validation_genes.txt'.format(rep,fold)\n",
    "        filename = data_dir + filename\n",
    "        self.val_genes = set()\n",
    "        print('Opening file',filename)\n",
    "        with open(filename,'r') as fin:\n",
    "            for line in fin:\n",
    "                gene_id = line.strip()\n",
    "                self.val_genes.add(gene_id)\n",
    "    def process(self,allids,allseq,gene_to_rci):\n",
    "        size = len(allids)\n",
    "        for t in range(size):\n",
    "            gene_id,tran_id = allids[t]\n",
    "            oneX            = allseq[t]\n",
    "            oneY            = gene_to_rci[gene_id]\n",
    "            in_middle = gene_to_rci[gene_id] >= -2 and gene_to_rci[gene_id] <= 0\n",
    "            in_tails = gene_to_rci[gene_id] < -2 or gene_to_rci[gene_id] > 0\n",
    "            if gene_id in self.val_genes:\n",
    "                if FILTER_TEST and (\\\n",
    "                    (FILTER_TAILS_TEST and in_tails) or \\\n",
    "                    (FILTER_MIDDLE_TEST and in_middle)):\n",
    "                    pass\n",
    "                else:\n",
    "                    self.val_ids.append(allids[t])\n",
    "                    self.val_seq.append(allseq[t])\n",
    "                    self.val_rci[gene_id]=gene_to_rci[gene_id]\n",
    "            else:\n",
    "                if FILTER_TRAIN and (\\\n",
    "                    (FILTER_TAILS_TRAIN and in_tails) or \\\n",
    "                    (FILTER_MIDDLE_TRAIN and in_middle)):\n",
    "                    pass\n",
    "                else:\n",
    "                    self.train_ids.append(allids[t])\n",
    "                    self.train_seq.append(allseq[t])\n",
    "                    self.train_rci[gene_id]=gene_to_rci[gene_id]\n",
    "    def get_ids(self):\n",
    "        return self.train_ids,self.val_ids\n",
    "    def get_seq(self):\n",
    "        return self.train_seq,self.val_seq\n",
    "    def get_rci(self):\n",
    "        return self.train_rci,self.val_rci"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 1000
    },
    "id": "XC9m0W-pFLBH",
    "outputId": "f1051adc-9a79-4671-dd2b-2d9617da1129",
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2023-04-17 18:38:08.084579\n",
      "Load RCI from /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/train.lncRNA_RCI.csv\n",
      "Number of RCI values loaded 4371\n",
      "Num RCI: 4371\n",
      "Load sequence from /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/train.canon_lncRNA_transcripts.csv\n",
      "Loaded 4371 sequences.\n",
      "\n",
      "Training # 1 1\n",
      "2023-04-17 18:38:14.249400\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.1.1.validation_genes.txt\n",
      "2023-04-17 18:38:14.257459\n",
      "Discriminator ZERO -1\n",
      "Train 1:0 1659:1837 47.45%\n",
      "Valid 1:0 402:473 45.94%\n",
      "FIT\n",
      "2023-04-17 18:38:14.300809\n",
      "\n",
      "Testing # 1 1\n",
      "2023-04-17 18:38:19.938358\n",
      "2023-04-17 18:38:19.938391\n",
      "PREDICT\n",
      "2023-04-17 18:38:19.943755\n",
      "debug pred [0.5, 0.44, 0.44]\n",
      "debug class [0 0 0]\n",
      "Test 1:0 402:473 45.94%\n",
      "Predict 1:0 356:519 40.69%\n",
      "Test sizes (875, 1025) (875,)\n",
      "Distrib of scores: 0.4825028571428571 mean 0.09785481223502 std\n",
      "Range of scores: 0.04 to 0.98\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[315 158]\n",
      " [204 198]]\n",
      "Normalized matrix\n",
      " [[0.36       0.18057143]\n",
      " [0.23314286 0.22628571]]\n",
      "Extreme scores correct, pos:neg 19 60\n",
      "Extreme scores incorrect pos:neg 9 20\n",
      "Accuracy: 58.63% Precision: 55.62% Recall: 49.25%\n",
      "F1: 52.24% MCC: 0.1608\n",
      "AUPRC: 58.49% AUROC: 63.21%\n",
      " accuracy [58.62857142857143]\n",
      " precision [55.61797752808989]\n",
      " recall [49.25373134328358]\n",
      " F1 [52.242744063324544]\n",
      " MCC [0.1607911545791673]\n",
      " AUPRC [58.48845142284694]\n",
      " AUROC [63.21458247872687]\n",
      "\n",
      "Training # 1 2\n",
      "2023-04-17 18:38:20.035262\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.1.2.validation_genes.txt\n",
      "2023-04-17 18:38:20.049072\n",
      "Discriminator ZERO -1\n",
      "Train 1:0 1643:1853 47.00%\n",
      "Valid 1:0 418:457 47.77%\n",
      "FIT\n",
      "2023-04-17 18:38:20.083423\n",
      "\n",
      "Testing # 1 2\n",
      "2023-04-17 18:38:25.644000\n",
      "2023-04-17 18:38:25.644030\n",
      "PREDICT\n",
      "2023-04-17 18:38:25.649251\n",
      "debug pred [0.54, 0.34, 0.55]\n",
      "debug class [1 0 1]\n",
      "Test 1:0 418:457 47.77%\n",
      "Predict 1:0 353:522 40.34%\n",
      "Test sizes (875, 1025) (875,)\n",
      "Distrib of scores: 0.47308571428571433 mean 0.09874219576498083 std\n",
      "Range of scores: 0.13 to 0.89\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[310 147]\n",
      " [212 206]]\n",
      "Normalized matrix\n",
      " [[0.35428571 0.168     ]\n",
      " [0.24228571 0.23542857]]\n",
      "Extreme scores correct, pos:neg 18 75\n",
      "Extreme scores incorrect pos:neg 3 25\n",
      "Accuracy: 58.97% Precision: 58.36% Recall: 49.28%\n",
      "F1: 53.44% MCC: 0.1743\n",
      "AUPRC: 59.81% AUROC: 63.08%\n",
      " accuracy [58.971428571428575]\n",
      " precision [58.35694050991501]\n",
      " recall [49.282296650717704]\n",
      " F1 [53.43709468223087]\n",
      " MCC [0.17427126336383483]\n",
      " AUPRC [59.80744637911614]\n",
      " AUROC [63.083035817113895]\n",
      "\n",
      "Training # 1 3\n",
      "2023-04-17 18:38:25.744544\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.1.3.validation_genes.txt\n",
      "2023-04-17 18:38:25.758331\n",
      "Discriminator ZERO -1\n",
      "Train 1:0 1671:1826 47.78%\n",
      "Valid 1:0 390:484 44.62%\n",
      "FIT\n",
      "2023-04-17 18:38:25.792910\n",
      "\n",
      "Testing # 1 3\n",
      "2023-04-17 18:38:31.299862\n",
      "2023-04-17 18:38:31.299890\n",
      "PREDICT\n",
      "2023-04-17 18:38:31.304986\n",
      "debug pred [0.52, 0.52, 0.39]\n",
      "debug class [1 1 0]\n",
      "Test 1:0 390:484 44.62%\n",
      "Predict 1:0 320:554 36.61%\n",
      "Test sizes (874, 1025) (874,)\n",
      "Distrib of scores: 0.4771395881006865 mean 0.0932201018029411 std\n",
      "Range of scores: 0.04 to 0.98\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[361 123]\n",
      " [193 197]]\n",
      "Normalized matrix\n",
      " [[0.41304348 0.14073227]\n",
      " [0.2208238  0.22540046]]\n",
      "Extreme scores correct, pos:neg 19 59\n",
      "Extreme scores incorrect pos:neg 9 16\n",
      "Accuracy: 63.84% Precision: 61.56% Recall: 50.51%\n",
      "F1: 55.49% MCC: 0.2590\n",
      "AUPRC: 59.10% AUROC: 66.17%\n",
      " accuracy [63.84439359267735]\n",
      " precision [61.5625]\n",
      " recall [50.51282051282051]\n",
      " F1 [55.492957746478865]\n",
      " MCC [0.2589952909073957]\n",
      " AUPRC [59.103333100273915]\n",
      " AUROC [66.16894469167195]\n",
      "\n",
      "Training # 1 4\n",
      "2023-04-17 18:38:31.398556\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.1.4.validation_genes.txt\n",
      "2023-04-17 18:38:31.414466\n",
      "Discriminator ZERO -1\n",
      "Train 1:0 1637:1860 46.81%\n",
      "Valid 1:0 424:450 48.51%\n",
      "FIT\n",
      "2023-04-17 18:38:31.448606\n",
      "\n",
      "Testing # 1 4\n",
      "2023-04-17 18:38:36.962441\n",
      "2023-04-17 18:38:36.962466\n",
      "PREDICT\n",
      "2023-04-17 18:38:36.968683\n",
      "debug pred [0.54, 0.49, 0.6]\n",
      "debug class [1 0 1]\n",
      "Test 1:0 424:450 48.51%\n",
      "Predict 1:0 296:578 33.87%\n",
      "Test sizes (874, 1025) (874,)\n",
      "Distrib of scores: 0.46473684210526317 mean 0.09624167159833329 std\n",
      "Range of scores: 0.13 to 0.94\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[329 121]\n",
      " [249 175]]\n",
      "Normalized matrix\n",
      " [[0.37643021 0.13844394]\n",
      " [0.28489703 0.20022883]]\n",
      "Extreme scores correct, pos:neg 19 73\n",
      "Extreme scores incorrect pos:neg 9 26\n",
      "Accuracy: 57.67% Precision: 59.12% Recall: 41.27%\n",
      "F1: 48.61% MCC: 0.1519\n",
      "AUPRC: 58.48% AUROC: 63.30%\n",
      " accuracy [57.66590389016019]\n",
      " precision [59.12162162162162]\n",
      " recall [41.27358490566038]\n",
      " F1 [48.61111111111111]\n",
      " MCC [0.15190780872860193]\n",
      " AUPRC [58.47660595814257]\n",
      " AUROC [63.3003144654088]\n",
      "\n",
      "Training # 1 5\n",
      "2023-04-17 18:38:37.070147\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.1.5.validation_genes.txt\n",
      "2023-04-17 18:38:37.087163\n",
      "Discriminator ZERO -1\n",
      "Train 1:0 1634:1864 46.71%\n",
      "Valid 1:0 427:446 48.91%\n",
      "FIT\n",
      "2023-04-17 18:38:37.121991\n",
      "\n",
      "Testing # 1 5\n",
      "2023-04-17 18:38:42.639665\n",
      "2023-04-17 18:38:42.639704\n",
      "PREDICT\n",
      "2023-04-17 18:38:42.644616\n",
      "debug pred [0.21, 0.48, 0.34]\n",
      "debug class [0 0 0]\n",
      "Test 1:0 427:446 48.91%\n",
      "Predict 1:0 298:575 34.14%\n",
      "Test sizes (873, 1025) (873,)\n",
      "Distrib of scores: 0.4700916380297823 mean 0.08832145202438005 std\n",
      "Range of scores: 0.19 to 0.87\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[329 117]\n",
      " [246 181]]\n",
      "Normalized matrix\n",
      " [[0.3768614  0.13402062]\n",
      " [0.28178694 0.20733104]]\n",
      "Extreme scores correct, pos:neg 15 71\n",
      "Extreme scores incorrect pos:neg 5 15\n",
      "Accuracy: 58.42% Precision: 60.74% Recall: 42.39%\n",
      "F1: 49.93% MCC: 0.1703\n",
      "AUPRC: 59.89% AUROC: 63.12%\n",
      " accuracy [58.4192439862543]\n",
      " precision [60.738255033557046]\n",
      " recall [42.38875878220141]\n",
      " F1 [49.93103448275861]\n",
      " MCC [0.17031846340058907]\n",
      " AUPRC [59.88855853592254]\n",
      " AUROC [63.12420579494019]\n",
      "\n",
      "Training # 2 1\n",
      "2023-04-17 18:38:42.735439\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.2.1.validation_genes.txt\n",
      "2023-04-17 18:38:42.748278\n",
      "Discriminator ZERO -1\n",
      "Train 1:0 1655:1841 47.34%\n",
      "Valid 1:0 406:469 46.40%\n",
      "FIT\n",
      "2023-04-17 18:38:42.783406\n",
      "\n",
      "Testing # 2 1\n",
      "2023-04-17 18:38:48.251934\n",
      "2023-04-17 18:38:48.251962\n",
      "PREDICT\n",
      "2023-04-17 18:38:48.257105\n",
      "debug pred [0.5, 0.52, 0.32]\n",
      "debug class [0 1 0]\n",
      "Test 1:0 406:469 46.40%\n",
      "Predict 1:0 305:570 34.86%\n",
      "Test sizes (875, 1025) (875,)\n",
      "Distrib of scores: 0.46780571428571427 mean 0.09053736069508904 std\n",
      "Range of scores: 0.1 to 0.75\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[337 132]\n",
      " [233 173]]\n",
      "Normalized matrix\n",
      " [[0.38514286 0.15085714]\n",
      " [0.26628571 0.19771429]]\n",
      "Extreme scores correct, pos:neg 8 68\n",
      "Extreme scores incorrect pos:neg 7 25\n",
      "Accuracy: 58.29% Precision: 56.72% Recall: 42.61%\n",
      "F1: 48.66% MCC: 0.1514\n",
      "AUPRC: 56.34% AUROC: 61.96%\n",
      " accuracy [58.285714285714285]\n",
      " precision [56.72131147540984]\n",
      " recall [42.610837438423644]\n",
      " F1 [48.66385372714487]\n",
      " MCC [0.15139306833142432]\n",
      " AUPRC [56.342673347659314]\n",
      " AUROC [61.96209312340479]\n",
      "\n",
      "Training # 2 2\n",
      "2023-04-17 18:38:48.349300\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.2.2.validation_genes.txt\n",
      "2023-04-17 18:38:48.362956\n",
      "Discriminator ZERO -1\n",
      "Train 1:0 1645:1851 47.05%\n",
      "Valid 1:0 416:459 47.54%\n",
      "FIT\n",
      "2023-04-17 18:38:48.397630\n",
      "\n",
      "Testing # 2 2\n",
      "2023-04-17 18:38:53.996436\n",
      "2023-04-17 18:38:53.996466\n",
      "PREDICT\n",
      "2023-04-17 18:38:54.002088\n",
      "debug pred [0.52, 0.53, 0.32]\n",
      "debug class [1 1 0]\n",
      "Test 1:0 416:459 47.54%\n",
      "Predict 1:0 347:528 39.66%\n",
      "Test sizes (875, 1025) (875,)\n",
      "Distrib of scores: 0.4777371428571429 mean 0.0981073437930232 std\n",
      "Range of scores: 0.08 to 0.87\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[321 138]\n",
      " [207 209]]\n",
      "Normalized matrix\n",
      " [[0.36685714 0.15771429]\n",
      " [0.23657143 0.23885714]]\n",
      "Extreme scores correct, pos:neg 18 67\n",
      "Extreme scores incorrect pos:neg 12 14\n",
      "Accuracy: 60.57% Precision: 60.23% Recall: 50.24%\n",
      "F1: 54.78% MCC: 0.2060\n",
      "AUPRC: 59.72% AUROC: 65.41%\n",
      " accuracy [60.57142857142858]\n",
      " precision [60.23054755043228]\n",
      " recall [50.24038461538461]\n",
      " F1 [54.78374836173001]\n",
      " MCC [0.20596119072330915]\n",
      " AUPRC [59.71660192733308]\n",
      " AUROC [65.41394335511983]\n",
      "\n",
      "Training # 2 3\n",
      "2023-04-17 18:38:54.094730\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.2.3.validation_genes.txt\n",
      "2023-04-17 18:38:54.109969\n",
      "Discriminator ZERO -1\n",
      "Train 1:0 1634:1863 46.73%\n",
      "Valid 1:0 427:447 48.86%\n",
      "FIT\n",
      "2023-04-17 18:38:54.146166\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Testing # 2 3\n",
      "2023-04-17 18:39:00.208047\n",
      "2023-04-17 18:39:00.208076\n",
      "PREDICT\n",
      "2023-04-17 18:39:00.213261\n",
      "debug pred [0.46, 0.55, 0.63]\n",
      "debug class [0 1 1]\n",
      "Test 1:0 427:447 48.86%\n",
      "Predict 1:0 325:549 37.19%\n",
      "Test sizes (874, 1025) (874,)\n",
      "Distrib of scores: 0.4750114416475973 mean 0.09408751151625869 std\n",
      "Range of scores: 0.17 to 0.91\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[318 129]\n",
      " [231 196]]\n",
      "Normalized matrix\n",
      " [[0.36384439 0.14759725]\n",
      " [0.26430206 0.22425629]]\n",
      "Extreme scores correct, pos:neg 22 69\n",
      "Extreme scores incorrect pos:neg 8 17\n",
      "Accuracy: 58.81% Precision: 60.31% Recall: 45.90%\n",
      "F1: 52.13% MCC: 0.1763\n",
      "AUPRC: 61.75% AUROC: 63.42%\n",
      " accuracy [58.81006864988558]\n",
      " precision [60.30769230769231]\n",
      " recall [45.90163934426229]\n",
      " F1 [52.127659574468076]\n",
      " MCC [0.17626868341781746]\n",
      " AUPRC [61.74953753544479]\n",
      " AUROC [63.42386663104014]\n",
      "\n",
      "Training # 2 4\n",
      "2023-04-17 18:39:00.307959\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.2.4.validation_genes.txt\n",
      "2023-04-17 18:39:00.323518\n",
      "Discriminator ZERO -1\n",
      "Train 1:0 1669:1829 47.71%\n",
      "Valid 1:0 392:481 44.90%\n",
      "FIT\n",
      "2023-04-17 18:39:00.358242\n",
      "\n",
      "Testing # 2 4\n",
      "2023-04-17 18:39:06.203055\n",
      "2023-04-17 18:39:06.203082\n",
      "PREDICT\n",
      "2023-04-17 18:39:06.208793\n",
      "debug pred [0.38, 0.55, 0.51]\n",
      "debug class [0 1 1]\n",
      "Test 1:0 392:481 44.90%\n",
      "Predict 1:0 323:550 37.00%\n",
      "Test sizes (873, 1025) (873,)\n",
      "Distrib of scores: 0.474581901489118 mean 0.09788293025044828 std\n",
      "Range of scores: 0.15 to 0.93\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[334 147]\n",
      " [216 176]]\n",
      "Normalized matrix\n",
      " [[0.38258877 0.16838488]\n",
      " [0.24742268 0.20160367]]\n",
      "Extreme scores correct, pos:neg 18 73\n",
      "Extreme scores incorrect pos:neg 8 31\n",
      "Accuracy: 58.42% Precision: 54.49% Recall: 44.90%\n",
      "F1: 49.23% MCC: 0.1477\n",
      "AUPRC: 55.38% AUROC: 61.10%\n",
      " accuracy [58.4192439862543]\n",
      " precision [54.4891640866873]\n",
      " recall [44.89795918367347]\n",
      " F1 [49.230769230769226]\n",
      " MCC [0.14769983194204786]\n",
      " AUPRC [55.3848193121017]\n",
      " AUROC [61.10224235224236]\n",
      "\n",
      "Training # 2 5\n",
      "2023-04-17 18:39:06.307207\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.2.5.validation_genes.txt\n",
      "2023-04-17 18:39:06.321248\n",
      "Discriminator ZERO -1\n",
      "Train 1:0 1641:1856 46.93%\n",
      "Valid 1:0 420:454 48.05%\n",
      "FIT\n",
      "2023-04-17 18:39:06.357829\n",
      "\n",
      "Testing # 2 5\n",
      "2023-04-17 18:39:12.230290\n",
      "2023-04-17 18:39:12.230318\n",
      "PREDICT\n",
      "2023-04-17 18:39:12.235441\n",
      "debug pred [0.5, 0.6, 0.58]\n",
      "debug class [0 1 1]\n",
      "Test 1:0 420:454 48.05%\n",
      "Predict 1:0 312:562 35.70%\n",
      "Test sizes (874, 1025) (874,)\n",
      "Distrib of scores: 0.4755148741418764 mean 0.09420884771080629 std\n",
      "Range of scores: 0.03 to 0.94\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[333 121]\n",
      " [229 191]]\n",
      "Normalized matrix\n",
      " [[0.38100686 0.13844394]\n",
      " [0.26201373 0.21853547]]\n",
      "Extreme scores correct, pos:neg 22 55\n",
      "Extreme scores incorrect pos:neg 3 23\n",
      "Accuracy: 59.95% Precision: 61.22% Recall: 45.48%\n",
      "F1: 52.19% MCC: 0.1963\n",
      "AUPRC: 63.48% AUROC: 65.67%\n",
      " accuracy [59.95423340961098]\n",
      " precision [61.21794871794872]\n",
      " recall [45.476190476190474]\n",
      " F1 [52.18579234972678]\n",
      " MCC [0.19630159666711008]\n",
      " AUPRC [63.484743542113996]\n",
      " AUROC [65.67311726452697]\n",
      "2023-04-17 18:39:12.333821\n"
     ]
    }
   ],
   "source": [
    "print(datetime.now())\n",
    "\n",
    "accuracy=[]\n",
    "precision=[]\n",
    "recall=[]\n",
    "f1=[]\n",
    "mcc=[]\n",
    "auprc=[]\n",
    "auroc=[]\n",
    "\n",
    "loader = DataLoader()\n",
    "filepath = DATA_DIR+RCI_FILE_TRAIN\n",
    "print(\"Load RCI from\",filepath)\n",
    "gene_to_rci = loader.load_gene_rci_values(filepath,EXCLUSIONS)\n",
    "print('Num RCI:', len(gene_to_rci.keys()))\n",
    "filepath = DATA_DIR+SEQ_FILE_TRAIN\n",
    "print('Load sequence from',filepath)\n",
    "allids,allseq = loader.load_spectra(filepath)  # for MLP (load_sequence() for CNN)\n",
    "print('Loaded',len(allseq),'sequences.')\n",
    "test_gene_to_rci = None\n",
    "test_allids = None\n",
    "test_allseq = None\n",
    "if SEQ_FILE_TEST is not None:\n",
    "    # Train on the entire train set (no cross-validation).\n",
    "    # Evaluate with the test files.\n",
    "    test_loader = DataLoader()\n",
    "    filepath = DATA_DIR+RCI_FILE_TEST\n",
    "    print(\"Load RCI from\",filepath)\n",
    "    test_gene_to_rci = test_loader.load_gene_rci_values(filepath,EXCLUSIONS)\n",
    "    print('Num RCI:', len(test_gene_to_rci.keys()))\n",
    "    filepath = DATA_DIR+SEQ_FILE_TEST\n",
    "    print('Load sequence from',filepath)\n",
    "    test_allids,test_allseq = test_loader.load_spectra(filepath)\n",
    "    print('Loaded',len(test_allseq),'sequences.')\n",
    "\n",
    "for repeat in range(REPEATS):\n",
    "    for fold in range(FOLDS):\n",
    "        show_r = repeat+1  # display one-based counting\n",
    "        show_f = fold+1    # display one-based counting\n",
    "\n",
    "        print()\n",
    "        print(\"Training #\",show_r,show_f)\n",
    "        print(datetime.now())\n",
    "        cvdo = CrossValidator(EPOCHS)\n",
    "        cvdo.set_threshold_mechanism(RCI_THRESHOLD_MECHANISM)\n",
    "        if SEQ_FILE_TEST is None:\n",
    "            # Train on 80% and evaluate on 20%.\n",
    "            separator = Separator()\n",
    "            separator.load(DATA_DIR,show_r,show_f)\n",
    "            separator.process(allids,allseq,gene_to_rci)\n",
    "            train_allids,test_allids = separator.get_ids()\n",
    "            train_allseq,test_allseq = separator.get_seq()\n",
    "            train_gene_to_rci,test_gene_to_rci = separator.get_rci()\n",
    "            cvdo.train_new_model(\n",
    "                train_allids,train_allseq,train_gene_to_rci,\n",
    "                test_allids,test_allseq,test_gene_to_rci)\n",
    "            if SAVE_MODEL_FILENAME is not None:\n",
    "                filename = f\"{SAVE_MODEL_FILENAME}.{show_r}.{show_f}.model\"\n",
    "                cvdo.save_model(filename)\n",
    "        else:\n",
    "            # Train on the entire train set (no cross-validation).\n",
    "            # Evaluate with the test files.\n",
    "            train_allids = allids\n",
    "            train_allseq = allseq\n",
    "            train_gene_to_rci = gene_to_rci\n",
    "            BREAK = True\n",
    "            cvdo.train_new_model(\n",
    "                train_allids,train_allseq,train_gene_to_rci,\n",
    "                None,None,None)\n",
    "\n",
    "        print()\n",
    "        print(\"Testing #\",show_r,show_f)\n",
    "        print(datetime.now())\n",
    "        cvdo.reset_statistics()\n",
    "        cvdo.test_without_training(\n",
    "            test_allids,test_allseq,test_gene_to_rci)\n",
    "        cv_accuracy,cv_precision,cv_recall,cv_f1,cv_mcc,cv_auprc,cv_auroc=\\\n",
    "            cvdo.get_statistics()\n",
    "\n",
    "        print(\" accuracy\" ,  cv_accuracy)\n",
    "        print(\" precision\" , cv_precision)\n",
    "        print(\" recall\" ,    cv_recall)\n",
    "        print(\" F1\" ,        cv_f1)\n",
    "        print(\" MCC\" ,       cv_mcc)\n",
    "        print(\" AUPRC\" ,     cv_auprc)\n",
    "        print(\" AUROC\" ,     cv_auroc)\n",
    "\n",
    "        accuracy.append(cv_accuracy)\n",
    "        precision.append(cv_precision)\n",
    "        recall.append(cv_recall)\n",
    "        f1.append(cv_f1)\n",
    "        mcc.append(cv_mcc)\n",
    "        auprc.append(cv_auprc)\n",
    "        auroc.append(cv_auroc)\n",
    "        if BREAK: break\n",
    "    if BREAK: break\n",
    "print(datetime.now())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "HkCeDg_HdQ36",
    "outputId": "b8486985-4447-4c8b-dde8-bbda95f9cc6d"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " accuracy mean 59.36% std 1.79\n",
      " precision mean 58.84% std 2.47\n",
      " recall mean 46.18% std 3.46\n",
      " F1 mean 51.67% std 2.49\n",
      " MCC mean 0.1794 std 0.0339\n",
      " AUPRC mean 59.24% std 2.35\n",
      " AUROC mean 63.65% std 1.63\n",
      " accuracy [[58.62857142857143], [58.971428571428575], [63.84439359267735], [57.66590389016019], [58.4192439862543], [58.285714285714285], [60.57142857142858], [58.81006864988558], [58.4192439862543], [59.95423340961098]]\n",
      " precision [[55.61797752808989], [58.35694050991501], [61.5625], [59.12162162162162], [60.738255033557046], [56.72131147540984], [60.23054755043228], [60.30769230769231], [54.4891640866873], [61.21794871794872]]\n",
      " recall [[49.25373134328358], [49.282296650717704], [50.51282051282051], [41.27358490566038], [42.38875878220141], [42.610837438423644], [50.24038461538461], [45.90163934426229], [44.89795918367347], [45.476190476190474]]\n",
      " F1 [[52.242744063324544], [53.43709468223087], [55.492957746478865], [48.61111111111111], [49.93103448275861], [48.66385372714487], [54.78374836173001], [52.127659574468076], [49.230769230769226], [52.18579234972678]]\n",
      " MCC [[0.1607911545791673], [0.17427126336383483], [0.2589952909073957], [0.15190780872860193], [0.17031846340058907], [0.15139306833142432], [0.20596119072330915], [0.17626868341781746], [0.14769983194204786], [0.19630159666711008]]\n",
      " AUPRC [[58.48845142284694], [59.80744637911614], [59.103333100273915], [58.47660595814257], [59.88855853592254], [56.342673347659314], [59.71660192733308], [61.74953753544479], [55.3848193121017], [63.484743542113996]]\n",
      " AUROC [[63.21458247872687], [63.083035817113895], [66.16894469167195], [63.3003144654088], [63.12420579494019], [61.96209312340479], [65.41394335511983], [63.42386663104014], [61.10224235224236], [65.67311726452697]]\n"
     ]
    }
   ],
   "source": [
    "def STD (values):\n",
    "    # ddof=1 reduces bias when extrapolating from sample to population\n",
    "    return np.std(values,ddof=1)\n",
    "\n",
    "print(\" accuracy mean %.2f%% std %.2f\" %  (np.mean(accuracy),  STD(accuracy)))\n",
    "print(\" precision mean %.2f%% std %.2f\" % (np.mean(precision), STD(precision)))\n",
    "print(\" recall mean %.2f%% std %.2f\" %    (np.mean(recall),    STD(recall)))\n",
    "print(\" F1 mean %.2f%% std %.2f\" %        (np.mean(f1),        STD(f1)))\n",
    "print(\" MCC mean %.4f std %.4f\" %       (np.mean(mcc),       STD(mcc)))\n",
    "print(\" AUPRC mean %.2f%% std %.2f\" %     (np.mean(auprc),     STD(auprc)))\n",
    "print(\" AUROC mean %.2f%% std %.2f\" %     (np.mean(auroc),     STD(auroc)))\n",
    "\n",
    "print(\" accuracy\"  , accuracy)\n",
    "print(\" precision\" , precision)\n",
    "print(\" recall\"    , recall)\n",
    "print(\" F1\"        , f1)\n",
    "print(\" MCC\"       , mcc)\n",
    "print(\" AUPRC\"     , auprc)\n",
    "print(\" AUROC\"     , auroc)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "QjSVa72v4IsA"
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "gpuClass": "standard",
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
