{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "PG-tGRnlFLA3"
   },
   "source": [
    "# Random Forest\n",
    "canonical lncRNA, -1 threshold, cross-valiation, middle exclusion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "0RmwUsVLFLA6",
    "outputId": "64cc44e8-c4cd-4eee-f276-47d1983f985c"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2023-04-17 18:41:29.408835\n"
     ]
    }
   ],
   "source": [
    "from datetime import datetime\n",
    "print(datetime.now())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "OlzN9OdsFWEU",
    "outputId": "20173346-1246-44f4-eaf7-2a5f2c032df7"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-04-17 18:41:29.484869: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GPU device not found\n",
      "/Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-04-17 18:41:40.055776: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\n"
     ]
    }
   ],
   "source": [
    "import tensorflow as tf\n",
    "dt='float32'\n",
    "tf.keras.backend.set_floatx('float32')\n",
    "# tf.random.set_seed(42) # supposedly leads to reproducible results\n",
    "\n",
    "device_name = tf.test.gpu_device_name()\n",
    "if device_name != '/device:GPU:0':\n",
    "    print('GPU device not found')\n",
    "else:\n",
    "    print('Found GPU at: {}'.format(device_name))\n",
    "\n",
    "try:\n",
    "    from google.colab import drive\n",
    "    IN_COLAB = True\n",
    "    print('Running on CoLab')\n",
    "    PATH='/content/drive/'\n",
    "    drive.mount(PATH)\n",
    "    DATA_DIR=PATH+'My Drive/data/Localization/TrainTest/TrainTest_ver43/'  # must end in \"/\"\n",
    "    MODEL_DIR=PATH+'My Drive/data/Localization/Models/'  # must end in \"/\"\n",
    "except:\n",
    "    IN_COLAB = False\n",
    "    DATA_DIR = 'D:/Adjeroh/Localization/TrainTest/'   # Windows\n",
    "    DATA_DIR = '/Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/'    # Mac\n",
    "    MODEL_DIR = '/Users/jasonmiller/WVU/Localization/Models/'    # Mac\n",
    "print(DATA_DIR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "PRX-UEr8FLA8",
    "outputId": "39890dac-1ca8-4ba2-9229-2b581f26b315"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Python 3.10.0\n",
      "sklearn 1.1.2\n"
     ]
    }
   ],
   "source": [
    "from platform import python_version\n",
    "print('Python',python_version())\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy.stats as ss\n",
    "import pickle\n",
    "import time # sleep function\n",
    "from os.path import isfile\n",
    "from matplotlib import pyplot as plt \n",
    "import sklearn   # pip install --upgrade scikit-learn\n",
    "print('sklearn',sklearn.__version__)\n",
    "\n",
    "from tensorflow import keras\n",
    "from sklearn.mixture import GaussianMixture\n",
    "from sklearn.model_selection import KFold\n",
    "from sklearn.metrics import accuracy_score\n",
    "from sklearn.metrics import precision_score\n",
    "from sklearn.metrics import recall_score\n",
    "from sklearn.metrics import f1_score\n",
    "from sklearn.metrics import precision_recall_curve\n",
    "from sklearn.metrics import roc_curve\n",
    "from sklearn.metrics import roc_auc_score\n",
    "from sklearn.metrics import auc\n",
    "from sklearn.metrics import matthews_corrcoef\n",
    "from sklearn.utils import class_weight\n",
    "from sklearn.metrics import confusion_matrix\n",
    "from sklearn.ensemble import RandomForestClassifier as RFC\n",
    "\n",
    "from keras.models import Sequential\n",
    "from keras.layers import Masking\n",
    "from keras.layers import Embedding\n",
    "from keras.layers import Conv1D\n",
    "from keras.layers import GlobalAveragePooling1D\n",
    "from keras.layers import GlobalMaxPooling1D\n",
    "from keras.layers import AveragePooling1D\n",
    "from keras.layers import MaxPooling1D\n",
    "from keras.layers import Bidirectional\n",
    "from keras.layers import LSTM\n",
    "from keras.layers import TimeDistributed\n",
    "from keras.layers import Flatten\n",
    "from keras.layers import Dense\n",
    "from keras.layers import Dropout\n",
    "from keras.losses import BinaryCrossentropy\n",
    "#from keras.losses import Hinge\n",
    "\n",
    "K=5\n",
    "ALPHABET=4**K + 1\n",
    "CFILTERS=64 \n",
    "FILTERSIZE=8\n",
    "RCELLS=32\n",
    "DCELLS=16\n",
    "EPOCHS=150 \n",
    "EMBED_DIMEN = 4 # arbitrary hyperparameter\n",
    "# MINLEN=200   # NO LONGER USED\n",
    "# MAXLEN=5000\n",
    "RCI_THRESHOLD_MECHANISM = 'ZERO'  # 'RCI_GMM' 'ZERO' 'THE_MEAN'\n",
    "BREAK = False   # break after first fold\n",
    "EXCLUSIONS = [1]   # possibly exclude cell line 1 = H1.hESC\n",
    "FILTER_TRAIN        = True\n",
    "FILTER_TAILS_TRAIN  = False\n",
    "FILTER_MIDDLE_TRAIN = True\n",
    "FILTER_TEST         = True\n",
    "FILTER_TAILS_TEST   = False\n",
    "FILTER_MIDDLE_TEST  = True\n",
    "REPEATS = 2\n",
    "FOLDS = 5\n",
    "\n",
    "SAVE_MODEL_FILENAME = None # 'RF_102'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "id": "LnkpVKdMFLA-"
   },
   "outputs": [],
   "source": [
    "RCI_FILE_TRAIN = 'train.lncRNA_RCI.csv'\n",
    "RCI_FILE_TEST  = None #'test.lncRNA_RCI.csv'\n",
    "\n",
    "SEQ_FILE_TRAIN = 'train.canon_lncRNA_transcripts.csv'\n",
    "SEQ_FILE_TEST  = None #'test.canon_lncRNA_transcripts.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "id": "e3p4QzQJFLA_"
   },
   "outputs": [],
   "source": [
    "def get_ordered_list():\n",
    "    ordered_list = \\\n",
    "    ['A549','H1.hESC','HeLa.S3','HepG2','HT1080',\\\n",
    "      'HUVEC','MCF.7','NCI.H460','NHEK','SK.MEL.5',\\\n",
    "      'SK.N.DZ','SK.N.SH','GM12878','K562','IMR.90']\n",
    "    return ordered_list\n",
    "all_cell_lines = get_ordered_list()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "UtqdpJOxFLBA"
   },
   "source": [
    "## Data Load"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "id": "p35ehKV3Kq0z"
   },
   "outputs": [],
   "source": [
    "class DataLoader():\n",
    "    def __init__(self):\n",
    "        self.cache=dict() \n",
    "        self.vals = {'A':0, 'C':1, 'G':2, 'T':3}\n",
    "        self.gene2rci = dict()\n",
    "        \n",
    "    def load_gene_rci_values(self,filepath,exclusions):\n",
    "        '''\n",
    "        Load all the genes from the given RCI csv file.\n",
    "        The given file usually contains train or test, not both.\n",
    "        Return dict with keys=gene:str and values=RCI:float.\n",
    "        '''\n",
    "        self.gene2rci = {}\n",
    "        overall_sum = 0\n",
    "        with open (filepath,'r') as handle:\n",
    "            header = None\n",
    "            for row in handle:\n",
    "                if header is None:\n",
    "                    header = row # skip file's header line\n",
    "                else:\n",
    "                    line = row.strip()\n",
    "                    fields = line.split(',')\n",
    "                    gene_id = fields.pop(0)\n",
    "                    log_mean = float(fields.pop(0))\n",
    "                    self.gene2rci[gene_id] = log_mean\n",
    "        print('Number of RCI values loaded',len(self.gene2rci.keys()))\n",
    "        return self.gene2rci\n",
    "\n",
    "    def _seq_to_kmer_values(self,rna,K):\n",
    "        vec=[] # seq converted to list of K-mers \n",
    "        N_indicator = 0 # indicator value\n",
    "        length = len(rna)\n",
    "        for i in range(length-K+1):\n",
    "            kmer = rna[i:i+K]\n",
    "            if 'N' in kmer:\n",
    "                value = N_indicator\n",
    "            elif kmer in self.cache.keys():\n",
    "                value = self.cache[kmer]\n",
    "            else:\n",
    "                value = 0\n",
    "                for j in range(K):\n",
    "                    value *= 4   \n",
    "                    nextnuc = kmer[j] \n",
    "                    nucval = self.vals[nextnuc]\n",
    "                    value += nucval\n",
    "                value += 1   # NNN => 0, AAA => 1\n",
    "                self.cache[kmer] = value\n",
    "            vec.append(value)\n",
    "        return vec\n",
    "\n",
    "    def load_sequence(self,filepath):\n",
    "        '''\n",
    "        Load all the sequences from the given file. \n",
    "        Load our version of GenCode -- csv, easier to parse than fasta.\n",
    "        Each line has IDs plus sequence.\n",
    "        The IDs in the file do not include dot-version numbers.\n",
    "        The file may already be filtered e.g. canonical transcripts.\n",
    "        '''\n",
    "        allids=[]\n",
    "        allseq=[]\n",
    "        #NREPEAT = str('N'*MAXLEN)   # not used for MLP\n",
    "        with open (filepath,'r') as handle:\n",
    "            header = None\n",
    "            for row in handle:\n",
    "                if header is None:\n",
    "                    header = row\n",
    "                else:\n",
    "                    line    = row.strip()\n",
    "                    fields  = line.split(',')\n",
    "                    tran_id = fields[0]  # without version number\n",
    "                    gene_id = fields[1]  # without version number\n",
    "                    seq_len = int(fields[3])\n",
    "                    seq_txt = fields[4]\n",
    "                    # Keep only transcripts having numeric RCI given the cell lines in use.\n",
    "                    # We have validated this by spot checking.\n",
    "                    # TO DO: validate this programmatically.\n",
    "                    if gene_id in self.gene2rci.keys():\n",
    "                        # no MAXLEN for MLP\n",
    "                        #if seq_len<=MAXLEN:\n",
    "                        #    seq_txt = seq_txt + NREPEAT\n",
    "                        #seq_txt = seq_txt[:MAXLEN]\n",
    "                        allids.append( (gene_id,tran_id) )\n",
    "                        hot_vec = self._seq_to_kmer_values(seq_txt,K)\n",
    "                        allseq.append(hot_vec)\n",
    "        self.cache=dict() # save RAM\n",
    "        return allids,allseq\n",
    "\n",
    "    def load_spectra(self,filepath):\n",
    "        '''\n",
    "        Load all (variable-length) sequences as lists of kmers.\n",
    "        Then convert each sequence to (fixed-length) kmer histograms.\n",
    "        '''\n",
    "        allids,allseq = self.load_sequence(filepath)\n",
    "        allspectra = []\n",
    "        for seq in allseq:\n",
    "            spectrum = np.zeros(ALPHABET)\n",
    "            for kmer in seq:\n",
    "                spectrum[kmer] += 1\n",
    "            spectrum /= len(seq)\n",
    "            allspectra.append(spectrum)\n",
    "        return allids,allspectra        \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "VDZ6siB_Kq04"
   },
   "source": [
    "## Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "id": "AwMbRjm0FLBF"
   },
   "outputs": [],
   "source": [
    "def build_model():\n",
    "    rfc = RFC()\n",
    "    return rfc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "clj-wufgFLBF",
    "outputId": "a17cf7a6-3755-48aa-d96b-dd0383c6f2f2"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2023-04-17 18:41:42.878452\n",
      "RandomForestClassifier()\n"
     ]
    }
   ],
   "source": [
    "print(datetime.now())\n",
    "model=build_model()\n",
    "print(model)  # Print this only once\n",
    "model=None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "MgrC1alOKq07"
   },
   "source": [
    "## Training"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "id": "W9xiFzNbFLBE"
   },
   "outputs": [],
   "source": [
    "import gc\n",
    "class CrossValidator():\n",
    "    def __init__(self,epochs,score_threshold=0.5):\n",
    "        self.epochs = epochs\n",
    "        self.score_threshold = score_threshold\n",
    "        self.mechanism = 'ZERO'\n",
    "        self.discriminator = -1 # or zero\n",
    "        self.flip = False\n",
    "        self.reset_statistics()\n",
    "        \n",
    "    def reset_statistics(self):\n",
    "        self.cv_accuracy=[]\n",
    "        self.cv_precision=[]\n",
    "        self.cv_recall=[]\n",
    "        self.cv_f1=[]\n",
    "        self.cv_auprc=[]\n",
    "        self.cv_auroc=[]\n",
    "        self.cv_mcc=[]\n",
    "        \n",
    "    def _get_X_y(self, all_ids, all_seqs, rci_map): \n",
    "        # Prepare X and y for training or testing.\n",
    "        subsetX=[]\n",
    "        subsetY=[]\n",
    "        for t in range(len(all_ids)):\n",
    "            gene_id,tran_id = all_ids[t]\n",
    "            oneX            = all_seqs[t]\n",
    "            oneY            = rci_map[gene_id]\n",
    "            subsetX.append(oneX)\n",
    "            subsetY.append(oneY)\n",
    "        subsetX = np.array(subsetX)\n",
    "        subsetY = np.array(subsetY).reshape((-1,1))\n",
    "        return subsetX,subsetY\n",
    "    \n",
    "    def set_threshold_mechanism(self, mechanism):\n",
    "        if mechanism not in ['RCI_GMM','THE_MEAN','ZERO']:\n",
    "            raise Exception('Unrecognized mechansm:',mechanism)\n",
    "        self.mechanism = mechanism\n",
    "    \n",
    "    def _apply_threshold(self, array_of_rci):\n",
    "        # Takes list of float, returns list of labels [0,1].\n",
    "        if self.mechanism == 'RCI_GMM':\n",
    "            labels = self.discriminator.predict(array_of_rci)\n",
    "            if self.flip:\n",
    "                IS_CYTO = lambda label: 1 if label==0 else 0\n",
    "                labels = np.array(list(map(IS_CYTO, labels)))\n",
    "        else:  # 'THE_MEAN' or 'ZERO'\n",
    "            rci_threshold = self.discriminator\n",
    "            IS_CYTO = lambda rci: 1 if rci>rci_threshold else 0\n",
    "            labels = np.array(list(map(IS_CYTO, array_of_rci)))\n",
    "        return labels\n",
    "    \n",
    "    def _prepare_threshold(self, rci_values, create=True):\n",
    "        if self.mechanism == 'RCI_GMM':\n",
    "            if create:  # during training, create a new GMM\n",
    "                gmm = GaussianMixture(n_components=2, verbose=0, \n",
    "                  covariance_type='spherical', n_init=100) # random_state=42) \n",
    "                gmm.fit(rci_values)\n",
    "            else:   # during testing, use existing GMM\n",
    "                gmm=self.discriminator\n",
    "            self.flip = False\n",
    "            # The GMM labels are arbitrary.\n",
    "            if gmm.means_[0][0] > gmm.means_[1][0]:\n",
    "                self.flip = True\n",
    "            self.discriminator = gmm   # redundant but consistent\n",
    "        elif self.mechanism == 'THE_MEAN':\n",
    "            self.discriminator = np.mean(rci_values)\n",
    "        elif self.mechanism == 'ZERO':\n",
    "            self.discriminator = -1   # 0 usually, -1 is as in Yuan et al.\n",
    "        else: # not expected\n",
    "            self.discriminator = 0\n",
    "    \n",
    "    def _explain_threshold(self):\n",
    "        if self.mechanism == 'RCI_GMM':\n",
    "            gmm=self.discriminator\n",
    "            print('Discriminator is GMM')\n",
    "            print('Means',[gmm.means_[0][0],gmm.means_[1][0]])\n",
    "            print('Variances',gmm.covariances_)\n",
    "            print('Priors',gmm.weights_)\n",
    "            test_rcis=[-5,-4,-3.5,-3,-2.5,-2,-1.5,-1,-0.5,0,0.5,1,1.5,2,3]\n",
    "            print(test_rcis)\n",
    "            print(self._apply_threshold(np.array(test_rcis).reshape((-1,1))))\n",
    "        else:\n",
    "            print('Discriminator',self.mechanism,self.discriminator)\n",
    "    \n",
    "    def _show_sizes(self,label,values):\n",
    "        a = np.count_nonzero(values==1)\n",
    "        b = np.count_nonzero(values==0)\n",
    "        print('%s 1:0 %d:%d %5.2f%%'%(label,a,b,100*a/(a+b)))\n",
    "        \n",
    "    def save_model(self,filename):\n",
    "        if self.model is not None:\n",
    "            filepath = MODEL_DIR + filename\n",
    "            #self.model.save(filepath)\n",
    "            print('? Saved model to',filepath)\n",
    "        \n",
    "    def load_model(self,filename):\n",
    "        filepath = MODEL_DIR + filename\n",
    "        #self.model = keras.models.load_model(filepath)\n",
    "        print('? Loaded model from',filepath)\n",
    "        \n",
    "    def train_new_model(self,train_ids,train_seq,train_rci,\n",
    "            valid_ids=None,valid_seq=None,valid_rci=None):\n",
    "        print(datetime.now())\n",
    "        X_train,y_rci = self._get_X_y(train_ids,train_seq,train_rci) \n",
    "        self._prepare_threshold(y_rci,True)  \n",
    "        self._explain_threshold()\n",
    "        y_train = self._apply_threshold(y_rci)\n",
    "        self._show_sizes('Train',y_train)\n",
    "        #cw = class_weight.compute_class_weight('balanced', classes=[0,1], y=y_train)\n",
    "        #cwd = {0: cw[0], 1: cw[1]}\n",
    "        #print('Computed class weights:',cwd)\n",
    "        if valid_ids is not None:\n",
    "            X_valid,y_rci = self._get_X_y(valid_ids,valid_seq,valid_rci) \n",
    "            y_valid = self._apply_threshold(y_rci)\n",
    "            self._show_sizes('Valid',y_valid)\n",
    "        y_rci = None\n",
    "\n",
    "        self.model=build_model()\n",
    "        \n",
    "        print(\"FIT\")\n",
    "        print(datetime.now())\n",
    "        self.model.fit(X_train, y_train) # sample weight\n",
    "\n",
    "    def test_without_training(self,test_ids,test_seq,test_rci):\n",
    "        # For final test, do no train.\n",
    "        # Assume set_sequences() set the test set.\n",
    "        print(datetime.now())\n",
    "        X_test,y_rci = self._get_X_y(test_ids,test_seq,test_rci) \n",
    "        y_test = self._apply_threshold(y_rci)\n",
    "        y_rci = None\n",
    "        \n",
    "        print(\"PREDICT\")\n",
    "        print(datetime.now())        \n",
    "        yhat_pairs=self.model.predict_proba(X_test)  # [ prob of 0, prob of 1 ]\n",
    "        yhat_pred=[pair[1] for pair in yhat_pairs]\n",
    "        yhat_classes=self.model.predict(X_test)  # 0 or 1\n",
    "        \n",
    "        print('debug pred',yhat_pred[:3])\n",
    "        print('debug class',yhat_classes[:3])\n",
    "\n",
    "        self._show_sizes('Test',y_test)\n",
    "        self._show_sizes('Predict',yhat_classes)\n",
    "        print('Test sizes',X_test.shape,y_test.shape)\n",
    "        print('Distrib of scores:',np.mean(yhat_pred),'mean',np.std(yhat_pred),'std')\n",
    "        print('Range of scores:',np.min(yhat_pred),'to',np.max(yhat_pred))\n",
    "        print('Score threshold',self.score_threshold)\n",
    "        cm1 = confusion_matrix(y_test,yhat_classes)\n",
    "        print('Confusion matrix\\n',cm1)\n",
    "        cm2 = confusion_matrix(y_test,yhat_classes,normalize='all')\n",
    "        print('Normalized matrix\\n',cm2)\n",
    "\n",
    "        accuracy = accuracy_score(y_test, yhat_classes)*100.\n",
    "        precision = precision_score(y_test, yhat_classes)*100.\n",
    "        recall = recall_score(y_test, yhat_classes)*100.\n",
    "        f1 = f1_score(y_test, yhat_classes)*100.\n",
    "        prc_Y, prc_X, prc_bins = precision_recall_curve(y_test, yhat_pred)\n",
    "        auprc = auc(prc_X,prc_Y)*100.\n",
    "        auroc = roc_auc_score(y_test, yhat_pred)*100.\n",
    "        mcc = matthews_corrcoef(y_test, yhat_classes)\n",
    "\n",
    "        correct_pos = 0\n",
    "        correct_neg = 0\n",
    "        wrong_pos = 0\n",
    "        wrong_neg = 0\n",
    "        for i in range(len(y_test)):\n",
    "            if yhat_pred[i]>=0.65:\n",
    "                if y_test[i]==1:\n",
    "                    correct_pos += 1\n",
    "                else:\n",
    "                    wrong_pos += 1\n",
    "            elif yhat_pred[i]<=0.35:\n",
    "                if y_test[i]==0:\n",
    "                    correct_neg += 1\n",
    "                else:\n",
    "                    wrong_neg += 1\n",
    "        print('Extreme scores correct, pos:neg',correct_pos,correct_neg)  \n",
    "        print('Extreme scores incorrect pos:neg',wrong_pos,wrong_neg)  \n",
    "\n",
    "        #self._pretty_pictures(y_test,yhat_pred,prc_X,prc_Y)\n",
    "        \n",
    "        print('Accuracy: %.2f%% Precision: %.2f%% Recall: %.2f%%' % (accuracy,precision,recall)) \n",
    "        print('F1: %.2f%% MCC: %.4f' % (f1,mcc)) \n",
    "        print('AUPRC: %.2f%% AUROC: %.2f%%' % (auprc,auroc)) \n",
    "\n",
    "        self.cv_accuracy.append(accuracy)\n",
    "        self.cv_precision.append(precision)\n",
    "        self.cv_recall.append(recall)\n",
    "        self.cv_f1.append(f1)\n",
    "        self.cv_mcc.append(mcc)\n",
    "        self.cv_auprc.append(auprc)\n",
    "        self.cv_auroc.append(auroc)\n",
    "\n",
    "    def _pretty_pictures(self,y_valid,yhat_pred,prc_X,prc_Y):\n",
    "        count_ones= len(y_valid[y_valid==1])\n",
    "        count_zeros= len(y_valid[y_valid==0])\n",
    "        guess = max(count_ones,count_zeros) / len(y_valid)\n",
    "        # PRC\n",
    "        plt.plot(prc_X, prc_Y, marker='.')\n",
    "        plt.plot([0, 1], [guess,guess], linestyle='--')\n",
    "        plt.xlabel('Recall')\n",
    "        plt.ylabel('Precision')\n",
    "        plt.show()\n",
    "        # ROC\n",
    "        fpr, tpr, roc_bins = roc_curve(y_valid, yhat_pred)\n",
    "        plt.plot(fpr, tpr, marker='.')\n",
    "        plt.xlabel('False Positive Rate')\n",
    "        plt.ylabel('True Positive Rate')\n",
    "        plt.show()\n",
    "        \n",
    "    def get_statistics(self):\n",
    "        return \\\n",
    "        self.cv_accuracy,\\\n",
    "        self.cv_precision,\\\n",
    "        self.cv_recall,\\\n",
    "        self.cv_f1,\\\n",
    "        self.cv_mcc,\\\n",
    "        self.cv_auprc,\\\n",
    "        self.cv_auroc\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "Td97uyyj5qDq"
   },
   "source": [
    "## Train"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "id": "poe3rwAF4Ir7"
   },
   "outputs": [],
   "source": [
    "class Separator():\n",
    "    def __init__(self):\n",
    "        self.train_ids = []\n",
    "        self.train_seq = []\n",
    "        self.train_rci = dict()\n",
    "        self.val_ids = []\n",
    "        self.val_seq = []\n",
    "        self.val_rci = dict()\n",
    "    def load(self,data_dir,rep,fold):\n",
    "        filename='cv.{}.{}.validation_genes.txt'.format(rep,fold)\n",
    "        filename = data_dir + filename\n",
    "        self.val_genes = set()\n",
    "        print('Opening file',filename)\n",
    "        with open(filename,'r') as fin:\n",
    "            for line in fin:\n",
    "                gene_id = line.strip()\n",
    "                self.val_genes.add(gene_id)\n",
    "    def process(self,allids,allseq,gene_to_rci):\n",
    "        size = len(allids)\n",
    "        for t in range(size):\n",
    "            gene_id,tran_id = allids[t]\n",
    "            oneX            = allseq[t]\n",
    "            oneY            = gene_to_rci[gene_id]\n",
    "            in_middle = gene_to_rci[gene_id] >= -2 and gene_to_rci[gene_id] <= 0\n",
    "            in_tails = gene_to_rci[gene_id] < -2 or gene_to_rci[gene_id] > 0\n",
    "            if gene_id in self.val_genes:\n",
    "                if FILTER_TEST and (\\\n",
    "                    (FILTER_TAILS_TEST and in_tails) or \\\n",
    "                    (FILTER_MIDDLE_TEST and in_middle)):\n",
    "                    pass\n",
    "                else:\n",
    "                    self.val_ids.append(allids[t])\n",
    "                    self.val_seq.append(allseq[t])\n",
    "                    self.val_rci[gene_id]=gene_to_rci[gene_id]\n",
    "            else:\n",
    "                if FILTER_TRAIN and (\\\n",
    "                    (FILTER_TAILS_TRAIN and in_tails) or \\\n",
    "                    (FILTER_MIDDLE_TRAIN and in_middle)):\n",
    "                    pass\n",
    "                else:\n",
    "                    self.train_ids.append(allids[t])\n",
    "                    self.train_seq.append(allseq[t])\n",
    "                    self.train_rci[gene_id]=gene_to_rci[gene_id]\n",
    "    def get_ids(self):\n",
    "        return self.train_ids,self.val_ids\n",
    "    def get_seq(self):\n",
    "        return self.train_seq,self.val_seq\n",
    "    def get_rci(self):\n",
    "        return self.train_rci,self.val_rci"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 1000
    },
    "id": "XC9m0W-pFLBH",
    "outputId": "f1051adc-9a79-4671-dd2b-2d9617da1129",
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2023-04-17 18:41:42.962818\n",
      "Load RCI from /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/train.lncRNA_RCI.csv\n",
      "Number of RCI values loaded 4371\n",
      "Num RCI: 4371\n",
      "Load sequence from /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/train.canon_lncRNA_transcripts.csv\n",
      "Loaded 4371 sequences.\n",
      "\n",
      "Training # 1 1\n",
      "2023-04-17 18:41:56.895332\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.1.1.validation_genes.txt\n",
      "2023-04-17 18:41:56.909925\n",
      "Discriminator ZERO -1\n",
      "Train 1:0 939:1175 44.42%\n",
      "Valid 1:0 228:311 42.30%\n",
      "FIT\n",
      "2023-04-17 18:41:56.962902\n",
      "\n",
      "Testing # 1 1\n",
      "2023-04-17 18:42:00.942770\n",
      "2023-04-17 18:42:00.942803\n",
      "PREDICT\n",
      "2023-04-17 18:42:00.946254\n",
      "debug pred [0.33, 0.4, 0.42]\n",
      "debug class [0 0 0]\n",
      "Test 1:0 228:311 42.30%\n",
      "Predict 1:0 153:386 28.39%\n",
      "Test sizes (539, 1025) (539,)\n",
      "Distrib of scores: 0.4469202226345083 mean 0.11515970655594178 std\n",
      "Range of scores: 0.05 to 0.96\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[251  60]\n",
      " [135  93]]\n",
      "Normalized matrix\n",
      " [[0.46567718 0.11131725]\n",
      " [0.25046382 0.17254174]]\n",
      "Extreme scores correct, pos:neg 17 93\n",
      "Extreme scores incorrect pos:neg 7 23\n",
      "Accuracy: 63.82% Precision: 60.78% Recall: 40.79%\n",
      "F1: 48.82% MCC: 0.2356\n",
      "AUPRC: 60.94% AUROC: 69.00%\n",
      " accuracy [63.821892393320965]\n",
      " precision [60.78431372549019]\n",
      " recall [40.78947368421053]\n",
      " F1 [48.818897637795274]\n",
      " MCC [0.2355501927041128]\n",
      " AUPRC [60.94141683491403]\n",
      " AUROC [68.9978563772776]\n",
      "\n",
      "Training # 1 2\n",
      "2023-04-17 18:42:01.026237\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.1.2.validation_genes.txt\n",
      "2023-04-17 18:42:01.036528\n",
      "Discriminator ZERO -1\n",
      "Train 1:0 926:1181 43.95%\n",
      "Valid 1:0 241:305 44.14%\n",
      "FIT\n",
      "2023-04-17 18:42:01.050933\n",
      "\n",
      "Testing # 1 2\n",
      "2023-04-17 18:42:04.199747\n",
      "2023-04-17 18:42:04.199779\n",
      "PREDICT\n",
      "2023-04-17 18:42:04.203765\n",
      "debug pred [0.4, 0.46, 0.26]\n",
      "debug class [0 0 0]\n",
      "Test 1:0 241:305 44.14%\n",
      "Predict 1:0 165:381 30.22%\n",
      "Test sizes (546, 1025) (546,)\n",
      "Distrib of scores: 0.44410256410256416 mean 0.11437847860087907 std\n",
      "Range of scores: 0.06 to 0.76\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[244  61]\n",
      " [137 104]]\n",
      "Normalized matrix\n",
      " [[0.44688645 0.11172161]\n",
      " [0.25091575 0.19047619]]\n",
      "Extreme scores correct, pos:neg 12 93\n",
      "Extreme scores incorrect pos:neg 5 19\n",
      "Accuracy: 63.74% Precision: 63.03% Recall: 43.15%\n",
      "F1: 51.23% MCC: 0.2504\n",
      "AUPRC: 64.02% AUROC: 71.46%\n",
      " accuracy [63.73626373626373]\n",
      " precision [63.030303030303024]\n",
      " recall [43.15352697095436]\n",
      " F1 [51.23152709359607]\n",
      " MCC [0.2503635709996857]\n",
      " AUPRC [64.02230771131678]\n",
      " AUROC [71.45908441602612]\n",
      "\n",
      "Training # 1 3\n",
      "2023-04-17 18:42:04.266396\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.1.3.validation_genes.txt\n",
      "2023-04-17 18:42:04.273393\n",
      "Discriminator ZERO -1\n",
      "Train 1:0 951:1183 44.56%\n",
      "Valid 1:0 216:303 41.62%\n",
      "FIT\n",
      "2023-04-17 18:42:04.293425\n",
      "\n",
      "Testing # 1 3\n",
      "2023-04-17 18:42:07.423694\n",
      "2023-04-17 18:42:07.423721\n",
      "PREDICT\n",
      "2023-04-17 18:42:07.426717\n",
      "debug pred [0.42, 0.31, 0.37]\n",
      "debug class [0 0 0]\n",
      "Test 1:0 216:303 41.62%\n",
      "Predict 1:0 161:358 31.02%\n",
      "Test sizes (519, 1025) (519,)\n",
      "Distrib of scores: 0.45246628131021194 mean 0.10806444912008756 std\n",
      "Range of scores: 0.11 to 0.93\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[240  63]\n",
      " [118  98]]\n",
      "Normalized matrix\n",
      " [[0.46242775 0.12138728]\n",
      " [0.22736031 0.18882466]]\n",
      "Extreme scores correct, pos:neg 12 75\n",
      "Extreme scores incorrect pos:neg 7 10\n",
      "Accuracy: 65.13% Precision: 60.87% Recall: 45.37%\n",
      "F1: 51.99% MCC: 0.2619\n",
      "AUPRC: 60.10% AUROC: 71.97%\n",
      " accuracy [65.1252408477842]\n",
      " precision [60.86956521739131]\n",
      " recall [45.370370370370374]\n",
      " F1 [51.9893899204244]\n",
      " MCC [0.2619059851756257]\n",
      " AUPRC [60.101585147697214]\n",
      " AUROC [71.96552988632196]\n",
      "\n",
      "Training # 1 4\n",
      "2023-04-17 18:42:07.486318\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.1.4.validation_genes.txt\n",
      "2023-04-17 18:42:07.493515\n",
      "Discriminator ZERO -1\n",
      "Train 1:0 936:1202 43.78%\n",
      "Valid 1:0 231:284 44.85%\n",
      "FIT\n",
      "2023-04-17 18:42:07.512330\n",
      "\n",
      "Testing # 1 4\n",
      "2023-04-17 18:42:10.581076\n",
      "2023-04-17 18:42:10.581105\n",
      "PREDICT\n",
      "2023-04-17 18:42:10.584032\n",
      "debug pred [0.36, 0.47, 0.42]\n",
      "debug class [0 0 0]\n",
      "Test 1:0 231:284 44.85%\n",
      "Predict 1:0 138:377 26.80%\n",
      "Test sizes (515, 1025) (515,)\n",
      "Distrib of scores: 0.44155339805825244 mean 0.1108936101843354 std\n",
      "Range of scores: 0.12 to 0.99\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[226  58]\n",
      " [151  80]]\n",
      "Normalized matrix\n",
      " [[0.43883495 0.11262136]\n",
      " [0.29320388 0.15533981]]\n",
      "Extreme scores correct, pos:neg 13 79\n",
      "Extreme scores incorrect pos:neg 3 22\n",
      "Accuracy: 59.42% Precision: 57.97% Recall: 34.63%\n",
      "F1: 43.36% MCC: 0.1596\n",
      "AUPRC: 58.15% AUROC: 64.43%\n",
      " accuracy [59.417475728155345]\n",
      " precision [57.971014492753625]\n",
      " recall [34.63203463203463]\n",
      " F1 [43.360433604336045]\n",
      " MCC [0.15956356113707784]\n",
      " AUPRC [58.148628781856516]\n",
      " AUROC [64.42976038046461]\n",
      "\n",
      "Training # 1 5\n",
      "2023-04-17 18:42:10.641613\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.1.5.validation_genes.txt\n",
      "2023-04-17 18:42:10.650675\n",
      "Discriminator ZERO -1\n",
      "Train 1:0 916:1203 43.23%\n",
      "Valid 1:0 251:283 47.00%\n",
      "FIT\n",
      "2023-04-17 18:42:10.665312\n",
      "\n",
      "Testing # 1 5\n",
      "2023-04-17 18:42:13.775340\n",
      "2023-04-17 18:42:13.775371\n",
      "PREDICT\n",
      "2023-04-17 18:42:13.778323\n",
      "debug pred [0.23, 0.4, 0.47]\n",
      "debug class [0 0 0]\n",
      "Test 1:0 251:283 47.00%\n",
      "Predict 1:0 149:385 27.90%\n",
      "Test sizes (534, 1025) (534,)\n",
      "Distrib of scores: 0.4431460674157303 mean 0.10251623244758067 std\n",
      "Range of scores: 0.11 to 0.74\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[239  44]\n",
      " [146 105]]\n",
      "Normalized matrix\n",
      " [[0.44756554 0.082397  ]\n",
      " [0.27340824 0.19662921]]\n",
      "Extreme scores correct, pos:neg 10 91\n",
      "Extreme scores incorrect pos:neg 0 22\n",
      "Accuracy: 64.42% Precision: 70.47% Recall: 41.83%\n",
      "F1: 52.50% MCC: 0.2925\n",
      "AUPRC: 69.06% AUROC: 72.00%\n",
      " accuracy [64.41947565543072]\n",
      " precision [70.46979865771812]\n",
      " recall [41.832669322709165]\n",
      " F1 [52.5]\n",
      " MCC [0.2924920063098742]\n",
      " AUPRC [69.05552768738255]\n",
      " AUROC [72.00244956569482]\n",
      "\n",
      "Training # 2 1\n",
      "2023-04-17 18:42:13.838631\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.2.1.validation_genes.txt\n",
      "2023-04-17 18:42:13.847400\n",
      "Discriminator ZERO -1\n",
      "Train 1:0 941:1187 44.22%\n",
      "Valid 1:0 226:299 43.05%\n",
      "FIT\n",
      "2023-04-17 18:42:13.860580\n",
      "\n",
      "Testing # 2 1\n",
      "2023-04-17 18:42:16.978277\n",
      "2023-04-17 18:42:16.978303\n",
      "PREDICT\n",
      "2023-04-17 18:42:16.980834\n",
      "debug pred [0.46, 0.33, 0.41]\n",
      "debug class [0 0 0]\n",
      "Test 1:0 226:299 43.05%\n",
      "Predict 1:0 159:366 30.29%\n",
      "Test sizes (525, 1025) (525,)\n",
      "Distrib of scores: 0.44388571428571433 mean 0.1110990389813748 std\n",
      "Range of scores: 0.09 to 0.77\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[236  63]\n",
      " [130  96]]\n",
      "Normalized matrix\n",
      " [[0.44952381 0.12      ]\n",
      " [0.24761905 0.18285714]]\n",
      "Extreme scores correct, pos:neg 7 81\n",
      "Extreme scores incorrect pos:neg 2 30\n",
      "Accuracy: 63.24% Precision: 60.38% Recall: 42.48%\n",
      "F1: 49.87% MCC: 0.2307\n",
      "AUPRC: 57.60% AUROC: 65.76%\n",
      " accuracy [63.23809523809524]\n",
      " precision [60.37735849056604]\n",
      " recall [42.47787610619469]\n",
      " F1 [49.87012987012987]\n",
      " MCC [0.23068509889216196]\n",
      " AUPRC [57.59958557050786]\n",
      " AUROC [65.76419924823156]\n",
      "\n",
      "Training # 2 2\n",
      "2023-04-17 18:42:17.047617\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.2.2.validation_genes.txt\n",
      "2023-04-17 18:42:17.056809\n",
      "Discriminator ZERO -1\n",
      "Train 1:0 929:1189 43.86%\n",
      "Valid 1:0 238:297 44.49%\n",
      "FIT\n",
      "2023-04-17 18:42:17.070788\n",
      "\n",
      "Testing # 2 2\n",
      "2023-04-17 18:42:20.095802\n",
      "2023-04-17 18:42:20.095829\n",
      "PREDICT\n",
      "2023-04-17 18:42:20.098723\n",
      "debug pred [0.33, 0.3, 0.43]\n",
      "debug class [0 0 0]\n",
      "Test 1:0 238:297 44.49%\n",
      "Predict 1:0 174:361 32.52%\n",
      "Test sizes (535, 1025) (535,)\n",
      "Distrib of scores: 0.45183177570093463 mean 0.11153814032569387 std\n",
      "Range of scores: 0.12 to 0.7\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[239  58]\n",
      " [122 116]]\n",
      "Normalized matrix\n",
      " [[0.44672897 0.10841121]\n",
      " [0.22803738 0.21682243]]\n",
      "Extreme scores correct, pos:neg 14 87\n",
      "Extreme scores incorrect pos:neg 3 13\n",
      "Accuracy: 66.36% Precision: 66.67% Recall: 48.74%\n",
      "F1: 56.31% MCC: 0.3099\n",
      "AUPRC: 67.24% AUROC: 74.05%\n",
      " accuracy [66.35514018691589]\n",
      " precision [66.66666666666666]\n",
      " recall [48.739495798319325]\n",
      " F1 [56.310679611650485]\n",
      " MCC [0.309872519465028]\n",
      " AUPRC [67.24319969997583]\n",
      " AUROC [74.05073140367257]\n",
      "\n",
      "Training # 2 3\n",
      "2023-04-17 18:42:20.158806\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.2.3.validation_genes.txt\n",
      "2023-04-17 18:42:20.167855\n",
      "Discriminator ZERO -1\n",
      "Train 1:0 924:1198 43.54%\n",
      "Valid 1:0 243:288 45.76%\n",
      "FIT\n",
      "2023-04-17 18:42:20.180723\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Testing # 2 3\n",
      "2023-04-17 18:42:23.455576\n",
      "2023-04-17 18:42:23.455603\n",
      "PREDICT\n",
      "2023-04-17 18:42:23.458663\n",
      "debug pred [0.56, 0.5, 0.47]\n",
      "debug class [1 0 0]\n",
      "Test 1:0 243:288 45.76%\n",
      "Predict 1:0 154:377 29.00%\n",
      "Test sizes (531, 1025) (531,)\n",
      "Distrib of scores: 0.4414689265536723 mean 0.1138857774923786 std\n",
      "Range of scores: 0.1 to 0.96\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[235  53]\n",
      " [142 101]]\n",
      "Normalized matrix\n",
      " [[0.44256121 0.09981168]\n",
      " [0.26741996 0.19020716]]\n",
      "Extreme scores correct, pos:neg 12 95\n",
      "Extreme scores incorrect pos:neg 6 26\n",
      "Accuracy: 63.28% Precision: 65.58% Recall: 41.56%\n",
      "F1: 50.88% MCC: 0.2543\n",
      "AUPRC: 61.72% AUROC: 69.54%\n",
      " accuracy [63.2768361581921]\n",
      " precision [65.5844155844156]\n",
      " recall [41.56378600823045]\n",
      " F1 [50.88161209068009]\n",
      " MCC [0.25428796487539973]\n",
      " AUPRC [61.719021254756434]\n",
      " AUROC [69.53732281664381]\n",
      "\n",
      "Training # 2 4\n",
      "2023-04-17 18:42:23.518056\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.2.4.validation_genes.txt\n",
      "2023-04-17 18:42:23.524822\n",
      "Discriminator ZERO -1\n",
      "Train 1:0 953:1173 44.83%\n",
      "Valid 1:0 214:313 40.61%\n",
      "FIT\n",
      "2023-04-17 18:42:23.537750\n",
      "\n",
      "Testing # 2 4\n",
      "2023-04-17 18:42:26.697516\n",
      "2023-04-17 18:42:26.697543\n",
      "PREDICT\n",
      "2023-04-17 18:42:26.700964\n",
      "debug pred [0.39, 0.58, 0.56]\n",
      "debug class [0 1 1]\n",
      "Test 1:0 214:313 40.61%\n",
      "Predict 1:0 154:373 29.22%\n",
      "Test sizes (527, 1025) (527,)\n",
      "Distrib of scores: 0.4465085388994307 mean 0.11531802171395754 std\n",
      "Range of scores: 0.1 to 0.94\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[254  59]\n",
      " [119  95]]\n",
      "Normalized matrix\n",
      " [[0.48197343 0.11195446]\n",
      " [0.22580645 0.18026565]]\n",
      "Extreme scores correct, pos:neg 19 87\n",
      "Extreme scores incorrect pos:neg 8 21\n",
      "Accuracy: 66.22% Precision: 61.69% Recall: 44.39%\n",
      "F1: 51.63% MCC: 0.2758\n",
      "AUPRC: 58.69% AUROC: 67.93%\n",
      " accuracy [66.22390891840607]\n",
      " precision [61.68831168831169]\n",
      " recall [44.39252336448598]\n",
      " F1 [51.63043478260869]\n",
      " MCC [0.27582320117194503]\n",
      " AUPRC [58.68620519570209]\n",
      " AUROC [67.92944373115164]\n",
      "\n",
      "Training # 2 5\n",
      "2023-04-17 18:42:26.761480\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.2.5.validation_genes.txt\n",
      "2023-04-17 18:42:26.769377\n",
      "Discriminator ZERO -1\n",
      "Train 1:0 921:1197 43.48%\n",
      "Valid 1:0 246:289 45.98%\n",
      "FIT\n",
      "2023-04-17 18:42:26.782418\n",
      "\n",
      "Testing # 2 5\n",
      "2023-04-17 18:42:30.101799\n",
      "2023-04-17 18:42:30.101840\n",
      "PREDICT\n",
      "2023-04-17 18:42:30.105552\n",
      "debug pred [0.57, 0.52, 0.19]\n",
      "debug class [1 1 0]\n",
      "Test 1:0 246:289 45.98%\n",
      "Predict 1:0 146:389 27.29%\n",
      "Test sizes (535, 1025) (535,)\n",
      "Distrib of scores: 0.4389906542056075 mean 0.11244483551569388 std\n",
      "Range of scores: 0.13 to 0.94\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[238  51]\n",
      " [151  95]]\n",
      "Normalized matrix\n",
      " [[0.44485981 0.0953271 ]\n",
      " [0.28224299 0.17757009]]\n",
      "Extreme scores correct, pos:neg 12 98\n",
      "Extreme scores incorrect pos:neg 4 21\n",
      "Accuracy: 62.24% Precision: 65.07% Recall: 38.62%\n",
      "F1: 48.47% MCC: 0.2346\n",
      "AUPRC: 64.63% AUROC: 71.27%\n",
      " accuracy [62.24299065420561]\n",
      " precision [65.06849315068493]\n",
      " recall [38.61788617886179]\n",
      " F1 [48.46938775510205]\n",
      " MCC [0.2346285254590521]\n",
      " AUPRC [64.62832454515643]\n",
      " AUROC [71.27324387430725]\n",
      "2023-04-17 18:42:30.194073\n"
     ]
    }
   ],
   "source": [
    "print(datetime.now())\n",
    "\n",
    "accuracy=[]\n",
    "precision=[]\n",
    "recall=[]\n",
    "f1=[]\n",
    "mcc=[]\n",
    "auprc=[]\n",
    "auroc=[]\n",
    "\n",
    "loader = DataLoader()\n",
    "filepath = DATA_DIR+RCI_FILE_TRAIN\n",
    "print(\"Load RCI from\",filepath)\n",
    "gene_to_rci = loader.load_gene_rci_values(filepath,EXCLUSIONS)\n",
    "print('Num RCI:', len(gene_to_rci.keys()))\n",
    "filepath = DATA_DIR+SEQ_FILE_TRAIN\n",
    "print('Load sequence from',filepath)\n",
    "allids,allseq = loader.load_spectra(filepath)  # for MLP (load_sequence() for CNN)\n",
    "print('Loaded',len(allseq),'sequences.')\n",
    "test_gene_to_rci = None\n",
    "test_allids = None\n",
    "test_allseq = None\n",
    "if SEQ_FILE_TEST is not None:\n",
    "    # Train on the entire train set (no cross-validation).\n",
    "    # Evaluate with the test files.\n",
    "    test_loader = DataLoader()\n",
    "    filepath = DATA_DIR+RCI_FILE_TEST\n",
    "    print(\"Load RCI from\",filepath)\n",
    "    test_gene_to_rci = test_loader.load_gene_rci_values(filepath,EXCLUSIONS)\n",
    "    print('Num RCI:', len(test_gene_to_rci.keys()))\n",
    "    filepath = DATA_DIR+SEQ_FILE_TEST\n",
    "    print('Load sequence from',filepath)\n",
    "    test_allids,test_allseq = test_loader.load_spectra(filepath)\n",
    "    print('Loaded',len(test_allseq),'sequences.')\n",
    "\n",
    "for repeat in range(REPEATS):\n",
    "    for fold in range(FOLDS):\n",
    "        show_r = repeat+1  # display one-based counting\n",
    "        show_f = fold+1    # display one-based counting\n",
    "\n",
    "        print()\n",
    "        print(\"Training #\",show_r,show_f)\n",
    "        print(datetime.now())\n",
    "        cvdo = CrossValidator(EPOCHS)\n",
    "        cvdo.set_threshold_mechanism(RCI_THRESHOLD_MECHANISM)\n",
    "        if SEQ_FILE_TEST is None:\n",
    "            # Train on 80% and evaluate on 20%.\n",
    "            separator = Separator()\n",
    "            separator.load(DATA_DIR,show_r,show_f)\n",
    "            separator.process(allids,allseq,gene_to_rci)\n",
    "            train_allids,test_allids = separator.get_ids()\n",
    "            train_allseq,test_allseq = separator.get_seq()\n",
    "            train_gene_to_rci,test_gene_to_rci = separator.get_rci()\n",
    "            cvdo.train_new_model(\n",
    "                train_allids,train_allseq,train_gene_to_rci,\n",
    "                test_allids,test_allseq,test_gene_to_rci)\n",
    "            if SAVE_MODEL_FILENAME is not None:\n",
    "                filename = f\"{SAVE_MODEL_FILENAME}.{show_r}.{show_f}.model\"\n",
    "                cvdo.save_model(filename)\n",
    "        else:\n",
    "            # Train on the entire train set (no cross-validation).\n",
    "            # Evaluate with the test files.\n",
    "            train_allids = allids\n",
    "            train_allseq = allseq\n",
    "            train_gene_to_rci = gene_to_rci\n",
    "            BREAK = True\n",
    "            cvdo.train_new_model(\n",
    "                train_allids,train_allseq,train_gene_to_rci,\n",
    "                None,None,None)\n",
    "\n",
    "        print()\n",
    "        print(\"Testing #\",show_r,show_f)\n",
    "        print(datetime.now())\n",
    "        cvdo.reset_statistics()\n",
    "        cvdo.test_without_training(\n",
    "            test_allids,test_allseq,test_gene_to_rci)\n",
    "        cv_accuracy,cv_precision,cv_recall,cv_f1,cv_mcc,cv_auprc,cv_auroc=\\\n",
    "            cvdo.get_statistics()\n",
    "\n",
    "        print(\" accuracy\" ,  cv_accuracy)\n",
    "        print(\" precision\" , cv_precision)\n",
    "        print(\" recall\" ,    cv_recall)\n",
    "        print(\" F1\" ,        cv_f1)\n",
    "        print(\" MCC\" ,       cv_mcc)\n",
    "        print(\" AUPRC\" ,     cv_auprc)\n",
    "        print(\" AUROC\" ,     cv_auroc)\n",
    "\n",
    "        accuracy.append(cv_accuracy)\n",
    "        precision.append(cv_precision)\n",
    "        recall.append(cv_recall)\n",
    "        f1.append(cv_f1)\n",
    "        mcc.append(cv_mcc)\n",
    "        auprc.append(cv_auprc)\n",
    "        auroc.append(cv_auroc)\n",
    "        if BREAK: break\n",
    "    if BREAK: break\n",
    "print(datetime.now())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "HkCeDg_HdQ36",
    "outputId": "b8486985-4447-4c8b-dde8-bbda95f9cc6d"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " accuracy mean 63.79% std 2.02\n",
      " precision mean 63.25% std 3.70\n",
      " recall mean 42.16% std 3.82\n",
      " F1 mean 50.51% std 3.34\n",
      " MCC mean 0.2505 std 0.0411\n",
      " AUPRC mean 62.21% std 3.91\n",
      " AUROC mean 69.74% std 3.02\n",
      " accuracy [[63.821892393320965], [63.73626373626373], [65.1252408477842], [59.417475728155345], [64.41947565543072], [63.23809523809524], [66.35514018691589], [63.2768361581921], [66.22390891840607], [62.24299065420561]]\n",
      " precision [[60.78431372549019], [63.030303030303024], [60.86956521739131], [57.971014492753625], [70.46979865771812], [60.37735849056604], [66.66666666666666], [65.5844155844156], [61.68831168831169], [65.06849315068493]]\n",
      " recall [[40.78947368421053], [43.15352697095436], [45.370370370370374], [34.63203463203463], [41.832669322709165], [42.47787610619469], [48.739495798319325], [41.56378600823045], [44.39252336448598], [38.61788617886179]]\n",
      " F1 [[48.818897637795274], [51.23152709359607], [51.9893899204244], [43.360433604336045], [52.5], [49.87012987012987], [56.310679611650485], [50.88161209068009], [51.63043478260869], [48.46938775510205]]\n",
      " MCC [[0.2355501927041128], [0.2503635709996857], [0.2619059851756257], [0.15956356113707784], [0.2924920063098742], [0.23068509889216196], [0.309872519465028], [0.25428796487539973], [0.27582320117194503], [0.2346285254590521]]\n",
      " AUPRC [[60.94141683491403], [64.02230771131678], [60.101585147697214], [58.148628781856516], [69.05552768738255], [57.59958557050786], [67.24319969997583], [61.719021254756434], [58.68620519570209], [64.62832454515643]]\n",
      " AUROC [[68.9978563772776], [71.45908441602612], [71.96552988632196], [64.42976038046461], [72.00244956569482], [65.76419924823156], [74.05073140367257], [69.53732281664381], [67.92944373115164], [71.27324387430725]]\n"
     ]
    }
   ],
   "source": [
    "def STD (values):\n",
    "    # ddof=1 reduces bias when extrapolating from sample to population\n",
    "    return np.std(values,ddof=1)\n",
    "\n",
    "print(\" accuracy mean %.2f%% std %.2f\" %  (np.mean(accuracy),  STD(accuracy)))\n",
    "print(\" precision mean %.2f%% std %.2f\" % (np.mean(precision), STD(precision)))\n",
    "print(\" recall mean %.2f%% std %.2f\" %    (np.mean(recall),    STD(recall)))\n",
    "print(\" F1 mean %.2f%% std %.2f\" %        (np.mean(f1),        STD(f1)))\n",
    "print(\" MCC mean %.4f std %.4f\" %       (np.mean(mcc),       STD(mcc)))\n",
    "print(\" AUPRC mean %.2f%% std %.2f\" %     (np.mean(auprc),     STD(auprc)))\n",
    "print(\" AUROC mean %.2f%% std %.2f\" %     (np.mean(auroc),     STD(auroc)))\n",
    "\n",
    "print(\" accuracy\"  , accuracy)\n",
    "print(\" precision\" , precision)\n",
    "print(\" recall\"    , recall)\n",
    "print(\" F1\"        , f1)\n",
    "print(\" MCC\"       , mcc)\n",
    "print(\" AUPRC\"     , auprc)\n",
    "print(\" AUROC\"     , auroc)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "QjSVa72v4IsA"
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "gpuClass": "standard",
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
