{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "PG-tGRnlFLA3"
   },
   "source": [
    "# Random Forest\n",
    "canonical lncRNA, GMM, cross-valiation, all the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "0RmwUsVLFLA6",
    "outputId": "64cc44e8-c4cd-4eee-f276-47d1983f985c"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2023-04-17 18:31:48.161156\n"
     ]
    }
   ],
   "source": [
    "from datetime import datetime\n",
    "print(datetime.now())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "OlzN9OdsFWEU",
    "outputId": "20173346-1246-44f4-eaf7-2a5f2c032df7"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-04-17 18:31:48.201695: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GPU device not found\n",
      "/Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-04-17 18:31:59.068755: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\n"
     ]
    }
   ],
   "source": [
    "import tensorflow as tf\n",
    "dt='float32'\n",
    "tf.keras.backend.set_floatx('float32')\n",
    "# tf.random.set_seed(42) # supposedly leads to reproducible results\n",
    "\n",
    "device_name = tf.test.gpu_device_name()\n",
    "if device_name != '/device:GPU:0':\n",
    "    print('GPU device not found')\n",
    "else:\n",
    "    print('Found GPU at: {}'.format(device_name))\n",
    "\n",
    "try:\n",
    "    from google.colab import drive\n",
    "    IN_COLAB = True\n",
    "    print('Running on CoLab')\n",
    "    PATH='/content/drive/'\n",
    "    drive.mount(PATH)\n",
    "    DATA_DIR=PATH+'My Drive/data/Localization/TrainTest/TrainTest_ver43/'  # must end in \"/\"\n",
    "    MODEL_DIR=PATH+'My Drive/data/Localization/Models/'  # must end in \"/\"\n",
    "except:\n",
    "    IN_COLAB = False\n",
    "    DATA_DIR = 'D:/Adjeroh/Localization/TrainTest/'   # Windows\n",
    "    DATA_DIR = '/Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/'    # Mac\n",
    "    MODEL_DIR = '/Users/jasonmiller/WVU/Localization/Models/'    # Mac\n",
    "print(DATA_DIR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "PRX-UEr8FLA8",
    "outputId": "39890dac-1ca8-4ba2-9229-2b581f26b315"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Python 3.10.0\n",
      "sklearn 1.1.2\n"
     ]
    }
   ],
   "source": [
    "from platform import python_version\n",
    "print('Python',python_version())\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy.stats as ss\n",
    "import pickle\n",
    "import time # sleep function\n",
    "from os.path import isfile\n",
    "from matplotlib import pyplot as plt \n",
    "import sklearn   # pip install --upgrade scikit-learn\n",
    "print('sklearn',sklearn.__version__)\n",
    "\n",
    "from tensorflow import keras\n",
    "from sklearn.mixture import GaussianMixture\n",
    "from sklearn.model_selection import KFold\n",
    "from sklearn.metrics import accuracy_score\n",
    "from sklearn.metrics import precision_score\n",
    "from sklearn.metrics import recall_score\n",
    "from sklearn.metrics import f1_score\n",
    "from sklearn.metrics import precision_recall_curve\n",
    "from sklearn.metrics import roc_curve\n",
    "from sklearn.metrics import roc_auc_score\n",
    "from sklearn.metrics import auc\n",
    "from sklearn.metrics import matthews_corrcoef\n",
    "from sklearn.utils import class_weight\n",
    "from sklearn.metrics import confusion_matrix\n",
    "from sklearn.ensemble import RandomForestClassifier as RFC\n",
    "\n",
    "from keras.models import Sequential\n",
    "from keras.layers import Masking\n",
    "from keras.layers import Embedding\n",
    "from keras.layers import Conv1D\n",
    "from keras.layers import GlobalAveragePooling1D\n",
    "from keras.layers import GlobalMaxPooling1D\n",
    "from keras.layers import AveragePooling1D\n",
    "from keras.layers import MaxPooling1D\n",
    "from keras.layers import Bidirectional\n",
    "from keras.layers import LSTM\n",
    "from keras.layers import TimeDistributed\n",
    "from keras.layers import Flatten\n",
    "from keras.layers import Dense\n",
    "from keras.layers import Dropout\n",
    "from keras.losses import BinaryCrossentropy\n",
    "#from keras.losses import Hinge\n",
    "\n",
    "K=5\n",
    "ALPHABET=4**K + 1\n",
    "CFILTERS=64 \n",
    "FILTERSIZE=8\n",
    "RCELLS=32\n",
    "DCELLS=16\n",
    "EPOCHS=150 \n",
    "EMBED_DIMEN = 4 # arbitrary hyperparameter\n",
    "# MINLEN=200   # NO LONGER USED\n",
    "# MAXLEN=5000\n",
    "RCI_THRESHOLD_MECHANISM = 'RCI_GMM'  # 'RCI_GMM' 'ZERO' 'THE_MEAN'\n",
    "BREAK = False   # break after first fold\n",
    "EXCLUSIONS = [1]   # possibly exclude cell line 1 = H1.hESC\n",
    "FILTER_TRAIN        = False\n",
    "FILTER_TAILS_TRAIN  = False\n",
    "FILTER_MIDDLE_TRAIN = False\n",
    "FILTER_TEST         = False\n",
    "FILTER_TAILS_TEST   = False\n",
    "FILTER_MIDDLE_TEST  = False\n",
    "REPEATS = 2\n",
    "FOLDS = 5\n",
    "\n",
    "SAVE_MODEL_FILENAME = None # 'RF_100'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "id": "LnkpVKdMFLA-"
   },
   "outputs": [],
   "source": [
    "RCI_FILE_TRAIN = 'train.lncRNA_RCI.csv'\n",
    "RCI_FILE_TEST  = None #'test.lncRNA_RCI.csv'\n",
    "\n",
    "SEQ_FILE_TRAIN = 'train.canon_lncRNA_transcripts.csv'\n",
    "SEQ_FILE_TEST  = None #'test.canon_lncRNA_transcripts.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "id": "e3p4QzQJFLA_"
   },
   "outputs": [],
   "source": [
    "def get_ordered_list():\n",
    "    ordered_list = \\\n",
    "    ['A549','H1.hESC','HeLa.S3','HepG2','HT1080',\\\n",
    "      'HUVEC','MCF.7','NCI.H460','NHEK','SK.MEL.5',\\\n",
    "      'SK.N.DZ','SK.N.SH','GM12878','K562','IMR.90']\n",
    "    return ordered_list\n",
    "all_cell_lines = get_ordered_list()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "UtqdpJOxFLBA"
   },
   "source": [
    "## Data Load"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "id": "p35ehKV3Kq0z"
   },
   "outputs": [],
   "source": [
    "class DataLoader():\n",
    "    def __init__(self):\n",
    "        self.cache=dict() \n",
    "        self.vals = {'A':0, 'C':1, 'G':2, 'T':3}\n",
    "        self.gene2rci = dict()\n",
    "        \n",
    "    def load_gene_rci_values(self,filepath,exclusions):\n",
    "        '''\n",
    "        Load all the genes from the given RCI csv file.\n",
    "        The given file usually contains train or test, not both.\n",
    "        Return dict with keys=gene:str and values=RCI:float.\n",
    "        '''\n",
    "        self.gene2rci = {}\n",
    "        overall_sum = 0\n",
    "        with open (filepath,'r') as handle:\n",
    "            header = None\n",
    "            for row in handle:\n",
    "                if header is None:\n",
    "                    header = row # skip file's header line\n",
    "                else:\n",
    "                    line = row.strip()\n",
    "                    fields = line.split(',')\n",
    "                    gene_id = fields.pop(0)\n",
    "                    log_mean = float(fields.pop(0))\n",
    "                    self.gene2rci[gene_id] = log_mean\n",
    "        print('Number of RCI values loaded',len(self.gene2rci.keys()))\n",
    "        return self.gene2rci\n",
    "\n",
    "    def _seq_to_kmer_values(self,rna,K):\n",
    "        vec=[] # seq converted to list of K-mers \n",
    "        N_indicator = 0 # indicator value\n",
    "        length = len(rna)\n",
    "        for i in range(length-K+1):\n",
    "            kmer = rna[i:i+K]\n",
    "            if 'N' in kmer:\n",
    "                value = N_indicator\n",
    "            elif kmer in self.cache.keys():\n",
    "                value = self.cache[kmer]\n",
    "            else:\n",
    "                value = 0\n",
    "                for j in range(K):\n",
    "                    value *= 4   \n",
    "                    nextnuc = kmer[j] \n",
    "                    nucval = self.vals[nextnuc]\n",
    "                    value += nucval\n",
    "                value += 1   # NNN => 0, AAA => 1\n",
    "                self.cache[kmer] = value\n",
    "            vec.append(value)\n",
    "        return vec\n",
    "\n",
    "    def load_sequence(self,filepath):\n",
    "        '''\n",
    "        Load all the sequences from the given file. \n",
    "        Load our version of GenCode -- csv, easier to parse than fasta.\n",
    "        Each line has IDs plus sequence.\n",
    "        The IDs in the file do not include dot-version numbers.\n",
    "        The file may already be filtered e.g. canonical transcripts.\n",
    "        '''\n",
    "        allids=[]\n",
    "        allseq=[]\n",
    "        #NREPEAT = str('N'*MAXLEN)   # not used for MLP\n",
    "        with open (filepath,'r') as handle:\n",
    "            header = None\n",
    "            for row in handle:\n",
    "                if header is None:\n",
    "                    header = row\n",
    "                else:\n",
    "                    line    = row.strip()\n",
    "                    fields  = line.split(',')\n",
    "                    tran_id = fields[0]  # without version number\n",
    "                    gene_id = fields[1]  # without version number\n",
    "                    seq_len = int(fields[3])\n",
    "                    seq_txt = fields[4]\n",
    "                    # Keep only transcripts having numeric RCI given the cell lines in use.\n",
    "                    # We have validated this by spot checking.\n",
    "                    # TO DO: validate this programmatically.\n",
    "                    if gene_id in self.gene2rci.keys():\n",
    "                        # no MAXLEN for MLP\n",
    "                        #if seq_len<=MAXLEN:\n",
    "                        #    seq_txt = seq_txt + NREPEAT\n",
    "                        #seq_txt = seq_txt[:MAXLEN]\n",
    "                        allids.append( (gene_id,tran_id) )\n",
    "                        hot_vec = self._seq_to_kmer_values(seq_txt,K)\n",
    "                        allseq.append(hot_vec)\n",
    "        self.cache=dict() # save RAM\n",
    "        return allids,allseq\n",
    "\n",
    "    def load_spectra(self,filepath):\n",
    "        '''\n",
    "        Load all (variable-length) sequences as lists of kmers.\n",
    "        Then convert each sequence to (fixed-length) kmer histograms.\n",
    "        '''\n",
    "        allids,allseq = self.load_sequence(filepath)\n",
    "        allspectra = []\n",
    "        for seq in allseq:\n",
    "            spectrum = np.zeros(ALPHABET)\n",
    "            for kmer in seq:\n",
    "                spectrum[kmer] += 1\n",
    "            spectrum /= len(seq)\n",
    "            allspectra.append(spectrum)\n",
    "        return allids,allspectra        \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "VDZ6siB_Kq04"
   },
   "source": [
    "## Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "id": "AwMbRjm0FLBF"
   },
   "outputs": [],
   "source": [
    "def build_model():\n",
    "    rfc = RFC()\n",
    "    return rfc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "clj-wufgFLBF",
    "outputId": "a17cf7a6-3755-48aa-d96b-dd0383c6f2f2"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2023-04-17 18:32:01.821163\n",
      "RandomForestClassifier()\n"
     ]
    }
   ],
   "source": [
    "print(datetime.now())\n",
    "model=build_model()\n",
    "print(model)  # Print this only once\n",
    "model=None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "MgrC1alOKq07"
   },
   "source": [
    "## Training"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "id": "W9xiFzNbFLBE"
   },
   "outputs": [],
   "source": [
    "import gc\n",
    "class CrossValidator():\n",
    "    def __init__(self,epochs,score_threshold=0.5):\n",
    "        self.epochs = epochs\n",
    "        self.score_threshold = score_threshold\n",
    "        self.mechanism = 'ZERO'\n",
    "        self.discriminator = -1 # or zero\n",
    "        self.flip = False\n",
    "        self.reset_statistics()\n",
    "        \n",
    "    def reset_statistics(self):\n",
    "        self.cv_accuracy=[]\n",
    "        self.cv_precision=[]\n",
    "        self.cv_recall=[]\n",
    "        self.cv_f1=[]\n",
    "        self.cv_auprc=[]\n",
    "        self.cv_auroc=[]\n",
    "        self.cv_mcc=[]\n",
    "        \n",
    "    def _get_X_y(self, all_ids, all_seqs, rci_map): \n",
    "        # Prepare X and y for training or testing.\n",
    "        subsetX=[]\n",
    "        subsetY=[]\n",
    "        for t in range(len(all_ids)):\n",
    "            gene_id,tran_id = all_ids[t]\n",
    "            oneX            = all_seqs[t]\n",
    "            oneY            = rci_map[gene_id]\n",
    "            subsetX.append(oneX)\n",
    "            subsetY.append(oneY)\n",
    "        subsetX = np.array(subsetX)\n",
    "        subsetY = np.array(subsetY).reshape((-1,1))\n",
    "        return subsetX,subsetY\n",
    "    \n",
    "    def set_threshold_mechanism(self, mechanism):\n",
    "        if mechanism not in ['RCI_GMM','THE_MEAN','ZERO']:\n",
    "            raise Exception('Unrecognized mechansm:',mechanism)\n",
    "        self.mechanism = mechanism\n",
    "    \n",
    "    def _apply_threshold(self, array_of_rci):\n",
    "        # Takes list of float, returns list of labels [0,1].\n",
    "        if self.mechanism == 'RCI_GMM':\n",
    "            labels = self.discriminator.predict(array_of_rci)\n",
    "            if self.flip:\n",
    "                IS_CYTO = lambda label: 1 if label==0 else 0\n",
    "                labels = np.array(list(map(IS_CYTO, labels)))\n",
    "        else:  # 'THE_MEAN' or 'ZERO'\n",
    "            rci_threshold = self.discriminator\n",
    "            IS_CYTO = lambda rci: 1 if rci>rci_threshold else 0\n",
    "            labels = np.array(list(map(IS_CYTO, array_of_rci)))\n",
    "        return labels\n",
    "    \n",
    "    def _prepare_threshold(self, rci_values, create=True):\n",
    "        if self.mechanism == 'RCI_GMM':\n",
    "            if create:  # during training, create a new GMM\n",
    "                gmm = GaussianMixture(n_components=2, verbose=0, \n",
    "                  covariance_type='spherical', n_init=100) # random_state=42) \n",
    "                gmm.fit(rci_values)\n",
    "            else:   # during testing, use existing GMM\n",
    "                gmm=self.discriminator\n",
    "            self.flip = False\n",
    "            # The GMM labels are arbitrary.\n",
    "            if gmm.means_[0][0] > gmm.means_[1][0]:\n",
    "                self.flip = True\n",
    "            self.discriminator = gmm   # redundant but consistent\n",
    "        elif self.mechanism == 'THE_MEAN':\n",
    "            self.discriminator = np.mean(rci_values)\n",
    "        elif self.mechanism == 'ZERO':\n",
    "            self.discriminator = -1   # 0 usually, -1 is as in Yuan et al.\n",
    "        else: # not expected\n",
    "            self.discriminator = 0\n",
    "    \n",
    "    def _explain_threshold(self):\n",
    "        if self.mechanism == 'RCI_GMM':\n",
    "            gmm=self.discriminator\n",
    "            print('Discriminator is GMM')\n",
    "            print('Means',[gmm.means_[0][0],gmm.means_[1][0]])\n",
    "            print('Variances',gmm.covariances_)\n",
    "            print('Priors',gmm.weights_)\n",
    "            test_rcis=[-5,-4,-3.5,-3,-2.5,-2,-1.5,-1,-0.5,0,0.5,1,1.5,2,3]\n",
    "            print(test_rcis)\n",
    "            print(self._apply_threshold(np.array(test_rcis).reshape((-1,1))))\n",
    "        else:\n",
    "            print('Discriminator',self.mechanism,self.discriminator)\n",
    "    \n",
    "    def _show_sizes(self,label,values):\n",
    "        a = np.count_nonzero(values==1)\n",
    "        b = np.count_nonzero(values==0)\n",
    "        print('%s 1:0 %d:%d %5.2f%%'%(label,a,b,100*a/(a+b)))\n",
    "        \n",
    "    def save_model(self,filename):\n",
    "        if self.model is not None:\n",
    "            filepath = MODEL_DIR + filename\n",
    "            #self.model.save(filepath)\n",
    "            print('? Saved model to',filepath)\n",
    "        \n",
    "    def load_model(self,filename):\n",
    "        filepath = MODEL_DIR + filename\n",
    "        #self.model = keras.models.load_model(filepath)\n",
    "        print('? Loaded model from',filepath)\n",
    "        \n",
    "    def train_new_model(self,train_ids,train_seq,train_rci,\n",
    "            valid_ids=None,valid_seq=None,valid_rci=None):\n",
    "        print(datetime.now())\n",
    "        X_train,y_rci = self._get_X_y(train_ids,train_seq,train_rci) \n",
    "        self._prepare_threshold(y_rci,True)  \n",
    "        self._explain_threshold()\n",
    "        y_train = self._apply_threshold(y_rci)\n",
    "        self._show_sizes('Train',y_train)\n",
    "        #cw = class_weight.compute_class_weight('balanced', classes=[0,1], y=y_train)\n",
    "        #cwd = {0: cw[0], 1: cw[1]}\n",
    "        #print('Computed class weights:',cwd)\n",
    "        if valid_ids is not None:\n",
    "            X_valid,y_rci = self._get_X_y(valid_ids,valid_seq,valid_rci) \n",
    "            y_valid = self._apply_threshold(y_rci)\n",
    "            self._show_sizes('Valid',y_valid)\n",
    "        y_rci = None\n",
    "\n",
    "        self.model=build_model()\n",
    "        \n",
    "        print(\"FIT\")\n",
    "        print(datetime.now())\n",
    "        self.model.fit(X_train, y_train) # sample weight\n",
    "\n",
    "    def test_without_training(self,test_ids,test_seq,test_rci):\n",
    "        # For final test, do no train.\n",
    "        # Assume set_sequences() set the test set.\n",
    "        print(datetime.now())\n",
    "        X_test,y_rci = self._get_X_y(test_ids,test_seq,test_rci) \n",
    "        y_test = self._apply_threshold(y_rci)\n",
    "        y_rci = None\n",
    "        \n",
    "        print(\"PREDICT\")\n",
    "        print(datetime.now())        \n",
    "        yhat_pairs=self.model.predict_proba(X_test)  # [ prob of 0, prob of 1 ]\n",
    "        yhat_pred=[pair[1] for pair in yhat_pairs]\n",
    "        yhat_classes=self.model.predict(X_test)  # 0 or 1\n",
    "        \n",
    "        print('debug pred',yhat_pred[:3])\n",
    "        print('debug class',yhat_classes[:3])\n",
    "\n",
    "        self._show_sizes('Test',y_test)\n",
    "        self._show_sizes('Predict',yhat_classes)\n",
    "        print('Test sizes',X_test.shape,y_test.shape)\n",
    "        print('Distrib of scores:',np.mean(yhat_pred),'mean',np.std(yhat_pred),'std')\n",
    "        print('Range of scores:',np.min(yhat_pred),'to',np.max(yhat_pred))\n",
    "        print('Score threshold',self.score_threshold)\n",
    "        cm1 = confusion_matrix(y_test,yhat_classes)\n",
    "        print('Confusion matrix\\n',cm1)\n",
    "        cm2 = confusion_matrix(y_test,yhat_classes,normalize='all')\n",
    "        print('Normalized matrix\\n',cm2)\n",
    "\n",
    "        accuracy = accuracy_score(y_test, yhat_classes)*100.\n",
    "        precision = precision_score(y_test, yhat_classes)*100.\n",
    "        recall = recall_score(y_test, yhat_classes)*100.\n",
    "        f1 = f1_score(y_test, yhat_classes)*100.\n",
    "        prc_Y, prc_X, prc_bins = precision_recall_curve(y_test, yhat_pred)\n",
    "        auprc = auc(prc_X,prc_Y)*100.\n",
    "        auroc = roc_auc_score(y_test, yhat_pred)*100.\n",
    "        mcc = matthews_corrcoef(y_test, yhat_classes)\n",
    "\n",
    "        correct_pos = 0\n",
    "        correct_neg = 0\n",
    "        wrong_pos = 0\n",
    "        wrong_neg = 0\n",
    "        for i in range(len(y_test)):\n",
    "            if yhat_pred[i]>=0.65:\n",
    "                if y_test[i]==1:\n",
    "                    correct_pos += 1\n",
    "                else:\n",
    "                    wrong_pos += 1\n",
    "            elif yhat_pred[i]<=0.35:\n",
    "                if y_test[i]==0:\n",
    "                    correct_neg += 1\n",
    "                else:\n",
    "                    wrong_neg += 1\n",
    "        print('Extreme scores correct, pos:neg',correct_pos,correct_neg)  \n",
    "        print('Extreme scores incorrect pos:neg',wrong_pos,wrong_neg)  \n",
    "\n",
    "        #self._pretty_pictures(y_test,yhat_pred,prc_X,prc_Y)\n",
    "        \n",
    "        print('Accuracy: %.2f%% Precision: %.2f%% Recall: %.2f%%' % (accuracy,precision,recall)) \n",
    "        print('F1: %.2f%% MCC: %.4f' % (f1,mcc)) \n",
    "        print('AUPRC: %.2f%% AUROC: %.2f%%' % (auprc,auroc)) \n",
    "\n",
    "        self.cv_accuracy.append(accuracy)\n",
    "        self.cv_precision.append(precision)\n",
    "        self.cv_recall.append(recall)\n",
    "        self.cv_f1.append(f1)\n",
    "        self.cv_mcc.append(mcc)\n",
    "        self.cv_auprc.append(auprc)\n",
    "        self.cv_auroc.append(auroc)\n",
    "\n",
    "    def _pretty_pictures(self,y_valid,yhat_pred,prc_X,prc_Y):\n",
    "        count_ones= len(y_valid[y_valid==1])\n",
    "        count_zeros= len(y_valid[y_valid==0])\n",
    "        guess = max(count_ones,count_zeros) / len(y_valid)\n",
    "        # PRC\n",
    "        plt.plot(prc_X, prc_Y, marker='.')\n",
    "        plt.plot([0, 1], [guess,guess], linestyle='--')\n",
    "        plt.xlabel('Recall')\n",
    "        plt.ylabel('Precision')\n",
    "        plt.show()\n",
    "        # ROC\n",
    "        fpr, tpr, roc_bins = roc_curve(y_valid, yhat_pred)\n",
    "        plt.plot(fpr, tpr, marker='.')\n",
    "        plt.xlabel('False Positive Rate')\n",
    "        plt.ylabel('True Positive Rate')\n",
    "        plt.show()\n",
    "        \n",
    "    def get_statistics(self):\n",
    "        return \\\n",
    "        self.cv_accuracy,\\\n",
    "        self.cv_precision,\\\n",
    "        self.cv_recall,\\\n",
    "        self.cv_f1,\\\n",
    "        self.cv_mcc,\\\n",
    "        self.cv_auprc,\\\n",
    "        self.cv_auroc\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "Td97uyyj5qDq"
   },
   "source": [
    "## Train"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "id": "poe3rwAF4Ir7"
   },
   "outputs": [],
   "source": [
    "class Separator():\n",
    "    def __init__(self):\n",
    "        self.train_ids = []\n",
    "        self.train_seq = []\n",
    "        self.train_rci = dict()\n",
    "        self.val_ids = []\n",
    "        self.val_seq = []\n",
    "        self.val_rci = dict()\n",
    "    def load(self,data_dir,rep,fold):\n",
    "        filename='cv.{}.{}.validation_genes.txt'.format(rep,fold)\n",
    "        filename = data_dir + filename\n",
    "        self.val_genes = set()\n",
    "        print('Opening file',filename)\n",
    "        with open(filename,'r') as fin:\n",
    "            for line in fin:\n",
    "                gene_id = line.strip()\n",
    "                self.val_genes.add(gene_id)\n",
    "    def process(self,allids,allseq,gene_to_rci):\n",
    "        size = len(allids)\n",
    "        for t in range(size):\n",
    "            gene_id,tran_id = allids[t]\n",
    "            oneX            = allseq[t]\n",
    "            oneY            = gene_to_rci[gene_id]\n",
    "            in_middle = gene_to_rci[gene_id] >= -2 and gene_to_rci[gene_id] <= 0\n",
    "            in_tails = gene_to_rci[gene_id] < -2 or gene_to_rci[gene_id] > 0\n",
    "            if gene_id in self.val_genes:\n",
    "                if FILTER_TEST and (\\\n",
    "                    (FILTER_TAILS_TEST and in_tails) or \\\n",
    "                    (FILTER_MIDDLE_TEST and in_middle)):\n",
    "                    pass\n",
    "                else:\n",
    "                    self.val_ids.append(allids[t])\n",
    "                    self.val_seq.append(allseq[t])\n",
    "                    self.val_rci[gene_id]=gene_to_rci[gene_id]\n",
    "            else:\n",
    "                if FILTER_TRAIN and (\\\n",
    "                    (FILTER_TAILS_TRAIN and in_tails) or \\\n",
    "                    (FILTER_MIDDLE_TRAIN and in_middle)):\n",
    "                    pass\n",
    "                else:\n",
    "                    self.train_ids.append(allids[t])\n",
    "                    self.train_seq.append(allseq[t])\n",
    "                    self.train_rci[gene_id]=gene_to_rci[gene_id]\n",
    "    def get_ids(self):\n",
    "        return self.train_ids,self.val_ids\n",
    "    def get_seq(self):\n",
    "        return self.train_seq,self.val_seq\n",
    "    def get_rci(self):\n",
    "        return self.train_rci,self.val_rci"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 1000
    },
    "id": "XC9m0W-pFLBH",
    "outputId": "f1051adc-9a79-4671-dd2b-2d9617da1129",
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2023-04-17 18:32:01.914476\n",
      "Load RCI from /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/train.lncRNA_RCI.csv\n",
      "Number of RCI values loaded 4371\n",
      "Num RCI: 4371\n",
      "Load sequence from /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/train.canon_lncRNA_transcripts.csv\n",
      "Loaded 4371 sequences.\n",
      "\n",
      "Training # 1 1\n",
      "2023-04-17 18:32:07.939875\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.1.1.validation_genes.txt\n",
      "2023-04-17 18:32:07.947729\n",
      "Discriminator is GMM\n",
      "Means [-0.1235364916771634, -2.7492144436660695]\n",
      "Variances [1.50711037 1.99461298]\n",
      "Priors [0.57052389 0.42947611]\n",
      "[-5, -4, -3.5, -3, -2.5, -2, -1.5, -1, -0.5, 0, 0.5, 1, 1.5, 2, 3]\n",
      "[0 0 0 0 0 0 1 1 1 1 1 1 1 1 1]\n",
      "Train 1:0 2091:1405 59.81%\n",
      "Valid 1:0 508:367 58.06%\n",
      "FIT\n",
      "2023-04-17 18:32:09.232044\n",
      "\n",
      "Testing # 1 1\n",
      "2023-04-17 18:32:14.889062\n",
      "2023-04-17 18:32:14.889090\n",
      "PREDICT\n",
      "2023-04-17 18:32:14.892879\n",
      "debug pred [0.54, 0.51, 0.55]\n",
      "debug class [1 1 1]\n",
      "Test 1:0 508:367 58.06%\n",
      "Predict 1:0 740:135 84.57%\n",
      "Test sizes (875, 1025) (875,)\n",
      "Distrib of scores: 0.5958971428571429 mean 0.09284562143907575 std\n",
      "Range of scores: 0.28 to 0.96\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[ 95 272]\n",
      " [ 40 468]]\n",
      "Normalized matrix\n",
      " [[0.10857143 0.31085714]\n",
      " [0.04571429 0.53485714]]\n",
      "Extreme scores correct, pos:neg 196 4\n",
      "Extreme scores incorrect pos:neg 75 3\n",
      "Accuracy: 64.34% Precision: 63.24% Recall: 92.13%\n",
      "F1: 75.00% MCC: 0.2461\n",
      "AUPRC: 69.55% AUROC: 66.24%\n",
      " accuracy [64.34285714285714]\n",
      " precision [63.24324324324324]\n",
      " recall [92.1259842519685]\n",
      " F1 [75.00000000000001]\n",
      " MCC [0.24605548996989562]\n",
      " AUPRC [69.54985943991282]\n",
      " AUROC [66.23586646355854]\n",
      "\n",
      "Training # 1 2\n",
      "2023-04-17 18:32:14.992379\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.1.2.validation_genes.txt\n",
      "2023-04-17 18:32:15.005590\n",
      "Discriminator is GMM\n",
      "Means [-0.15507654504140794, -2.755657665223358]\n",
      "Variances [1.42953516 2.06354601]\n",
      "Priors [0.567216 0.432784]\n",
      "[-5, -4, -3.5, -3, -2.5, -2, -1.5, -1, -0.5, 0, 0.5, 1, 1.5, 2, 3]\n",
      "[0 0 0 0 0 0 1 1 1 1 1 1 1 1 1]\n",
      "Train 1:0 2091:1405 59.81%\n",
      "Valid 1:0 516:359 58.97%\n",
      "FIT\n",
      "2023-04-17 18:32:16.258356\n",
      "\n",
      "Testing # 1 2\n",
      "2023-04-17 18:32:21.859457\n",
      "2023-04-17 18:32:21.859488\n",
      "PREDICT\n",
      "2023-04-17 18:32:21.863198\n",
      "debug pred [0.58, 0.42, 0.64]\n",
      "debug class [1 0 1]\n",
      "Test 1:0 516:359 58.97%\n",
      "Predict 1:0 720:155 82.29%\n",
      "Test sizes (875, 1025) (875,)\n",
      "Distrib of scores: 0.5855771428571429 mean 0.09241016358980152 std\n",
      "Range of scores: 0.22 to 0.9\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[105 254]\n",
      " [ 50 466]]\n",
      "Normalized matrix\n",
      " [[0.12       0.29028571]\n",
      " [0.05714286 0.53257143]]\n",
      "Extreme scores correct, pos:neg 172 8\n",
      "Extreme scores incorrect pos:neg 62 5\n",
      "Accuracy: 65.26% Precision: 64.72% Recall: 90.31%\n",
      "F1: 75.40% MCC: 0.2520\n",
      "AUPRC: 74.50% AUROC: 67.60%\n",
      " accuracy [65.25714285714285]\n",
      " precision [64.72222222222223]\n",
      " recall [90.31007751937985]\n",
      " F1 [75.40453074433657]\n",
      " MCC [0.2519788918294678]\n",
      " AUPRC [74.50263885679924]\n",
      " AUROC [67.603809030252]\n",
      "\n",
      "Training # 1 3\n",
      "2023-04-17 18:32:21.953621\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.1.3.validation_genes.txt\n",
      "2023-04-17 18:32:21.969194\n",
      "Discriminator is GMM\n",
      "Means [-0.10647496210854528, -2.78199592941014]\n",
      "Variances [1.49864906 2.0823436 ]\n",
      "Priors [0.57174195 0.42825805]\n",
      "[-5, -4, -3.5, -3, -2.5, -2, -1.5, -1, -0.5, 0, 0.5, 1, 1.5, 2, 3]\n",
      "[0 0 0 0 0 0 1 1 1 1 1 1 1 1 1]\n",
      "Train 1:0 2095:1402 59.91%\n",
      "Valid 1:0 510:364 58.35%\n",
      "FIT\n",
      "2023-04-17 18:32:23.207406\n",
      "\n",
      "Testing # 1 3\n",
      "2023-04-17 18:32:28.813016\n",
      "2023-04-17 18:32:28.813046\n",
      "PREDICT\n",
      "2023-04-17 18:32:28.816455\n",
      "debug pred [0.63, 0.62, 0.4]\n",
      "debug class [1 1 0]\n",
      "Test 1:0 510:364 58.35%\n",
      "Predict 1:0 720:154 82.38%\n",
      "Test sizes (874, 1025) (874,)\n",
      "Distrib of scores: 0.5879290617848969 mean 0.09500181620467112 std\n",
      "Range of scores: 0.27 to 0.94\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[ 93 271]\n",
      " [ 61 449]]\n",
      "Normalized matrix\n",
      " [[0.10640732 0.31006865]\n",
      " [0.06979405 0.51372998]]\n",
      "Extreme scores correct, pos:neg 142 7\n",
      "Extreme scores incorrect pos:neg 73 2\n",
      "Accuracy: 62.01% Precision: 62.36% Recall: 88.04%\n",
      "F1: 73.01% MCC: 0.1758\n",
      "AUPRC: 66.44% AUROC: 61.78%\n",
      " accuracy [62.013729977116704]\n",
      " precision [62.361111111111114]\n",
      " recall [88.0392156862745]\n",
      " F1 [73.00813008130082]\n",
      " MCC [0.17582717682293555]\n",
      " AUPRC [66.44407027191536]\n",
      " AUROC [61.775748761042884]\n",
      "\n",
      "Training # 1 4\n",
      "2023-04-17 18:32:28.910099\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.1.4.validation_genes.txt\n",
      "2023-04-17 18:32:28.925308\n",
      "Discriminator is GMM\n",
      "Means [-2.7618271525964446, -0.10819901877560016]\n",
      "Variances [1.96087741 1.47141094]\n",
      "Priors [0.4408093 0.5591907]\n",
      "[-5, -4, -3.5, -3, -2.5, -2, -1.5, -1, -0.5, 0, 0.5, 1, 1.5, 2, 3]\n",
      "[0 0 0 0 0 0 1 1 1 1 1 1 1 1 1]\n",
      "Train 1:0 2025:1472 57.91%\n",
      "Valid 1:0 532:342 60.87%\n",
      "FIT\n",
      "2023-04-17 18:32:30.176055\n",
      "\n",
      "Testing # 1 4\n",
      "2023-04-17 18:32:36.186896\n",
      "2023-04-17 18:32:36.186925\n",
      "PREDICT\n",
      "2023-04-17 18:32:36.190387\n",
      "debug pred [0.56, 0.48, 0.78]\n",
      "debug class [1 0 1]\n",
      "Test 1:0 532:342 60.87%\n",
      "Predict 1:0 662:212 75.74%\n",
      "Test sizes (874, 1025) (874,)\n",
      "Distrib of scores: 0.5670594965675058 mean 0.09511250056909598 std\n",
      "Range of scores: 0.21 to 0.97\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[116 226]\n",
      " [ 96 436]]\n",
      "Normalized matrix\n",
      " [[0.13272311 0.25858124]\n",
      " [0.10983982 0.49885584]]\n",
      "Extreme scores correct, pos:neg 138 10\n",
      "Extreme scores incorrect pos:neg 44 3\n",
      "Accuracy: 63.16% Precision: 65.86% Recall: 81.95%\n",
      "F1: 73.03% MCC: 0.1807\n",
      "AUPRC: 71.95% AUROC: 64.40%\n",
      " accuracy [63.1578947368421]\n",
      " precision [65.86102719033232]\n",
      " recall [81.95488721804512]\n",
      " F1 [73.03182579564488]\n",
      " MCC [0.1807304768810816]\n",
      " AUPRC [71.95253109891995]\n",
      " AUROC [64.39508859868971]\n",
      "\n",
      "Training # 1 5\n",
      "2023-04-17 18:32:36.285165\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.1.5.validation_genes.txt\n",
      "2023-04-17 18:32:36.300842\n",
      "Discriminator is GMM\n",
      "Means [-2.800912846557972, -0.1593377607506266]\n",
      "Variances [2.07062503 1.51592921]\n",
      "Priors [0.42874853 0.57125147]\n",
      "[-5, -4, -3.5, -3, -2.5, -2, -1.5, -1, -0.5, 0, 0.5, 1, 1.5, 2, 3]\n",
      "[0 0 0 0 0 0 1 1 1 1 1 1 1 1 1]\n",
      "Train 1:0 2093:1405 59.83%\n",
      "Valid 1:0 540:333 61.86%\n",
      "FIT\n",
      "2023-04-17 18:32:37.545125\n",
      "\n",
      "Testing # 1 5\n",
      "2023-04-17 18:32:43.280955\n",
      "2023-04-17 18:32:43.280986\n",
      "PREDICT\n",
      "2023-04-17 18:32:43.284177\n",
      "debug pred [0.28, 0.7, 0.57]\n",
      "debug class [0 1 1]\n",
      "Test 1:0 540:333 61.86%\n",
      "Predict 1:0 712:161 81.56%\n",
      "Test sizes (873, 1025) (873,)\n",
      "Distrib of scores: 0.5880183276059564 mean 0.09146680308891413 std\n",
      "Range of scores: 0.25 to 0.91\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[ 95 238]\n",
      " [ 66 474]]\n",
      "Normalized matrix\n",
      " [[0.10882016 0.27262314]\n",
      " [0.07560137 0.54295533]]\n",
      "Extreme scores correct, pos:neg 161 7\n",
      "Extreme scores incorrect pos:neg 68 0\n",
      "Accuracy: 65.18% Precision: 66.57% Recall: 87.78%\n",
      "F1: 75.72% MCC: 0.2042\n",
      "AUPRC: 69.82% AUROC: 62.11%\n",
      " accuracy [65.17754868270332]\n",
      " precision [66.57303370786516]\n",
      " recall [87.77777777777777]\n",
      " F1 [75.7188498402556]\n",
      " MCC [0.20423099721334534]\n",
      " AUPRC [69.82109468824706]\n",
      " AUROC [62.11294627961295]\n",
      "\n",
      "Training # 2 1\n",
      "2023-04-17 18:32:43.379709\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.2.1.validation_genes.txt\n",
      "2023-04-17 18:32:43.394075\n",
      "Discriminator is GMM\n",
      "Means [-0.11372879952619426, -2.7425220412195848]\n",
      "Variances [1.4819813 2.0539251]\n",
      "Priors [0.56613822 0.43386178]\n",
      "[-5, -4, -3.5, -3, -2.5, -2, -1.5, -1, -0.5, 0, 0.5, 1, 1.5, 2, 3]\n",
      "[0 0 0 0 0 0 1 1 1 1 1 1 1 1 1]\n",
      "Train 1:0 2086:1410 59.67%\n",
      "Valid 1:0 503:372 57.49%\n",
      "FIT\n",
      "2023-04-17 18:32:44.658104\n",
      "\n",
      "Testing # 2 1\n",
      "2023-04-17 18:32:50.198336\n",
      "2023-04-17 18:32:50.198377\n",
      "PREDICT\n",
      "2023-04-17 18:32:50.202041\n",
      "debug pred [0.55, 0.57, 0.43]\n",
      "debug class [1 1 0]\n",
      "Test 1:0 503:372 57.49%\n",
      "Predict 1:0 696:179 79.54%\n",
      "Test sizes (875, 1025) (875,)\n",
      "Distrib of scores: 0.58048 mean 0.09136175129669966 std\n",
      "Range of scores: 0.19 to 0.84\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[113 259]\n",
      " [ 66 437]]\n",
      "Normalized matrix\n",
      " [[0.12914286 0.296     ]\n",
      " [0.07542857 0.49942857]]\n",
      "Extreme scores correct, pos:neg 143 6\n",
      "Extreme scores incorrect pos:neg 62 0\n",
      "Accuracy: 62.86% Precision: 62.79% Recall: 86.88%\n",
      "F1: 72.89% MCC: 0.2115\n",
      "AUPRC: 68.69% AUROC: 64.08%\n",
      " accuracy [62.857142857142854]\n",
      " precision [62.787356321839084]\n",
      " recall [86.87872763419483]\n",
      " F1 [72.89407839866556]\n",
      " MCC [0.2114662656143013]\n",
      " AUPRC [68.68562004757442]\n",
      " AUROC [64.08324248060028]\n",
      "\n",
      "Training # 2 2\n",
      "2023-04-17 18:32:50.294488\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.2.2.validation_genes.txt\n",
      "2023-04-17 18:32:50.309334\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Discriminator is GMM\n",
      "Means [-2.7902173367548957, -0.15669726413134297]\n",
      "Variances [2.02828573 1.47347722]\n",
      "Priors [0.42643664 0.57356336]\n",
      "[-5, -4, -3.5, -3, -2.5, -2, -1.5, -1, -0.5, 0, 0.5, 1, 1.5, 2, 3]\n",
      "[0 0 0 0 0 0 1 1 1 1 1 1 1 1 1]\n",
      "Train 1:0 2104:1392 60.18%\n",
      "Valid 1:0 524:351 59.89%\n",
      "FIT\n",
      "2023-04-17 18:32:51.630078\n",
      "\n",
      "Testing # 2 2\n",
      "2023-04-17 18:32:57.352071\n",
      "2023-04-17 18:32:57.352130\n",
      "PREDICT\n",
      "2023-04-17 18:32:57.355334\n",
      "debug pred [0.78, 0.64, 0.45]\n",
      "debug class [1 1 0]\n",
      "Test 1:0 524:351 59.89%\n",
      "Predict 1:0 730:145 83.43%\n",
      "Test sizes (875, 1025) (875,)\n",
      "Distrib of scores: 0.6004571428571429 mean 0.09504174807710042 std\n",
      "Range of scores: 0.23 to 0.93\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[100 251]\n",
      " [ 45 479]]\n",
      "Normalized matrix\n",
      " [[0.11428571 0.28685714]\n",
      " [0.05142857 0.54742857]]\n",
      "Extreme scores correct, pos:neg 206 5\n",
      "Extreme scores incorrect pos:neg 71 1\n",
      "Accuracy: 66.17% Precision: 65.62% Recall: 91.41%\n",
      "F1: 76.40% MCC: 0.2623\n",
      "AUPRC: 72.89% AUROC: 67.63%\n",
      " accuracy [66.17142857142856]\n",
      " precision [65.61643835616438]\n",
      " recall [91.41221374045801]\n",
      " F1 [76.39553429027113]\n",
      " MCC [0.26234680426989065]\n",
      " AUPRC [72.88524876553087]\n",
      " AUROC [67.63472956221048]\n",
      "\n",
      "Training # 2 3\n",
      "2023-04-17 18:32:57.445592\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.2.3.validation_genes.txt\n",
      "2023-04-17 18:32:57.460368\n",
      "Discriminator is GMM\n",
      "Means [-2.7899005057965844, -0.14175559866600776]\n",
      "Variances [2.01263245 1.46799439]\n",
      "Priors [0.43538125 0.56461875]\n",
      "[-5, -4, -3.5, -3, -2.5, -2, -1.5, -1, -0.5, 0, 0.5, 1, 1.5, 2, 3]\n",
      "[0 0 0 0 0 0 1 1 1 1 1 1 1 1 1]\n",
      "Train 1:0 2072:1425 59.25%\n",
      "Valid 1:0 533:341 60.98%\n",
      "FIT\n",
      "2023-04-17 18:32:58.733778\n",
      "\n",
      "Testing # 2 3\n",
      "2023-04-17 18:33:05.255709\n",
      "2023-04-17 18:33:05.255738\n",
      "PREDICT\n",
      "2023-04-17 18:33:05.259292\n",
      "debug pred [0.55, 0.73, 0.79]\n",
      "debug class [1 1 1]\n",
      "Test 1:0 533:341 60.98%\n",
      "Predict 1:0 690:184 78.95%\n",
      "Test sizes (874, 1025) (874,)\n",
      "Distrib of scores: 0.5837757437070938 mean 0.10161476611741417 std\n",
      "Range of scores: 0.24 to 0.96\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[103 238]\n",
      " [ 81 452]]\n",
      "Normalized matrix\n",
      " [[0.11784897 0.27231121]\n",
      " [0.09267735 0.51716247]]\n",
      "Extreme scores correct, pos:neg 166 4\n",
      "Extreme scores incorrect pos:neg 65 3\n",
      "Accuracy: 63.50% Precision: 65.51% Recall: 84.80%\n",
      "F1: 73.92% MCC: 0.1796\n",
      "AUPRC: 71.78% AUROC: 64.37%\n",
      " accuracy [63.50114416475973]\n",
      " precision [65.5072463768116]\n",
      " recall [84.80300187617262]\n",
      " F1 [73.91659852820932]\n",
      " MCC [0.17957185713876297]\n",
      " AUPRC [71.77865772725896]\n",
      " AUROC [64.36867617040708]\n",
      "\n",
      "Training # 2 4\n",
      "2023-04-17 18:33:05.352634\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.2.4.validation_genes.txt\n",
      "2023-04-17 18:33:05.366583\n",
      "Discriminator is GMM\n",
      "Means [-2.7775134423966796, -0.10928572591265324]\n",
      "Variances [2.08879181 1.51554389]\n",
      "Priors [0.42693146 0.57306854]\n",
      "[-5, -4, -3.5, -3, -2.5, -2, -1.5, -1, -0.5, 0, 0.5, 1, 1.5, 2, 3]\n",
      "[0 0 0 0 0 0 1 1 1 1 1 1 1 1 1]\n",
      "Train 1:0 2103:1395 60.12%\n",
      "Valid 1:0 506:367 57.96%\n",
      "FIT\n",
      "2023-04-17 18:33:06.764169\n",
      "\n",
      "Testing # 2 4\n",
      "2023-04-17 18:33:13.024632\n",
      "2023-04-17 18:33:13.024672\n",
      "PREDICT\n",
      "2023-04-17 18:33:13.027718\n",
      "debug pred [0.52, 0.62, 0.7]\n",
      "debug class [1 1 1]\n",
      "Test 1:0 506:367 57.96%\n",
      "Predict 1:0 706:167 80.87%\n",
      "Test sizes (873, 1025) (873,)\n",
      "Distrib of scores: 0.5863230240549828 mean 0.10003605825969966 std\n",
      "Range of scores: 0.15 to 0.94\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[103 264]\n",
      " [ 64 442]]\n",
      "Normalized matrix\n",
      " [[0.11798396 0.3024055 ]\n",
      " [0.07331042 0.50630011]]\n",
      "Extreme scores correct, pos:neg 155 6\n",
      "Extreme scores incorrect pos:neg 66 3\n",
      "Accuracy: 62.43% Precision: 62.61% Recall: 87.35%\n",
      "F1: 72.94% MCC: 0.1935\n",
      "AUPRC: 68.59% AUROC: 64.05%\n",
      " accuracy [62.42840778923253]\n",
      " precision [62.606232294617556]\n",
      " recall [87.35177865612648]\n",
      " F1 [72.93729372937293]\n",
      " MCC [0.19348730313591153]\n",
      " AUPRC [68.58787002290349]\n",
      " AUROC [64.04642922531798]\n",
      "\n",
      "Training # 2 5\n",
      "2023-04-17 18:33:13.170341\n",
      "Opening file /Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/cv.2.5.validation_genes.txt\n",
      "2023-04-17 18:33:13.186009\n",
      "Discriminator is GMM\n",
      "Means [-2.7357969593256457, -0.12175395649636288]\n",
      "Variances [1.99856448 1.47543451]\n",
      "Priors [0.441858 0.558142]\n",
      "[-5, -4, -3.5, -3, -2.5, -2, -1.5, -1, -0.5, 0, 0.5, 1, 1.5, 2, 3]\n",
      "[0 0 0 0 0 0 1 1 1 1 1 1 1 1 1]\n",
      "Train 1:0 2022:1475 57.82%\n",
      "Valid 1:0 533:341 60.98%\n",
      "FIT\n",
      "2023-04-17 18:33:14.495838\n",
      "\n",
      "Testing # 2 5\n",
      "2023-04-17 18:33:20.435260\n",
      "2023-04-17 18:33:20.435289\n",
      "PREDICT\n",
      "2023-04-17 18:33:20.438445\n",
      "debug pred [0.6, 0.63, 0.76]\n",
      "debug class [1 1 1]\n",
      "Test 1:0 533:341 60.98%\n",
      "Predict 1:0 678:196 77.57%\n",
      "Test sizes (874, 1025) (874,)\n",
      "Distrib of scores: 0.5731006864988558 mean 0.09149719043583955 std\n",
      "Range of scores: 0.27 to 0.94\n",
      "Score threshold 0.5\n",
      "Confusion matrix\n",
      " [[120 221]\n",
      " [ 76 457]]\n",
      "Normalized matrix\n",
      " [[0.13729977 0.25286041]\n",
      " [0.08695652 0.5228833 ]]\n",
      "Extreme scores correct, pos:neg 146 2\n",
      "Extreme scores incorrect pos:neg 45 3\n",
      "Accuracy: 66.02% Precision: 67.40% Recall: 85.74%\n",
      "F1: 75.47% MCC: 0.2448\n",
      "AUPRC: 74.00% AUROC: 66.47%\n",
      " accuracy [66.01830663615561]\n",
      " precision [67.40412979351032]\n",
      " recall [85.74108818011257]\n",
      " F1 [75.47481420313791]\n",
      " MCC [0.2447949737289193]\n",
      " AUPRC [74.00451544657345]\n",
      " AUROC [66.47290553663488]\n",
      "2023-04-17 18:33:20.530377\n"
     ]
    }
   ],
   "source": [
    "print(datetime.now())\n",
    "\n",
    "accuracy=[]\n",
    "precision=[]\n",
    "recall=[]\n",
    "f1=[]\n",
    "mcc=[]\n",
    "auprc=[]\n",
    "auroc=[]\n",
    "\n",
    "loader = DataLoader()\n",
    "filepath = DATA_DIR+RCI_FILE_TRAIN\n",
    "print(\"Load RCI from\",filepath)\n",
    "gene_to_rci = loader.load_gene_rci_values(filepath,EXCLUSIONS)\n",
    "print('Num RCI:', len(gene_to_rci.keys()))\n",
    "filepath = DATA_DIR+SEQ_FILE_TRAIN\n",
    "print('Load sequence from',filepath)\n",
    "allids,allseq = loader.load_spectra(filepath)  # for MLP (load_sequence() for CNN)\n",
    "print('Loaded',len(allseq),'sequences.')\n",
    "test_gene_to_rci = None\n",
    "test_allids = None\n",
    "test_allseq = None\n",
    "if SEQ_FILE_TEST is not None:\n",
    "    # Train on the entire train set (no cross-validation).\n",
    "    # Evaluate with the test files.\n",
    "    test_loader = DataLoader()\n",
    "    filepath = DATA_DIR+RCI_FILE_TEST\n",
    "    print(\"Load RCI from\",filepath)\n",
    "    test_gene_to_rci = test_loader.load_gene_rci_values(filepath,EXCLUSIONS)\n",
    "    print('Num RCI:', len(test_gene_to_rci.keys()))\n",
    "    filepath = DATA_DIR+SEQ_FILE_TEST\n",
    "    print('Load sequence from',filepath)\n",
    "    test_allids,test_allseq = test_loader.load_spectra(filepath)\n",
    "    print('Loaded',len(test_allseq),'sequences.')\n",
    "\n",
    "for repeat in range(REPEATS):\n",
    "    for fold in range(FOLDS):\n",
    "        show_r = repeat+1  # display one-based counting\n",
    "        show_f = fold+1    # display one-based counting\n",
    "\n",
    "        print()\n",
    "        print(\"Training #\",show_r,show_f)\n",
    "        print(datetime.now())\n",
    "        cvdo = CrossValidator(EPOCHS)\n",
    "        cvdo.set_threshold_mechanism(RCI_THRESHOLD_MECHANISM)\n",
    "        if SEQ_FILE_TEST is None:\n",
    "            # Train on 80% and evaluate on 20%.\n",
    "            separator = Separator()\n",
    "            separator.load(DATA_DIR,show_r,show_f)\n",
    "            separator.process(allids,allseq,gene_to_rci)\n",
    "            train_allids,test_allids = separator.get_ids()\n",
    "            train_allseq,test_allseq = separator.get_seq()\n",
    "            train_gene_to_rci,test_gene_to_rci = separator.get_rci()\n",
    "            cvdo.train_new_model(\n",
    "                train_allids,train_allseq,train_gene_to_rci,\n",
    "                test_allids,test_allseq,test_gene_to_rci)\n",
    "            if SAVE_MODEL_FILENAME is not None:\n",
    "                filename = f\"{SAVE_MODEL_FILENAME}.{show_r}.{show_f}.model\"\n",
    "                cvdo.save_model(filename)\n",
    "        else:\n",
    "            # Train on the entire train set (no cross-validation).\n",
    "            # Evaluate with the test files.\n",
    "            train_allids = allids\n",
    "            train_allseq = allseq\n",
    "            train_gene_to_rci = gene_to_rci\n",
    "            BREAK = True\n",
    "            cvdo.train_new_model(\n",
    "                train_allids,train_allseq,train_gene_to_rci,\n",
    "                None,None,None)\n",
    "\n",
    "        print()\n",
    "        print(\"Testing #\",show_r,show_f)\n",
    "        print(datetime.now())\n",
    "        cvdo.reset_statistics()\n",
    "        cvdo.test_without_training(\n",
    "            test_allids,test_allseq,test_gene_to_rci)\n",
    "        cv_accuracy,cv_precision,cv_recall,cv_f1,cv_mcc,cv_auprc,cv_auroc=\\\n",
    "            cvdo.get_statistics()\n",
    "\n",
    "        print(\" accuracy\" ,  cv_accuracy)\n",
    "        print(\" precision\" , cv_precision)\n",
    "        print(\" recall\" ,    cv_recall)\n",
    "        print(\" F1\" ,        cv_f1)\n",
    "        print(\" MCC\" ,       cv_mcc)\n",
    "        print(\" AUPRC\" ,     cv_auprc)\n",
    "        print(\" AUROC\" ,     cv_auroc)\n",
    "\n",
    "        accuracy.append(cv_accuracy)\n",
    "        precision.append(cv_precision)\n",
    "        recall.append(cv_recall)\n",
    "        f1.append(cv_f1)\n",
    "        mcc.append(cv_mcc)\n",
    "        auprc.append(cv_auprc)\n",
    "        auroc.append(cv_auroc)\n",
    "        if BREAK: break\n",
    "    if BREAK: break\n",
    "print(datetime.now())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "HkCeDg_HdQ36",
    "outputId": "b8486985-4447-4c8b-dde8-bbda95f9cc6d"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " accuracy mean 64.09% std 1.51\n",
      " precision mean 64.67% std 1.80\n",
      " recall mean 87.64% std 3.10\n",
      " F1 mean 74.38% std 1.36\n",
      " MCC mean 0.2150 std 0.0334\n",
      " AUPRC mean 70.82% std 2.61\n",
      " AUROC mean 64.87% std 2.07\n",
      " accuracy [[64.34285714285714], [65.25714285714285], [62.013729977116704], [63.1578947368421], [65.17754868270332], [62.857142857142854], [66.17142857142856], [63.50114416475973], [62.42840778923253], [66.01830663615561]]\n",
      " precision [[63.24324324324324], [64.72222222222223], [62.361111111111114], [65.86102719033232], [66.57303370786516], [62.787356321839084], [65.61643835616438], [65.5072463768116], [62.606232294617556], [67.40412979351032]]\n",
      " recall [[92.1259842519685], [90.31007751937985], [88.0392156862745], [81.95488721804512], [87.77777777777777], [86.87872763419483], [91.41221374045801], [84.80300187617262], [87.35177865612648], [85.74108818011257]]\n",
      " F1 [[75.00000000000001], [75.40453074433657], [73.00813008130082], [73.03182579564488], [75.7188498402556], [72.89407839866556], [76.39553429027113], [73.91659852820932], [72.93729372937293], [75.47481420313791]]\n",
      " MCC [[0.24605548996989562], [0.2519788918294678], [0.17582717682293555], [0.1807304768810816], [0.20423099721334534], [0.2114662656143013], [0.26234680426989065], [0.17957185713876297], [0.19348730313591153], [0.2447949737289193]]\n",
      " AUPRC [[69.54985943991282], [74.50263885679924], [66.44407027191536], [71.95253109891995], [69.82109468824706], [68.68562004757442], [72.88524876553087], [71.77865772725896], [68.58787002290349], [74.00451544657345]]\n",
      " AUROC [[66.23586646355854], [67.603809030252], [61.775748761042884], [64.39508859868971], [62.11294627961295], [64.08324248060028], [67.63472956221048], [64.36867617040708], [64.04642922531798], [66.47290553663488]]\n"
     ]
    }
   ],
   "source": [
    "def STD (values):\n",
    "    # ddof=1 reduces bias when extrapolating from sample to population\n",
    "    return np.std(values,ddof=1)\n",
    "\n",
    "print(\" accuracy mean %.2f%% std %.2f\" %  (np.mean(accuracy),  STD(accuracy)))\n",
    "print(\" precision mean %.2f%% std %.2f\" % (np.mean(precision), STD(precision)))\n",
    "print(\" recall mean %.2f%% std %.2f\" %    (np.mean(recall),    STD(recall)))\n",
    "print(\" F1 mean %.2f%% std %.2f\" %        (np.mean(f1),        STD(f1)))\n",
    "print(\" MCC mean %.4f std %.4f\" %       (np.mean(mcc),       STD(mcc)))\n",
    "print(\" AUPRC mean %.2f%% std %.2f\" %     (np.mean(auprc),     STD(auprc)))\n",
    "print(\" AUROC mean %.2f%% std %.2f\" %     (np.mean(auroc),     STD(auroc)))\n",
    "\n",
    "print(\" accuracy\"  , accuracy)\n",
    "print(\" precision\" , precision)\n",
    "print(\" recall\"    , recall)\n",
    "print(\" F1\"        , f1)\n",
    "print(\" MCC\"       , mcc)\n",
    "print(\" AUPRC\"     , auprc)\n",
    "print(\" AUROC\"     , auroc)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "QjSVa72v4IsA"
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "gpuClass": "standard",
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
