{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "PG-tGRnlFLA3"
      },
      "source": [
        "# Support Vector Machine\n",
        "canonical lncRNA, -1 threshold, cross-valiation, all the data"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 35,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "0RmwUsVLFLA6",
        "outputId": "d4a8f05f-af80-4210-a96e-0c6f930dfcbe"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2023-04-18 13:34:11.481982\n"
          ]
        }
      ],
      "source": [
        "from datetime import datetime\n",
        "print(datetime.now())"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 36,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "OlzN9OdsFWEU",
        "outputId": "b754d583-8472-4122-b596-ba8378df26f9"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "GPU device not found\n",
            "Running on CoLab\n",
            "Drive already mounted at /content/drive/; to attempt to forcibly remount, call drive.mount(\"/content/drive/\", force_remount=True).\n",
            "/content/drive/My Drive/data/Localization/TrainTest/TrainTest_ver43/\n"
          ]
        }
      ],
      "source": [
        "import tensorflow as tf\n",
        "dt='float32'\n",
        "tf.keras.backend.set_floatx('float32')\n",
        "# tf.random.set_seed(42) # supposedly leads to reproducible results\n",
        "\n",
        "device_name = tf.test.gpu_device_name()\n",
        "if device_name != '/device:GPU:0':\n",
        "    print('GPU device not found')\n",
        "else:\n",
        "    print('Found GPU at: {}'.format(device_name))\n",
        "\n",
        "try:\n",
        "    from google.colab import drive\n",
        "    IN_COLAB = True\n",
        "    print('Running on CoLab')\n",
        "    PATH='/content/drive/'\n",
        "    drive.mount(PATH)\n",
        "    DATA_DIR=PATH+'My Drive/data/Localization/TrainTest/TrainTest_ver43/'  # must end in \"/\"\n",
        "    MODEL_DIR=PATH+'My Drive/data/Localization/Models/'  # must end in \"/\"\n",
        "except:\n",
        "    IN_COLAB = False\n",
        "    DATA_DIR = 'D:/Adjeroh/Localization/TrainTest/'   # Windows\n",
        "    DATA_DIR = '/Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/'    # Mac\n",
        "    MODEL_DIR = '/Users/jasonmiller/WVU/Localization/Models/'    # Mac\n",
        "print(DATA_DIR)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 37,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "PRX-UEr8FLA8",
        "outputId": "e5dd9d39-4b7f-436d-8595-773853466dfd"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Python 3.9.16\n",
            "sklearn 1.2.2\n"
          ]
        }
      ],
      "source": [
        "from platform import python_version\n",
        "print('Python',python_version())\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "import scipy.stats as ss\n",
        "import pickle\n",
        "import time # sleep function\n",
        "from os.path import isfile\n",
        "from matplotlib import pyplot as plt \n",
        "import sklearn   # pip install --upgrade scikit-learn\n",
        "print('sklearn',sklearn.__version__)\n",
        "\n",
        "from tensorflow import keras\n",
        "from sklearn.mixture import GaussianMixture\n",
        "from sklearn.model_selection import KFold\n",
        "from sklearn.metrics import accuracy_score\n",
        "from sklearn.metrics import precision_score\n",
        "from sklearn.metrics import recall_score\n",
        "from sklearn.metrics import f1_score\n",
        "from sklearn.metrics import precision_recall_curve\n",
        "from sklearn.metrics import roc_curve\n",
        "from sklearn.metrics import roc_auc_score\n",
        "from sklearn.metrics import auc\n",
        "from sklearn.metrics import matthews_corrcoef\n",
        "from sklearn.utils import class_weight\n",
        "from sklearn.metrics import confusion_matrix\n",
        "from sklearn.svm import SVC\n",
        "\n",
        "from keras.models import Sequential\n",
        "from keras.layers import Masking\n",
        "from keras.layers import Embedding\n",
        "from keras.layers import Conv1D\n",
        "from keras.layers import GlobalAveragePooling1D\n",
        "from keras.layers import GlobalMaxPooling1D\n",
        "from keras.layers import AveragePooling1D\n",
        "from keras.layers import MaxPooling1D\n",
        "from keras.layers import Bidirectional\n",
        "from keras.layers import LSTM\n",
        "from keras.layers import TimeDistributed\n",
        "from keras.layers import Flatten\n",
        "from keras.layers import Dense\n",
        "from keras.layers import Dropout\n",
        "from keras.losses import BinaryCrossentropy\n",
        "#from keras.losses import Hinge\n",
        "\n",
        "K=5\n",
        "ALPHABET=4**K + 1\n",
        "CFILTERS=64 \n",
        "FILTERSIZE=8\n",
        "RCELLS=32\n",
        "DCELLS=16\n",
        "EPOCHS=150 \n",
        "EMBED_DIMEN = 4 # arbitrary hyperparameter\n",
        "# MINLEN=200   # NO LONGER USED\n",
        "# MAXLEN=5000\n",
        "RCI_THRESHOLD_MECHANISM = 'ZERO'  # 'RCI_GMM' 'ZERO' 'THE_MEAN'\n",
        "BREAK = False   # break after first fold\n",
        "EXCLUSIONS = [1]   # possibly exclude cell line 1 = H1.hESC\n",
        "FILTER_TRAIN        = False\n",
        "FILTER_TAILS_TRAIN  = False\n",
        "FILTER_MIDDLE_TRAIN = False\n",
        "FILTER_TEST         = False\n",
        "FILTER_TAILS_TEST   = False\n",
        "FILTER_MIDDLE_TEST  = False\n",
        "REPEATS = 2\n",
        "FOLDS = 5\n",
        "\n",
        "SAVE_MODEL_FILENAME = None # 'RF_101'"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 38,
      "metadata": {
        "id": "LnkpVKdMFLA-"
      },
      "outputs": [],
      "source": [
        "RCI_FILE_TRAIN = 'train.lncRNA_RCI.csv'\n",
        "RCI_FILE_TEST  = None #'test.lncRNA_RCI.csv'\n",
        "\n",
        "SEQ_FILE_TRAIN = 'train.canon_lncRNA_transcripts.csv'\n",
        "SEQ_FILE_TEST  = None #'test.canon_lncRNA_transcripts.csv'"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 39,
      "metadata": {
        "id": "e3p4QzQJFLA_"
      },
      "outputs": [],
      "source": [
        "def get_ordered_list():\n",
        "    ordered_list = \\\n",
        "    ['A549','H1.hESC','HeLa.S3','HepG2','HT1080',\\\n",
        "      'HUVEC','MCF.7','NCI.H460','NHEK','SK.MEL.5',\\\n",
        "      'SK.N.DZ','SK.N.SH','GM12878','K562','IMR.90']\n",
        "    return ordered_list\n",
        "all_cell_lines = get_ordered_list()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "UtqdpJOxFLBA"
      },
      "source": [
        "## Data Load"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 40,
      "metadata": {
        "id": "p35ehKV3Kq0z"
      },
      "outputs": [],
      "source": [
        "class DataLoader():\n",
        "    def __init__(self):\n",
        "        self.cache=dict() \n",
        "        self.vals = {'A':0, 'C':1, 'G':2, 'T':3}\n",
        "        self.gene2rci = dict()\n",
        "        \n",
        "    def load_gene_rci_values(self,filepath,exclusions):\n",
        "        '''\n",
        "        Load all the genes from the given RCI csv file.\n",
        "        The given file usually contains train or test, not both.\n",
        "        Return dict with keys=gene:str and values=RCI:float.\n",
        "        '''\n",
        "        self.gene2rci = {}\n",
        "        overall_sum = 0\n",
        "        with open (filepath,'r') as handle:\n",
        "            header = None\n",
        "            for row in handle:\n",
        "                if header is None:\n",
        "                    header = row # skip file's header line\n",
        "                else:\n",
        "                    line = row.strip()\n",
        "                    fields = line.split(',')\n",
        "                    gene_id = fields.pop(0)\n",
        "                    log_mean = float(fields.pop(0))\n",
        "                    self.gene2rci[gene_id] = log_mean\n",
        "        print('Number of RCI values loaded',len(self.gene2rci.keys()))\n",
        "        return self.gene2rci\n",
        "\n",
        "    def _seq_to_kmer_values(self,rna,K):\n",
        "        vec=[] # seq converted to list of K-mers \n",
        "        N_indicator = 0 # indicator value\n",
        "        length = len(rna)\n",
        "        for i in range(length-K+1):\n",
        "            kmer = rna[i:i+K]\n",
        "            if 'N' in kmer:\n",
        "                value = N_indicator\n",
        "            elif kmer in self.cache.keys():\n",
        "                value = self.cache[kmer]\n",
        "            else:\n",
        "                value = 0\n",
        "                for j in range(K):\n",
        "                    value *= 4   \n",
        "                    nextnuc = kmer[j] \n",
        "                    nucval = self.vals[nextnuc]\n",
        "                    value += nucval\n",
        "                value += 1   # NNN => 0, AAA => 1\n",
        "                self.cache[kmer] = value\n",
        "            vec.append(value)\n",
        "        return vec\n",
        "\n",
        "    def load_sequence(self,filepath):\n",
        "        '''\n",
        "        Load all the sequences from the given file. \n",
        "        Load our version of GenCode -- csv, easier to parse than fasta.\n",
        "        Each line has IDs plus sequence.\n",
        "        The IDs in the file do not include dot-version numbers.\n",
        "        The file may already be filtered e.g. canonical transcripts.\n",
        "        '''\n",
        "        allids=[]\n",
        "        allseq=[]\n",
        "        #NREPEAT = str('N'*MAXLEN)   # not used for MLP\n",
        "        with open (filepath,'r') as handle:\n",
        "            header = None\n",
        "            for row in handle:\n",
        "                if header is None:\n",
        "                    header = row\n",
        "                else:\n",
        "                    line    = row.strip()\n",
        "                    fields  = line.split(',')\n",
        "                    tran_id = fields[0]  # without version number\n",
        "                    gene_id = fields[1]  # without version number\n",
        "                    seq_len = int(fields[3])\n",
        "                    seq_txt = fields[4]\n",
        "                    # Keep only transcripts having numeric RCI given the cell lines in use.\n",
        "                    # We have validated this by spot checking.\n",
        "                    # TO DO: validate this programmatically.\n",
        "                    if gene_id in self.gene2rci.keys():\n",
        "                        # no MAXLEN for MLP\n",
        "                        #if seq_len<=MAXLEN:\n",
        "                        #    seq_txt = seq_txt + NREPEAT\n",
        "                        #seq_txt = seq_txt[:MAXLEN]\n",
        "                        allids.append( (gene_id,tran_id) )\n",
        "                        hot_vec = self._seq_to_kmer_values(seq_txt,K)\n",
        "                        allseq.append(hot_vec)\n",
        "        self.cache=dict() # save RAM\n",
        "        return allids,allseq\n",
        "\n",
        "    def load_spectra(self,filepath):\n",
        "        '''\n",
        "        Load all (variable-length) sequences as lists of kmers.\n",
        "        Then convert each sequence to (fixed-length) kmer histograms.\n",
        "        '''\n",
        "        allids,allseq = self.load_sequence(filepath)\n",
        "        allspectra = []\n",
        "        for seq in allseq:\n",
        "            spectrum = np.zeros(ALPHABET)\n",
        "            for kmer in seq:\n",
        "                spectrum[kmer] += 1\n",
        "            spectrum /= len(seq)\n",
        "            allspectra.append(spectrum)\n",
        "        return allids,allspectra        \n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "VDZ6siB_Kq04"
      },
      "source": [
        "## Model"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 41,
      "metadata": {
        "id": "AwMbRjm0FLBF"
      },
      "outputs": [],
      "source": [
        "def build_model():\n",
        "    '''\n",
        "    Probability is not the default with SVM.\n",
        "    We use probablity to compute the ARPRC statistic.\n",
        "    The sklearn documenation says:\n",
        "    Whether to enable probability estimates. \n",
        "    This must be enabled prior to calling fit, \n",
        "    will slow down that method as it internally uses 5-fold cross-validation, \n",
        "    and predict_proba may be inconsistent with predict.\n",
        "    '''\n",
        "    svc = SVC(probability=True)\n",
        "    return svc"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 42,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "clj-wufgFLBF",
        "outputId": "ea354f6b-ca88-454e-a8cd-233a6264863a"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2023-04-18 13:34:12.974331\n",
            "SVC(probability=True)\n"
          ]
        }
      ],
      "source": [
        "print(datetime.now())\n",
        "model=build_model()\n",
        "print(model)  # Print this only once\n",
        "model=None"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "MgrC1alOKq07"
      },
      "source": [
        "## Training"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 43,
      "metadata": {
        "id": "W9xiFzNbFLBE"
      },
      "outputs": [],
      "source": [
        "import gc\n",
        "class CrossValidator():\n",
        "    def __init__(self,epochs,score_threshold=0.5):\n",
        "        self.epochs = epochs\n",
        "        self.score_threshold = score_threshold\n",
        "        self.mechanism = 'ZERO'\n",
        "        self.discriminator = -1 # or zero\n",
        "        self.flip = False\n",
        "        self.reset_statistics()\n",
        "        \n",
        "    def reset_statistics(self):\n",
        "        self.cv_accuracy=[]\n",
        "        self.cv_precision=[]\n",
        "        self.cv_recall=[]\n",
        "        self.cv_f1=[]\n",
        "        self.cv_auprc=[]\n",
        "        self.cv_auroc=[]\n",
        "        self.cv_mcc=[]\n",
        "        \n",
        "    def _get_X_y(self, all_ids, all_seqs, rci_map): \n",
        "        # Prepare X and y for training or testing.\n",
        "        subsetX=[]\n",
        "        subsetY=[]\n",
        "        for t in range(len(all_ids)):\n",
        "            gene_id,tran_id = all_ids[t]\n",
        "            oneX            = all_seqs[t]\n",
        "            oneY            = rci_map[gene_id]\n",
        "            subsetX.append(oneX)\n",
        "            subsetY.append(oneY)\n",
        "        subsetX = np.array(subsetX)\n",
        "        subsetY = np.array(subsetY).reshape((-1,1))\n",
        "        return subsetX,subsetY\n",
        "    \n",
        "    def set_threshold_mechanism(self, mechanism):\n",
        "        if mechanism not in ['RCI_GMM','THE_MEAN','ZERO']:\n",
        "            raise Exception('Unrecognized mechansm:',mechanism)\n",
        "        self.mechanism = mechanism\n",
        "    \n",
        "    def _apply_threshold(self, array_of_rci):\n",
        "        # Takes list of float, returns list of labels [0,1].\n",
        "        if self.mechanism == 'RCI_GMM':\n",
        "            labels = self.discriminator.predict(array_of_rci)\n",
        "            if self.flip:\n",
        "                IS_CYTO = lambda label: 1 if label==0 else 0\n",
        "                labels = np.array(list(map(IS_CYTO, labels)))\n",
        "        else:  # 'THE_MEAN' or 'ZERO'\n",
        "            rci_threshold = self.discriminator\n",
        "            IS_CYTO = lambda rci: 1 if rci>rci_threshold else 0\n",
        "            labels = np.array(list(map(IS_CYTO, array_of_rci)))\n",
        "        return labels\n",
        "    \n",
        "    def _prepare_threshold(self, rci_values, create=True):\n",
        "        if self.mechanism == 'RCI_GMM':\n",
        "            if create:  # during training, create a new GMM\n",
        "                gmm = GaussianMixture(n_components=2, verbose=0, \n",
        "                  covariance_type='spherical', n_init=100) # random_state=42) \n",
        "                gmm.fit(rci_values)\n",
        "            else:   # during testing, use existing GMM\n",
        "                gmm=self.discriminator\n",
        "            self.flip = False\n",
        "            # The GMM labels are arbitrary.\n",
        "            if gmm.means_[0][0] > gmm.means_[1][0]:\n",
        "                self.flip = True\n",
        "            self.discriminator = gmm   # redundant but consistent\n",
        "        elif self.mechanism == 'THE_MEAN':\n",
        "            self.discriminator = np.mean(rci_values)\n",
        "        elif self.mechanism == 'ZERO':\n",
        "            self.discriminator = -1   # 0 usually, -1 is as in Yuan et al.\n",
        "        else: # not expected\n",
        "            self.discriminator = 0\n",
        "    \n",
        "    def _explain_threshold(self):\n",
        "        if self.mechanism == 'RCI_GMM':\n",
        "            gmm=self.discriminator\n",
        "            print('Discriminator is GMM')\n",
        "            print('Means',[gmm.means_[0][0],gmm.means_[1][0]])\n",
        "            print('Variances',gmm.covariances_)\n",
        "            print('Priors',gmm.weights_)\n",
        "            test_rcis=[-5,-4,-3.5,-3,-2.5,-2,-1.5,-1,-0.5,0,0.5,1,1.5,2,3]\n",
        "            print(test_rcis)\n",
        "            print(self._apply_threshold(np.array(test_rcis).reshape((-1,1))))\n",
        "        else:\n",
        "            print('Discriminator',self.mechanism,self.discriminator)\n",
        "    \n",
        "    def _show_sizes(self,label,values):\n",
        "        a = np.count_nonzero(values==1)\n",
        "        b = np.count_nonzero(values==0)\n",
        "        print('%s 1:0 %d:%d %5.2f%%'%(label,a,b,100*a/(a+b)))\n",
        "        \n",
        "    def save_model(self,filename):\n",
        "        if self.model is not None:\n",
        "            filepath = MODEL_DIR + filename\n",
        "            #self.model.save(filepath)\n",
        "            print('? Saved model to',filepath)\n",
        "        \n",
        "    def load_model(self,filename):\n",
        "        filepath = MODEL_DIR + filename\n",
        "        #self.model = keras.models.load_model(filepath)\n",
        "        print('? Loaded model from',filepath)\n",
        "        \n",
        "    def train_new_model(self,train_ids,train_seq,train_rci,\n",
        "            valid_ids=None,valid_seq=None,valid_rci=None):\n",
        "        print(datetime.now())\n",
        "        X_train,y_rci = self._get_X_y(train_ids,train_seq,train_rci) \n",
        "        self._prepare_threshold(y_rci,True)  \n",
        "        self._explain_threshold()\n",
        "        y_train = self._apply_threshold(y_rci)\n",
        "        self._show_sizes('Train',y_train)\n",
        "        #cw = class_weight.compute_class_weight('balanced', classes=[0,1], y=y_train)\n",
        "        #cwd = {0: cw[0], 1: cw[1]}\n",
        "        #print('Computed class weights:',cwd)\n",
        "        if valid_ids is not None:\n",
        "            X_valid,y_rci = self._get_X_y(valid_ids,valid_seq,valid_rci) \n",
        "            y_valid = self._apply_threshold(y_rci)\n",
        "            self._show_sizes('Valid',y_valid)\n",
        "        y_rci = None\n",
        "\n",
        "        self.model=build_model()\n",
        "        \n",
        "        print(\"FIT\")\n",
        "        print(datetime.now())\n",
        "        self.model.fit(X_train, y_train) # sample weight\n",
        "\n",
        "    def test_without_training(self,test_ids,test_seq,test_rci):\n",
        "        # For final test, do no train.\n",
        "        # Assume set_sequences() set the test set.\n",
        "        print(datetime.now())\n",
        "        X_test,y_rci = self._get_X_y(test_ids,test_seq,test_rci) \n",
        "        y_test = self._apply_threshold(y_rci)\n",
        "        y_rci = None\n",
        "        \n",
        "        print(\"PREDICT\")\n",
        "        print(datetime.now())        \n",
        "        yhat_pairs=self.model.predict_proba(X_test)  # [ prob of 0, prob of 1 ]\n",
        "        yhat_pred=[pair[1] for pair in yhat_pairs]\n",
        "        yhat_classes=self.model.predict(X_test)  # 0 or 1\n",
        "        \n",
        "        print('debug pred',yhat_pred[:3])\n",
        "        print('debug class',yhat_classes[:3])\n",
        "\n",
        "        self._show_sizes('Test',y_test)\n",
        "        self._show_sizes('Predict',yhat_classes)\n",
        "        print('Test sizes',X_test.shape,y_test.shape)\n",
        "        print('Distrib of scores:',np.mean(yhat_pred),'mean',np.std(yhat_pred),'std')\n",
        "        print('Range of scores:',np.min(yhat_pred),'to',np.max(yhat_pred))\n",
        "        print('Score threshold',self.score_threshold)\n",
        "        cm1 = confusion_matrix(y_test,yhat_classes)\n",
        "        print('Confusion matrix\\n',cm1)\n",
        "        cm2 = confusion_matrix(y_test,yhat_classes,normalize='all')\n",
        "        print('Normalized matrix\\n',cm2)\n",
        "\n",
        "        accuracy = accuracy_score(y_test, yhat_classes)*100.\n",
        "        precision = precision_score(y_test, yhat_classes)*100.\n",
        "        recall = recall_score(y_test, yhat_classes)*100.\n",
        "        f1 = f1_score(y_test, yhat_classes)*100.\n",
        "        prc_Y, prc_X, prc_bins = precision_recall_curve(y_test, yhat_pred)\n",
        "        auprc = auc(prc_X,prc_Y)*100.\n",
        "        auroc = roc_auc_score(y_test, yhat_pred)*100.\n",
        "        mcc = matthews_corrcoef(y_test, yhat_classes)\n",
        "\n",
        "        correct_pos = 0\n",
        "        correct_neg = 0\n",
        "        wrong_pos = 0\n",
        "        wrong_neg = 0\n",
        "        for i in range(len(y_test)):\n",
        "            if yhat_pred[i]>=0.65:\n",
        "                if y_test[i]==1:\n",
        "                    correct_pos += 1\n",
        "                else:\n",
        "                    wrong_pos += 1\n",
        "            elif yhat_pred[i]<=0.35:\n",
        "                if y_test[i]==0:\n",
        "                    correct_neg += 1\n",
        "                else:\n",
        "                    wrong_neg += 1\n",
        "        print('Extreme scores correct, pos:neg',correct_pos,correct_neg)  \n",
        "        print('Extreme scores incorrect pos:neg',wrong_pos,wrong_neg)  \n",
        "\n",
        "        #self._pretty_pictures(y_test,yhat_pred,prc_X,prc_Y)\n",
        "        \n",
        "        print('Accuracy: %.2f%% Precision: %.2f%% Recall: %.2f%%' % (accuracy,precision,recall)) \n",
        "        print('F1: %.2f%% MCC: %.4f' % (f1,mcc)) \n",
        "        print('AUPRC: %.2f%% AUROC: %.2f%%' % (auprc,auroc)) \n",
        "\n",
        "        self.cv_accuracy.append(accuracy)\n",
        "        self.cv_precision.append(precision)\n",
        "        self.cv_recall.append(recall)\n",
        "        self.cv_f1.append(f1)\n",
        "        self.cv_mcc.append(mcc)\n",
        "        self.cv_auprc.append(auprc)\n",
        "        self.cv_auroc.append(auroc)\n",
        "\n",
        "    def _pretty_pictures(self,y_valid,yhat_pred,prc_X,prc_Y):\n",
        "        count_ones= len(y_valid[y_valid==1])\n",
        "        count_zeros= len(y_valid[y_valid==0])\n",
        "        guess = max(count_ones,count_zeros) / len(y_valid)\n",
        "        # PRC\n",
        "        plt.plot(prc_X, prc_Y, marker='.')\n",
        "        plt.plot([0, 1], [guess,guess], linestyle='--')\n",
        "        plt.xlabel('Recall')\n",
        "        plt.ylabel('Precision')\n",
        "        plt.show()\n",
        "        # ROC\n",
        "        fpr, tpr, roc_bins = roc_curve(y_valid, yhat_pred)\n",
        "        plt.plot(fpr, tpr, marker='.')\n",
        "        plt.xlabel('False Positive Rate')\n",
        "        plt.ylabel('True Positive Rate')\n",
        "        plt.show()\n",
        "        \n",
        "    def get_statistics(self):\n",
        "        return \\\n",
        "        self.cv_accuracy,\\\n",
        "        self.cv_precision,\\\n",
        "        self.cv_recall,\\\n",
        "        self.cv_f1,\\\n",
        "        self.cv_mcc,\\\n",
        "        self.cv_auprc,\\\n",
        "        self.cv_auroc\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Td97uyyj5qDq"
      },
      "source": [
        "## Train"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 44,
      "metadata": {
        "id": "poe3rwAF4Ir7"
      },
      "outputs": [],
      "source": [
        "class Separator():\n",
        "    def __init__(self):\n",
        "        self.train_ids = []\n",
        "        self.train_seq = []\n",
        "        self.train_rci = dict()\n",
        "        self.val_ids = []\n",
        "        self.val_seq = []\n",
        "        self.val_rci = dict()\n",
        "    def load(self,data_dir,rep,fold):\n",
        "        filename='cv.{}.{}.validation_genes.txt'.format(rep,fold)\n",
        "        filename = data_dir + filename\n",
        "        self.val_genes = set()\n",
        "        print('Opening file',filename)\n",
        "        with open(filename,'r') as fin:\n",
        "            for line in fin:\n",
        "                gene_id = line.strip()\n",
        "                self.val_genes.add(gene_id)\n",
        "    def process(self,allids,allseq,gene_to_rci):\n",
        "        size = len(allids)\n",
        "        for t in range(size):\n",
        "            gene_id,tran_id = allids[t]\n",
        "            oneX            = allseq[t]\n",
        "            oneY            = gene_to_rci[gene_id]\n",
        "            in_middle = gene_to_rci[gene_id] >= -2 and gene_to_rci[gene_id] <= 0\n",
        "            in_tails = gene_to_rci[gene_id] < -2 or gene_to_rci[gene_id] > 0\n",
        "            if gene_id in self.val_genes:\n",
        "                if FILTER_TEST and (\\\n",
        "                    (FILTER_TAILS_TEST and in_tails) or \\\n",
        "                    (FILTER_MIDDLE_TEST and in_middle)):\n",
        "                    pass\n",
        "                else:\n",
        "                    self.val_ids.append(allids[t])\n",
        "                    self.val_seq.append(allseq[t])\n",
        "                    self.val_rci[gene_id]=gene_to_rci[gene_id]\n",
        "            else:\n",
        "                if FILTER_TRAIN and (\\\n",
        "                    (FILTER_TAILS_TRAIN and in_tails) or \\\n",
        "                    (FILTER_MIDDLE_TRAIN and in_middle)):\n",
        "                    pass\n",
        "                else:\n",
        "                    self.train_ids.append(allids[t])\n",
        "                    self.train_seq.append(allseq[t])\n",
        "                    self.train_rci[gene_id]=gene_to_rci[gene_id]\n",
        "    def get_ids(self):\n",
        "        return self.train_ids,self.val_ids\n",
        "    def get_seq(self):\n",
        "        return self.train_seq,self.val_seq\n",
        "    def get_rci(self):\n",
        "        return self.train_rci,self.val_rci"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 45,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "XC9m0W-pFLBH",
        "outputId": "2f66cf14-ce54-4128-883d-3ddd313d2d98",
        "scrolled": false
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2023-04-18 13:34:13.081787\n",
            "Load RCI from /content/drive/My Drive/data/Localization/TrainTest/TrainTest_ver43/train.lncRNA_RCI.csv\n",
            "Number of RCI values loaded 4371\n",
            "Num RCI: 4371\n",
            "Load sequence from /content/drive/My Drive/data/Localization/TrainTest/TrainTest_ver43/train.canon_lncRNA_transcripts.csv\n",
            "Loaded 4371 sequences.\n",
            "\n",
            "Training # 1 1\n",
            "2023-04-18 13:34:22.426235\n",
            "Opening file /content/drive/My Drive/data/Localization/TrainTest/TrainTest_ver43/cv.1.1.validation_genes.txt\n",
            "2023-04-18 13:34:22.437597\n",
            "Discriminator ZERO -1\n",
            "Train 1:0 1659:1837 47.45%\n",
            "Valid 1:0 402:473 45.94%\n",
            "FIT\n",
            "2023-04-18 13:34:22.462487\n",
            "\n",
            "Testing # 1 1\n",
            "2023-04-18 13:35:15.983824\n",
            "2023-04-18 13:35:15.984240\n",
            "PREDICT\n",
            "2023-04-18 13:35:15.993249\n",
            "debug pred [0.5510112759434617, 0.34690888753798743, 0.31154063459540865]\n",
            "debug class [1 0 0]\n",
            "Test 1:0 402:473 45.94%\n",
            "Predict 1:0 379:496 43.31%\n",
            "Test sizes (875, 1025) (875,)\n",
            "Distrib of scores: 0.4839059502195316 mean 0.13261193909708255 std\n",
            "Range of scores: 0.14209271607107005 to 0.8605794048614055\n",
            "Score threshold 0.5\n",
            "Confusion matrix\n",
            " [[322 151]\n",
            " [174 228]]\n",
            "Normalized matrix\n",
            " [[0.368      0.17257143]\n",
            " [0.19885714 0.26057143]]\n",
            "Extreme scores correct, pos:neg 72 111\n",
            "Extreme scores incorrect pos:neg 31 35\n",
            "Accuracy: 62.86% Precision: 60.16% Recall: 56.72%\n",
            "F1: 58.39% MCC: 0.2493\n",
            "AUPRC: 60.74% AUROC: 66.74%\n",
            " accuracy [62.857142857142854]\n",
            " precision [60.15831134564644]\n",
            " recall [56.71641791044776]\n",
            " F1 [58.38668373879643]\n",
            " MCC [0.24934689434881505]\n",
            " AUPRC [60.742126215874656]\n",
            " AUROC [66.7366129184942]\n",
            "\n",
            "Training # 1 2\n",
            "2023-04-18 13:35:22.982686\n",
            "Opening file /content/drive/My Drive/data/Localization/TrainTest/TrainTest_ver43/cv.1.2.validation_genes.txt\n",
            "2023-04-18 13:35:22.999156\n",
            "Discriminator ZERO -1\n",
            "Train 1:0 1643:1853 47.00%\n",
            "Valid 1:0 418:457 47.77%\n",
            "FIT\n",
            "2023-04-18 13:35:23.026393\n",
            "\n",
            "Testing # 1 2\n",
            "2023-04-18 13:36:15.376092\n",
            "2023-04-18 13:36:15.376165\n",
            "PREDICT\n",
            "2023-04-18 13:36:15.383382\n",
            "debug pred [0.39328266874615037, 0.3152039627458052, 0.682451983853544]\n",
            "debug class [0 0 1]\n",
            "Test 1:0 418:457 47.77%\n",
            "Predict 1:0 367:508 41.94%\n",
            "Test sizes (875, 1025) (875,)\n",
            "Distrib of scores: 0.47043512141794186 mean 0.13630688974273758 std\n",
            "Range of scores: 0.09211522089392801 to 0.8280361482919905\n",
            "Score threshold 0.5\n",
            "Confusion matrix\n",
            " [[310 147]\n",
            " [198 220]]\n",
            "Normalized matrix\n",
            " [[0.35428571 0.168     ]\n",
            " [0.22628571 0.25142857]]\n",
            "Extreme scores correct, pos:neg 61 129\n",
            "Extreme scores incorrect pos:neg 26 44\n",
            "Accuracy: 60.57% Precision: 59.95% Recall: 52.63%\n",
            "F1: 56.05% MCC: 0.2072\n",
            "AUPRC: 61.35% AUROC: 64.57%\n",
            " accuracy [60.57142857142858]\n",
            " precision [59.945504087193456]\n",
            " recall [52.63157894736842]\n",
            " F1 [56.05095541401274]\n",
            " MCC [0.20715669777440116]\n",
            " AUPRC [61.35212370793991]\n",
            " AUROC [64.57340885533905]\n",
            "\n",
            "Training # 1 3\n",
            "2023-04-18 13:36:22.417302\n",
            "Opening file /content/drive/My Drive/data/Localization/TrainTest/TrainTest_ver43/cv.1.3.validation_genes.txt\n",
            "2023-04-18 13:36:22.436263\n",
            "Discriminator ZERO -1\n",
            "Train 1:0 1671:1826 47.78%\n",
            "Valid 1:0 390:484 44.62%\n",
            "FIT\n",
            "2023-04-18 13:36:22.474725\n",
            "\n",
            "Testing # 1 3\n",
            "2023-04-18 13:37:13.943685\n",
            "2023-04-18 13:37:13.943906\n",
            "PREDICT\n",
            "2023-04-18 13:37:13.950380\n",
            "debug pred [0.5385012923825232, 0.5615515696929416, 0.3072172995618429]\n",
            "debug class [1 1 0]\n",
            "Test 1:0 390:484 44.62%\n",
            "Predict 1:0 362:512 41.42%\n",
            "Test sizes (874, 1025) (874,)\n",
            "Distrib of scores: 0.4710077257177034 mean 0.13378311427839948 std\n",
            "Range of scores: 0.09670309731246599 to 0.8774499490923737\n",
            "Score threshold 0.5\n",
            "Confusion matrix\n",
            " [[327 157]\n",
            " [185 205]]\n",
            "Normalized matrix\n",
            " [[0.37414188 0.17963387]\n",
            " [0.21167048 0.23455378]]\n",
            "Extreme scores correct, pos:neg 53 124\n",
            "Extreme scores incorrect pos:neg 32 40\n",
            "Accuracy: 60.87% Precision: 56.63% Recall: 52.56%\n",
            "F1: 54.52% MCC: 0.2031\n",
            "AUPRC: 57.42% AUROC: 65.46%\n",
            " accuracy [60.86956521739131]\n",
            " precision [56.62983425414365]\n",
            " recall [52.56410256410257]\n",
            " F1 [54.52127659574468]\n",
            " MCC [0.20310707109522025]\n",
            " AUPRC [57.420244255591435]\n",
            " AUROC [65.45984318711592]\n",
            "\n",
            "Training # 1 4\n",
            "2023-04-18 13:37:21.578145\n",
            "Opening file /content/drive/My Drive/data/Localization/TrainTest/TrainTest_ver43/cv.1.4.validation_genes.txt\n",
            "2023-04-18 13:37:21.596982\n",
            "Discriminator ZERO -1\n",
            "Train 1:0 1637:1860 46.81%\n",
            "Valid 1:0 424:450 48.51%\n",
            "FIT\n",
            "2023-04-18 13:37:21.634755\n",
            "\n",
            "Testing # 1 4\n",
            "2023-04-18 13:38:12.643955\n",
            "2023-04-18 13:38:12.644008\n",
            "PREDICT\n",
            "2023-04-18 13:38:12.651010\n",
            "debug pred [0.429914370322482, 0.43171374979053656, 0.6579061766855253]\n",
            "debug class [0 0 1]\n",
            "Test 1:0 424:450 48.51%\n",
            "Predict 1:0 350:524 40.05%\n",
            "Test sizes (874, 1025) (874,)\n",
            "Distrib of scores: 0.46825329615123307 mean 0.13468110256142193 std\n",
            "Range of scores: 0.10325363639847226 to 0.8782589328037946\n",
            "Score threshold 0.5\n",
            "Confusion matrix\n",
            " [[315 135]\n",
            " [209 215]]\n",
            "Normalized matrix\n",
            " [[0.3604119  0.15446224]\n",
            " [0.23913043 0.24599542]]\n",
            "Extreme scores correct, pos:neg 58 122\n",
            "Extreme scores incorrect pos:neg 20 51\n",
            "Accuracy: 60.64% Precision: 61.43% Recall: 50.71%\n",
            "F1: 55.56% MCC: 0.2112\n",
            "AUPRC: 60.49% AUROC: 64.50%\n",
            " accuracy [60.64073226544623]\n",
            " precision [61.42857142857143]\n",
            " recall [50.70754716981132]\n",
            " F1 [55.55555555555556]\n",
            " MCC [0.21121180052880498]\n",
            " AUPRC [60.48610729670993]\n",
            " AUROC [64.50052410901468]\n",
            "\n",
            "Training # 1 5\n",
            "2023-04-18 13:38:20.463593\n",
            "Opening file /content/drive/My Drive/data/Localization/TrainTest/TrainTest_ver43/cv.1.5.validation_genes.txt\n",
            "2023-04-18 13:38:20.473444\n",
            "Discriminator ZERO -1\n",
            "Train 1:0 1634:1864 46.71%\n",
            "Valid 1:0 427:446 48.91%\n",
            "FIT\n",
            "2023-04-18 13:38:20.502709\n",
            "\n",
            "Testing # 1 5\n",
            "2023-04-18 13:39:09.788383\n",
            "2023-04-18 13:39:09.788444\n",
            "PREDICT\n",
            "2023-04-18 13:39:09.796447\n",
            "debug pred [0.32203701859119055, 0.47564085630691216, 0.31001333012615784]\n",
            "debug class [0 0 0]\n",
            "Test 1:0 427:446 48.91%\n",
            "Predict 1:0 331:542 37.92%\n",
            "Test sizes (873, 1025) (873,)\n",
            "Distrib of scores: 0.4649000693870361 mean 0.13165897446173833 std\n",
            "Range of scores: 0.13435624977817948 to 0.8482494285670282\n",
            "Score threshold 0.5\n",
            "Confusion matrix\n",
            " [[322 124]\n",
            " [220 207]]\n",
            "Normalized matrix\n",
            " [[0.36884307 0.14203895]\n",
            " [0.25200458 0.2371134 ]]\n",
            "Extreme scores correct, pos:neg 59 122\n",
            "Extreme scores incorrect pos:neg 17 47\n",
            "Accuracy: 60.60% Precision: 62.54% Recall: 48.48%\n",
            "F1: 54.62% MCC: 0.2130\n",
            "AUPRC: 62.36% AUROC: 65.46%\n",
            " accuracy [60.595647193585336]\n",
            " precision [62.53776435045317]\n",
            " recall [48.47775175644028]\n",
            " F1 [54.61741424802111]\n",
            " MCC [0.21301713040868467]\n",
            " AUPRC [62.361772146575035]\n",
            " AUROC [65.45824975583118]\n",
            "\n",
            "Training # 2 1\n",
            "2023-04-18 13:39:17.640513\n",
            "Opening file /content/drive/My Drive/data/Localization/TrainTest/TrainTest_ver43/cv.2.1.validation_genes.txt\n",
            "2023-04-18 13:39:17.651990\n",
            "Discriminator ZERO -1\n",
            "Train 1:0 1655:1841 47.34%\n",
            "Valid 1:0 406:469 46.40%\n",
            "FIT\n",
            "2023-04-18 13:39:17.678992\n",
            "\n",
            "Testing # 2 1\n",
            "2023-04-18 13:40:07.432966\n",
            "2023-04-18 13:40:07.433366\n",
            "PREDICT\n",
            "2023-04-18 13:40:07.441761\n",
            "debug pred [0.5746166508231456, 0.49009567111187297, 0.32820994462569447]\n",
            "debug class [1 0 0]\n",
            "Test 1:0 406:469 46.40%\n",
            "Predict 1:0 342:533 39.09%\n",
            "Test sizes (875, 1025) (875,)\n",
            "Distrib of scores: 0.46805802718712847 mean 0.12780724300617666 std\n",
            "Range of scores: 0.10388258073698357 to 0.8211612032759313\n",
            "Score threshold 0.5\n",
            "Confusion matrix\n",
            " [[321 148]\n",
            " [212 194]]\n",
            "Normalized matrix\n",
            " [[0.36685714 0.16914286]\n",
            " [0.24228571 0.22171429]]\n",
            "Extreme scores correct, pos:neg 47 115\n",
            "Extreme scores incorrect pos:neg 23 46\n",
            "Accuracy: 58.86% Precision: 56.73% Recall: 47.78%\n",
            "F1: 51.87% MCC: 0.1658\n",
            "AUPRC: 58.08% AUROC: 63.86%\n",
            " accuracy [58.857142857142854]\n",
            " precision [56.72514619883041]\n",
            " recall [47.783251231527096]\n",
            " F1 [51.87165775401069]\n",
            " MCC [0.16584572192504507]\n",
            " AUPRC [58.07571833438998]\n",
            " AUROC [63.86111315344459]\n",
            "\n",
            "Training # 2 2\n",
            "2023-04-18 13:40:15.216103\n",
            "Opening file /content/drive/My Drive/data/Localization/TrainTest/TrainTest_ver43/cv.2.2.validation_genes.txt\n",
            "2023-04-18 13:40:15.229777\n",
            "Discriminator ZERO -1\n",
            "Train 1:0 1645:1851 47.05%\n",
            "Valid 1:0 416:459 47.54%\n",
            "FIT\n",
            "2023-04-18 13:40:15.260360\n",
            "\n",
            "Testing # 2 2\n",
            "2023-04-18 13:41:06.289307\n",
            "2023-04-18 13:41:06.289706\n",
            "PREDICT\n",
            "2023-04-18 13:41:06.298481\n",
            "debug pred [0.4892536695909835, 0.41904631946491216, 0.23715583884204655]\n",
            "debug class [0 0 0]\n",
            "Test 1:0 416:459 47.54%\n",
            "Predict 1:0 393:482 44.91%\n",
            "Test sizes (875, 1025) (875,)\n",
            "Distrib of scores: 0.48330309939059135 mean 0.13755194861759587 std\n",
            "Range of scores: 0.1081366746869374 to 0.8382291048589937\n",
            "Score threshold 0.5\n",
            "Confusion matrix\n",
            " [[310 149]\n",
            " [172 244]]\n",
            "Normalized matrix\n",
            " [[0.35428571 0.17028571]\n",
            " [0.19657143 0.27885714]]\n",
            "Extreme scores correct, pos:neg 82 114\n",
            "Extreme scores incorrect pos:neg 25 35\n",
            "Accuracy: 63.31% Precision: 62.09% Recall: 58.65%\n",
            "F1: 60.32% MCC: 0.2630\n",
            "AUPRC: 64.74% AUROC: 68.47%\n",
            " accuracy [63.31428571428571]\n",
            " precision [62.086513994910945]\n",
            " recall [58.65384615384615]\n",
            " F1 [60.32138442521631]\n",
            " MCC [0.2629671018882356]\n",
            " AUPRC [64.7418357725581]\n",
            " AUROC [68.47112242332831]\n",
            "\n",
            "Training # 2 3\n",
            "2023-04-18 13:41:13.419411\n",
            "Opening file /content/drive/My Drive/data/Localization/TrainTest/TrainTest_ver43/cv.2.3.validation_genes.txt\n",
            "2023-04-18 13:41:13.433959\n",
            "Discriminator ZERO -1\n",
            "Train 1:0 1634:1863 46.73%\n",
            "Valid 1:0 427:447 48.86%\n",
            "FIT\n",
            "2023-04-18 13:41:13.464087\n",
            "\n",
            "Testing # 2 3\n",
            "2023-04-18 13:42:04.476551\n",
            "2023-04-18 13:42:04.476630\n",
            "PREDICT\n",
            "2023-04-18 13:42:04.482770\n",
            "debug pred [0.4382321720759271, 0.5958285135343099, 0.684258870254376]\n",
            "debug class [0 1 1]\n",
            "Test 1:0 427:447 48.86%\n",
            "Predict 1:0 340:534 38.90%\n",
            "Test sizes (874, 1025) (874,)\n",
            "Distrib of scores: 0.46901823802186643 mean 0.14001785014896245 std\n",
            "Range of scores: 0.10816089419145156 to 0.858241651498708\n",
            "Score threshold 0.5\n",
            "Confusion matrix\n",
            " [[314 133]\n",
            " [220 207]]\n",
            "Normalized matrix\n",
            " [[0.35926773 0.15217391]\n",
            " [0.25171625 0.23684211]]\n",
            "Extreme scores correct, pos:neg 58 132\n",
            "Extreme scores incorrect pos:neg 30 47\n",
            "Accuracy: 59.61% Precision: 60.88% Recall: 48.48%\n",
            "F1: 53.98% MCC: 0.1920\n",
            "AUPRC: 62.09% AUROC: 65.29%\n",
            " accuracy [59.610983981693366]\n",
            " precision [60.882352941176464]\n",
            " recall [48.47775175644028]\n",
            " F1 [53.976531942633635]\n",
            " MCC [0.19197844003043096]\n",
            " AUPRC [62.09011947242835]\n",
            " AUROC [65.29242569511025]\n",
            "\n",
            "Training # 2 4\n",
            "2023-04-18 13:42:11.443684\n",
            "Opening file /content/drive/My Drive/data/Localization/TrainTest/TrainTest_ver43/cv.2.4.validation_genes.txt\n",
            "2023-04-18 13:42:11.459524\n",
            "Discriminator ZERO -1\n",
            "Train 1:0 1669:1829 47.71%\n",
            "Valid 1:0 392:481 44.90%\n",
            "FIT\n",
            "2023-04-18 13:42:11.488080\n",
            "\n",
            "Testing # 2 4\n",
            "2023-04-18 13:43:02.460947\n",
            "2023-04-18 13:43:02.461024\n",
            "PREDICT\n",
            "2023-04-18 13:43:02.469095\n",
            "debug pred [0.3704574267081894, 0.46654749230119713, 0.64531737353393]\n",
            "debug class [0 0 1]\n",
            "Test 1:0 392:481 44.90%\n",
            "Predict 1:0 350:523 40.09%\n",
            "Test sizes (873, 1025) (873,)\n",
            "Distrib of scores: 0.46858084152169166 mean 0.13876397015899908 std\n",
            "Range of scores: 0.10863820965752602 to 0.8745069735945462\n",
            "Score threshold 0.5\n",
            "Confusion matrix\n",
            " [[332 149]\n",
            " [191 201]]\n",
            "Normalized matrix\n",
            " [[0.38029782 0.17067583]\n",
            " [0.2187858  0.23024055]]\n",
            "Extreme scores correct, pos:neg 57 130\n",
            "Extreme scores incorrect pos:neg 31 43\n",
            "Accuracy: 61.05% Precision: 57.43% Recall: 51.28%\n",
            "F1: 54.18% MCC: 0.2060\n",
            "AUPRC: 56.32% AUROC: 64.55%\n",
            " accuracy [61.05383734249714]\n",
            " precision [57.42857142857143]\n",
            " recall [51.275510204081634]\n",
            " F1 [54.17789757412399]\n",
            " MCC [0.20601178594770397]\n",
            " AUPRC [56.31896108809661]\n",
            " AUROC [64.54718061860919]\n",
            "\n",
            "Training # 2 5\n",
            "2023-04-18 13:43:09.716139\n",
            "Opening file /content/drive/My Drive/data/Localization/TrainTest/TrainTest_ver43/cv.2.5.validation_genes.txt\n",
            "2023-04-18 13:43:09.738028\n",
            "Discriminator ZERO -1\n",
            "Train 1:0 1641:1856 46.93%\n",
            "Valid 1:0 420:454 48.05%\n",
            "FIT\n",
            "2023-04-18 13:43:09.785787\n",
            "\n",
            "Testing # 2 5\n",
            "2023-04-18 13:44:00.573975\n",
            "2023-04-18 13:44:00.574035\n",
            "PREDICT\n",
            "2023-04-18 13:44:00.580880\n",
            "debug pred [0.63936500644787, 0.5424669619341783, 0.5282967640092678]\n",
            "debug class [1 1 1]\n",
            "Test 1:0 420:454 48.05%\n",
            "Predict 1:0 357:517 40.85%\n",
            "Test sizes (874, 1025) (874,)\n",
            "Distrib of scores: 0.4749078280795988 mean 0.1254734570694403 std\n",
            "Range of scores: 0.18025574377477893 to 0.8147030657282672\n",
            "Score threshold 0.5\n",
            "Confusion matrix\n",
            " [[327 127]\n",
            " [190 230]]\n",
            "Normalized matrix\n",
            " [[0.37414188 0.14530892]\n",
            " [0.2173913  0.26315789]]\n",
            "Extreme scores correct, pos:neg 58 98\n",
            "Extreme scores incorrect pos:neg 30 40\n",
            "Accuracy: 63.73% Precision: 64.43% Recall: 54.76%\n",
            "F1: 59.20% MCC: 0.2723\n",
            "AUPRC: 62.97% AUROC: 66.07%\n",
            " accuracy [63.7299771167048]\n",
            " precision [64.42577030812325]\n",
            " recall [54.761904761904766]\n",
            " F1 [59.202059202059196]\n",
            " MCC [0.2722820033743004]\n",
            " AUPRC [62.973258365105124]\n",
            " AUROC [66.07300188797987]\n",
            "2023-04-18 13:44:08.492871\n"
          ]
        }
      ],
      "source": [
        "print(datetime.now())\n",
        "\n",
        "accuracy=[]\n",
        "precision=[]\n",
        "recall=[]\n",
        "f1=[]\n",
        "mcc=[]\n",
        "auprc=[]\n",
        "auroc=[]\n",
        "\n",
        "loader = DataLoader()\n",
        "filepath = DATA_DIR+RCI_FILE_TRAIN\n",
        "print(\"Load RCI from\",filepath)\n",
        "gene_to_rci = loader.load_gene_rci_values(filepath,EXCLUSIONS)\n",
        "print('Num RCI:', len(gene_to_rci.keys()))\n",
        "filepath = DATA_DIR+SEQ_FILE_TRAIN\n",
        "print('Load sequence from',filepath)\n",
        "allids,allseq = loader.load_spectra(filepath)  # for MLP (load_sequence() for CNN)\n",
        "print('Loaded',len(allseq),'sequences.')\n",
        "test_gene_to_rci = None\n",
        "test_allids = None\n",
        "test_allseq = None\n",
        "if SEQ_FILE_TEST is not None:\n",
        "    # Train on the entire train set (no cross-validation).\n",
        "    # Evaluate with the test files.\n",
        "    test_loader = DataLoader()\n",
        "    filepath = DATA_DIR+RCI_FILE_TEST\n",
        "    print(\"Load RCI from\",filepath)\n",
        "    test_gene_to_rci = test_loader.load_gene_rci_values(filepath,EXCLUSIONS)\n",
        "    print('Num RCI:', len(test_gene_to_rci.keys()))\n",
        "    filepath = DATA_DIR+SEQ_FILE_TEST\n",
        "    print('Load sequence from',filepath)\n",
        "    test_allids,test_allseq = test_loader.load_spectra(filepath)\n",
        "    print('Loaded',len(test_allseq),'sequences.')\n",
        "\n",
        "for repeat in range(REPEATS):\n",
        "    for fold in range(FOLDS):\n",
        "        show_r = repeat+1  # display one-based counting\n",
        "        show_f = fold+1    # display one-based counting\n",
        "\n",
        "        print()\n",
        "        print(\"Training #\",show_r,show_f)\n",
        "        print(datetime.now())\n",
        "        cvdo = CrossValidator(EPOCHS)\n",
        "        cvdo.set_threshold_mechanism(RCI_THRESHOLD_MECHANISM)\n",
        "        if SEQ_FILE_TEST is None:\n",
        "            # Train on 80% and evaluate on 20%.\n",
        "            separator = Separator()\n",
        "            separator.load(DATA_DIR,show_r,show_f)\n",
        "            separator.process(allids,allseq,gene_to_rci)\n",
        "            train_allids,test_allids = separator.get_ids()\n",
        "            train_allseq,test_allseq = separator.get_seq()\n",
        "            train_gene_to_rci,test_gene_to_rci = separator.get_rci()\n",
        "            cvdo.train_new_model(\n",
        "                train_allids,train_allseq,train_gene_to_rci,\n",
        "                test_allids,test_allseq,test_gene_to_rci)\n",
        "            if SAVE_MODEL_FILENAME is not None:\n",
        "                filename = f\"{SAVE_MODEL_FILENAME}.{show_r}.{show_f}.model\"\n",
        "                cvdo.save_model(filename)\n",
        "        else:\n",
        "            # Train on the entire train set (no cross-validation).\n",
        "            # Evaluate with the test files.\n",
        "            train_allids = allids\n",
        "            train_allseq = allseq\n",
        "            train_gene_to_rci = gene_to_rci\n",
        "            BREAK = True\n",
        "            cvdo.train_new_model(\n",
        "                train_allids,train_allseq,train_gene_to_rci,\n",
        "                None,None,None)\n",
        "\n",
        "        print()\n",
        "        print(\"Testing #\",show_r,show_f)\n",
        "        print(datetime.now())\n",
        "        cvdo.reset_statistics()\n",
        "        cvdo.test_without_training(\n",
        "            test_allids,test_allseq,test_gene_to_rci)\n",
        "        cv_accuracy,cv_precision,cv_recall,cv_f1,cv_mcc,cv_auprc,cv_auroc=\\\n",
        "            cvdo.get_statistics()\n",
        "\n",
        "        print(\" accuracy\" ,  cv_accuracy)\n",
        "        print(\" precision\" , cv_precision)\n",
        "        print(\" recall\" ,    cv_recall)\n",
        "        print(\" F1\" ,        cv_f1)\n",
        "        print(\" MCC\" ,       cv_mcc)\n",
        "        print(\" AUPRC\" ,     cv_auprc)\n",
        "        print(\" AUROC\" ,     cv_auroc)\n",
        "\n",
        "        accuracy.append(cv_accuracy)\n",
        "        precision.append(cv_precision)\n",
        "        recall.append(cv_recall)\n",
        "        f1.append(cv_f1)\n",
        "        mcc.append(cv_mcc)\n",
        "        auprc.append(cv_auprc)\n",
        "        auroc.append(cv_auroc)\n",
        "        if BREAK: break\n",
        "    if BREAK: break\n",
        "print(datetime.now())"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 46,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "HkCeDg_HdQ36",
        "outputId": "b56e7a08-a105-4c8e-e7b1-9fe8b3da62d0"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            " accuracy mean 61.21% std 1.59\n",
            " precision mean 60.22% std 2.61\n",
            " recall mean 52.20% std 3.64\n",
            " F1 mean 55.87% std 2.65\n",
            " MCC mean 0.2183 std 0.0332\n",
            " AUPRC mean 60.66% std 2.66\n",
            " AUROC mean 65.50% std 1.34\n",
            " accuracy [[62.857142857142854], [60.57142857142858], [60.86956521739131], [60.64073226544623], [60.595647193585336], [58.857142857142854], [63.31428571428571], [59.610983981693366], [61.05383734249714], [63.7299771167048]]\n",
            " precision [[60.15831134564644], [59.945504087193456], [56.62983425414365], [61.42857142857143], [62.53776435045317], [56.72514619883041], [62.086513994910945], [60.882352941176464], [57.42857142857143], [64.42577030812325]]\n",
            " recall [[56.71641791044776], [52.63157894736842], [52.56410256410257], [50.70754716981132], [48.47775175644028], [47.783251231527096], [58.65384615384615], [48.47775175644028], [51.275510204081634], [54.761904761904766]]\n",
            " F1 [[58.38668373879643], [56.05095541401274], [54.52127659574468], [55.55555555555556], [54.61741424802111], [51.87165775401069], [60.32138442521631], [53.976531942633635], [54.17789757412399], [59.202059202059196]]\n",
            " MCC [[0.24934689434881505], [0.20715669777440116], [0.20310707109522025], [0.21121180052880498], [0.21301713040868467], [0.16584572192504507], [0.2629671018882356], [0.19197844003043096], [0.20601178594770397], [0.2722820033743004]]\n",
            " AUPRC [[60.742126215874656], [61.35212370793991], [57.420244255591435], [60.48610729670993], [62.361772146575035], [58.07571833438998], [64.7418357725581], [62.09011947242835], [56.31896108809661], [62.973258365105124]]\n",
            " AUROC [[66.7366129184942], [64.57340885533905], [65.45984318711592], [64.50052410901468], [65.45824975583118], [63.86111315344459], [68.47112242332831], [65.29242569511025], [64.54718061860919], [66.07300188797987]]\n"
          ]
        }
      ],
      "source": [
        "def STD (values):\n",
        "    # ddof=1 reduces bias when extrapolating from sample to population\n",
        "    return np.std(values,ddof=1)\n",
        "\n",
        "print(\" accuracy mean %.2f%% std %.2f\" %  (np.mean(accuracy),  STD(accuracy)))\n",
        "print(\" precision mean %.2f%% std %.2f\" % (np.mean(precision), STD(precision)))\n",
        "print(\" recall mean %.2f%% std %.2f\" %    (np.mean(recall),    STD(recall)))\n",
        "print(\" F1 mean %.2f%% std %.2f\" %        (np.mean(f1),        STD(f1)))\n",
        "print(\" MCC mean %.4f std %.4f\" %       (np.mean(mcc),       STD(mcc)))\n",
        "print(\" AUPRC mean %.2f%% std %.2f\" %     (np.mean(auprc),     STD(auprc)))\n",
        "print(\" AUROC mean %.2f%% std %.2f\" %     (np.mean(auroc),     STD(auroc)))\n",
        "\n",
        "print(\" accuracy\"  , accuracy)\n",
        "print(\" precision\" , precision)\n",
        "print(\" recall\"    , recall)\n",
        "print(\" F1\"        , f1)\n",
        "print(\" MCC\"       , mcc)\n",
        "print(\" AUPRC\"     , auprc)\n",
        "print(\" AUROC\"     , auroc)\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 46,
      "metadata": {
        "id": "QjSVa72v4IsA"
      },
      "outputs": [],
      "source": []
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "gpuClass": "standard",
    "kernelspec": {
      "display_name": "Python 3 (ipykernel)",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.10.0"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}