{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "PG-tGRnlFLA3"
      },
      "source": [
        "# Support Vector Machine\n",
        "canonical lncRNA, -1 threshold, cross-valiation, all the data"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 33,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "0RmwUsVLFLA6",
        "outputId": "1beb7e04-197a-4e3c-83b7-d9996a1b3817"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2023-04-18 13:34:33.902945\n"
          ]
        }
      ],
      "source": [
        "from datetime import datetime\n",
        "print(datetime.now())"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 34,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "OlzN9OdsFWEU",
        "outputId": "fd421aff-2860-48fc-a672-734e5078493c"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "GPU device not found\n",
            "Running on CoLab\n",
            "Drive already mounted at /content/drive/; to attempt to forcibly remount, call drive.mount(\"/content/drive/\", force_remount=True).\n",
            "/content/drive/My Drive/data/Localization/TrainTest/TrainTest_ver43/\n"
          ]
        }
      ],
      "source": [
        "import tensorflow as tf\n",
        "dt='float32'\n",
        "tf.keras.backend.set_floatx('float32')\n",
        "# tf.random.set_seed(42) # supposedly leads to reproducible results\n",
        "\n",
        "device_name = tf.test.gpu_device_name()\n",
        "if device_name != '/device:GPU:0':\n",
        "    print('GPU device not found')\n",
        "else:\n",
        "    print('Found GPU at: {}'.format(device_name))\n",
        "\n",
        "try:\n",
        "    from google.colab import drive\n",
        "    IN_COLAB = True\n",
        "    print('Running on CoLab')\n",
        "    PATH='/content/drive/'\n",
        "    drive.mount(PATH)\n",
        "    DATA_DIR=PATH+'My Drive/data/Localization/TrainTest/TrainTest_ver43/'  # must end in \"/\"\n",
        "    MODEL_DIR=PATH+'My Drive/data/Localization/Models/'  # must end in \"/\"\n",
        "except:\n",
        "    IN_COLAB = False\n",
        "    DATA_DIR = 'D:/Adjeroh/Localization/TrainTest/'   # Windows\n",
        "    DATA_DIR = '/Users/jasonmiller/WVU/Localization/TrainTest/TrainTest_ver43/'    # Mac\n",
        "    MODEL_DIR = '/Users/jasonmiller/WVU/Localization/Models/'    # Mac\n",
        "print(DATA_DIR)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 35,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "PRX-UEr8FLA8",
        "outputId": "998ac085-550b-4243-aef0-a19d6362d2f7"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Python 3.9.16\n",
            "sklearn 1.2.2\n"
          ]
        }
      ],
      "source": [
        "from platform import python_version\n",
        "print('Python',python_version())\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "import scipy.stats as ss\n",
        "import pickle\n",
        "import time # sleep function\n",
        "from os.path import isfile\n",
        "from matplotlib import pyplot as plt \n",
        "import sklearn   # pip install --upgrade scikit-learn\n",
        "print('sklearn',sklearn.__version__)\n",
        "\n",
        "from tensorflow import keras\n",
        "from sklearn.mixture import GaussianMixture\n",
        "from sklearn.model_selection import KFold\n",
        "from sklearn.metrics import accuracy_score\n",
        "from sklearn.metrics import precision_score\n",
        "from sklearn.metrics import recall_score\n",
        "from sklearn.metrics import f1_score\n",
        "from sklearn.metrics import precision_recall_curve\n",
        "from sklearn.metrics import roc_curve\n",
        "from sklearn.metrics import roc_auc_score\n",
        "from sklearn.metrics import auc\n",
        "from sklearn.metrics import matthews_corrcoef\n",
        "from sklearn.utils import class_weight\n",
        "from sklearn.metrics import confusion_matrix\n",
        "from sklearn.svm import SVC\n",
        "\n",
        "from keras.models import Sequential\n",
        "from keras.layers import Masking\n",
        "from keras.layers import Embedding\n",
        "from keras.layers import Conv1D\n",
        "from keras.layers import GlobalAveragePooling1D\n",
        "from keras.layers import GlobalMaxPooling1D\n",
        "from keras.layers import AveragePooling1D\n",
        "from keras.layers import MaxPooling1D\n",
        "from keras.layers import Bidirectional\n",
        "from keras.layers import LSTM\n",
        "from keras.layers import TimeDistributed\n",
        "from keras.layers import Flatten\n",
        "from keras.layers import Dense\n",
        "from keras.layers import Dropout\n",
        "from keras.losses import BinaryCrossentropy\n",
        "#from keras.losses import Hinge\n",
        "\n",
        "K=5\n",
        "ALPHABET=4**K + 1\n",
        "CFILTERS=64 \n",
        "FILTERSIZE=8\n",
        "RCELLS=32\n",
        "DCELLS=16\n",
        "EPOCHS=150 \n",
        "EMBED_DIMEN = 4 # arbitrary hyperparameter\n",
        "# MINLEN=200   # NO LONGER USED\n",
        "# MAXLEN=5000\n",
        "RCI_THRESHOLD_MECHANISM = 'ZERO'  # 'RCI_GMM' 'ZERO' 'THE_MEAN'\n",
        "BREAK = False   # break after first fold\n",
        "EXCLUSIONS = [1]   # possibly exclude cell line 1 = H1.hESC\n",
        "FILTER_TRAIN        = True\n",
        "FILTER_TAILS_TRAIN  = False\n",
        "FILTER_MIDDLE_TRAIN = True\n",
        "FILTER_TEST         = True\n",
        "FILTER_TAILS_TEST   = False\n",
        "FILTER_MIDDLE_TEST  = True\n",
        "REPEATS = 2\n",
        "FOLDS = 5\n",
        "\n",
        "SAVE_MODEL_FILENAME = None # 'RF_101'"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 36,
      "metadata": {
        "id": "LnkpVKdMFLA-"
      },
      "outputs": [],
      "source": [
        "RCI_FILE_TRAIN = 'train.lncRNA_RCI.csv'\n",
        "RCI_FILE_TEST  = None #'test.lncRNA_RCI.csv'\n",
        "\n",
        "SEQ_FILE_TRAIN = 'train.canon_lncRNA_transcripts.csv'\n",
        "SEQ_FILE_TEST  = None #'test.canon_lncRNA_transcripts.csv'"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 37,
      "metadata": {
        "id": "e3p4QzQJFLA_"
      },
      "outputs": [],
      "source": [
        "def get_ordered_list():\n",
        "    ordered_list = \\\n",
        "    ['A549','H1.hESC','HeLa.S3','HepG2','HT1080',\\\n",
        "      'HUVEC','MCF.7','NCI.H460','NHEK','SK.MEL.5',\\\n",
        "      'SK.N.DZ','SK.N.SH','GM12878','K562','IMR.90']\n",
        "    return ordered_list\n",
        "all_cell_lines = get_ordered_list()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "UtqdpJOxFLBA"
      },
      "source": [
        "## Data Load"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 38,
      "metadata": {
        "id": "p35ehKV3Kq0z"
      },
      "outputs": [],
      "source": [
        "class DataLoader():\n",
        "    def __init__(self):\n",
        "        self.cache=dict() \n",
        "        self.vals = {'A':0, 'C':1, 'G':2, 'T':3}\n",
        "        self.gene2rci = dict()\n",
        "        \n",
        "    def load_gene_rci_values(self,filepath,exclusions):\n",
        "        '''\n",
        "        Load all the genes from the given RCI csv file.\n",
        "        The given file usually contains train or test, not both.\n",
        "        Return dict with keys=gene:str and values=RCI:float.\n",
        "        '''\n",
        "        self.gene2rci = {}\n",
        "        overall_sum = 0\n",
        "        with open (filepath,'r') as handle:\n",
        "            header = None\n",
        "            for row in handle:\n",
        "                if header is None:\n",
        "                    header = row # skip file's header line\n",
        "                else:\n",
        "                    line = row.strip()\n",
        "                    fields = line.split(',')\n",
        "                    gene_id = fields.pop(0)\n",
        "                    log_mean = float(fields.pop(0))\n",
        "                    self.gene2rci[gene_id] = log_mean\n",
        "        print('Number of RCI values loaded',len(self.gene2rci.keys()))\n",
        "        return self.gene2rci\n",
        "\n",
        "    def _seq_to_kmer_values(self,rna,K):\n",
        "        vec=[] # seq converted to list of K-mers \n",
        "        N_indicator = 0 # indicator value\n",
        "        length = len(rna)\n",
        "        for i in range(length-K+1):\n",
        "            kmer = rna[i:i+K]\n",
        "            if 'N' in kmer:\n",
        "                value = N_indicator\n",
        "            elif kmer in self.cache.keys():\n",
        "                value = self.cache[kmer]\n",
        "            else:\n",
        "                value = 0\n",
        "                for j in range(K):\n",
        "                    value *= 4   \n",
        "                    nextnuc = kmer[j] \n",
        "                    nucval = self.vals[nextnuc]\n",
        "                    value += nucval\n",
        "                value += 1   # NNN => 0, AAA => 1\n",
        "                self.cache[kmer] = value\n",
        "            vec.append(value)\n",
        "        return vec\n",
        "\n",
        "    def load_sequence(self,filepath):\n",
        "        '''\n",
        "        Load all the sequences from the given file. \n",
        "        Load our version of GenCode -- csv, easier to parse than fasta.\n",
        "        Each line has IDs plus sequence.\n",
        "        The IDs in the file do not include dot-version numbers.\n",
        "        The file may already be filtered e.g. canonical transcripts.\n",
        "        '''\n",
        "        allids=[]\n",
        "        allseq=[]\n",
        "        #NREPEAT = str('N'*MAXLEN)   # not used for MLP\n",
        "        with open (filepath,'r') as handle:\n",
        "            header = None\n",
        "            for row in handle:\n",
        "                if header is None:\n",
        "                    header = row\n",
        "                else:\n",
        "                    line    = row.strip()\n",
        "                    fields  = line.split(',')\n",
        "                    tran_id = fields[0]  # without version number\n",
        "                    gene_id = fields[1]  # without version number\n",
        "                    seq_len = int(fields[3])\n",
        "                    seq_txt = fields[4]\n",
        "                    # Keep only transcripts having numeric RCI given the cell lines in use.\n",
        "                    # We have validated this by spot checking.\n",
        "                    # TO DO: validate this programmatically.\n",
        "                    if gene_id in self.gene2rci.keys():\n",
        "                        # no MAXLEN for MLP\n",
        "                        #if seq_len<=MAXLEN:\n",
        "                        #    seq_txt = seq_txt + NREPEAT\n",
        "                        #seq_txt = seq_txt[:MAXLEN]\n",
        "                        allids.append( (gene_id,tran_id) )\n",
        "                        hot_vec = self._seq_to_kmer_values(seq_txt,K)\n",
        "                        allseq.append(hot_vec)\n",
        "        self.cache=dict() # save RAM\n",
        "        return allids,allseq\n",
        "\n",
        "    def load_spectra(self,filepath):\n",
        "        '''\n",
        "        Load all (variable-length) sequences as lists of kmers.\n",
        "        Then convert each sequence to (fixed-length) kmer histograms.\n",
        "        '''\n",
        "        allids,allseq = self.load_sequence(filepath)\n",
        "        allspectra = []\n",
        "        for seq in allseq:\n",
        "            spectrum = np.zeros(ALPHABET)\n",
        "            for kmer in seq:\n",
        "                spectrum[kmer] += 1\n",
        "            spectrum /= len(seq)\n",
        "            allspectra.append(spectrum)\n",
        "        return allids,allspectra        \n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "VDZ6siB_Kq04"
      },
      "source": [
        "## Model"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 39,
      "metadata": {
        "id": "AwMbRjm0FLBF"
      },
      "outputs": [],
      "source": [
        "def build_model():\n",
        "    '''\n",
        "    Probability is not the default with SVM.\n",
        "    We use probablity to compute the ARPRC statistic.\n",
        "    The sklearn documenation says:\n",
        "    Whether to enable probability estimates. \n",
        "    This must be enabled prior to calling fit, \n",
        "    will slow down that method as it internally uses 5-fold cross-validation, \n",
        "    and predict_proba may be inconsistent with predict.\n",
        "    '''    \n",
        "    svc = SVC(probability=True)\n",
        "    return svc"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 40,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "clj-wufgFLBF",
        "outputId": "85eb2963-59e1-4c36-ed4b-6589ecd08c5a"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2023-04-18 13:34:35.706510\n",
            "SVC(probability=True)\n"
          ]
        }
      ],
      "source": [
        "print(datetime.now())\n",
        "model=build_model()\n",
        "print(model)  # Print this only once\n",
        "model=None"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "MgrC1alOKq07"
      },
      "source": [
        "## Training"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 41,
      "metadata": {
        "id": "W9xiFzNbFLBE"
      },
      "outputs": [],
      "source": [
        "import gc\n",
        "class CrossValidator():\n",
        "    def __init__(self,epochs,score_threshold=0.5):\n",
        "        self.epochs = epochs\n",
        "        self.score_threshold = score_threshold\n",
        "        self.mechanism = 'ZERO'\n",
        "        self.discriminator = -1 # or zero\n",
        "        self.flip = False\n",
        "        self.reset_statistics()\n",
        "        \n",
        "    def reset_statistics(self):\n",
        "        self.cv_accuracy=[]\n",
        "        self.cv_precision=[]\n",
        "        self.cv_recall=[]\n",
        "        self.cv_f1=[]\n",
        "        self.cv_auprc=[]\n",
        "        self.cv_auroc=[]\n",
        "        self.cv_mcc=[]\n",
        "        \n",
        "    def _get_X_y(self, all_ids, all_seqs, rci_map): \n",
        "        # Prepare X and y for training or testing.\n",
        "        subsetX=[]\n",
        "        subsetY=[]\n",
        "        for t in range(len(all_ids)):\n",
        "            gene_id,tran_id = all_ids[t]\n",
        "            oneX            = all_seqs[t]\n",
        "            oneY            = rci_map[gene_id]\n",
        "            subsetX.append(oneX)\n",
        "            subsetY.append(oneY)\n",
        "        subsetX = np.array(subsetX)\n",
        "        subsetY = np.array(subsetY).reshape((-1,1))\n",
        "        return subsetX,subsetY\n",
        "    \n",
        "    def set_threshold_mechanism(self, mechanism):\n",
        "        if mechanism not in ['RCI_GMM','THE_MEAN','ZERO']:\n",
        "            raise Exception('Unrecognized mechansm:',mechanism)\n",
        "        self.mechanism = mechanism\n",
        "    \n",
        "    def _apply_threshold(self, array_of_rci):\n",
        "        # Takes list of float, returns list of labels [0,1].\n",
        "        if self.mechanism == 'RCI_GMM':\n",
        "            labels = self.discriminator.predict(array_of_rci)\n",
        "            if self.flip:\n",
        "                IS_CYTO = lambda label: 1 if label==0 else 0\n",
        "                labels = np.array(list(map(IS_CYTO, labels)))\n",
        "        else:  # 'THE_MEAN' or 'ZERO'\n",
        "            rci_threshold = self.discriminator\n",
        "            IS_CYTO = lambda rci: 1 if rci>rci_threshold else 0\n",
        "            labels = np.array(list(map(IS_CYTO, array_of_rci)))\n",
        "        return labels\n",
        "    \n",
        "    def _prepare_threshold(self, rci_values, create=True):\n",
        "        if self.mechanism == 'RCI_GMM':\n",
        "            if create:  # during training, create a new GMM\n",
        "                gmm = GaussianMixture(n_components=2, verbose=0, \n",
        "                  covariance_type='spherical', n_init=100) # random_state=42) \n",
        "                gmm.fit(rci_values)\n",
        "            else:   # during testing, use existing GMM\n",
        "                gmm=self.discriminator\n",
        "            self.flip = False\n",
        "            # The GMM labels are arbitrary.\n",
        "            if gmm.means_[0][0] > gmm.means_[1][0]:\n",
        "                self.flip = True\n",
        "            self.discriminator = gmm   # redundant but consistent\n",
        "        elif self.mechanism == 'THE_MEAN':\n",
        "            self.discriminator = np.mean(rci_values)\n",
        "        elif self.mechanism == 'ZERO':\n",
        "            self.discriminator = -1   # 0 usually, -1 is as in Yuan et al.\n",
        "        else: # not expected\n",
        "            self.discriminator = 0\n",
        "    \n",
        "    def _explain_threshold(self):\n",
        "        if self.mechanism == 'RCI_GMM':\n",
        "            gmm=self.discriminator\n",
        "            print('Discriminator is GMM')\n",
        "            print('Means',[gmm.means_[0][0],gmm.means_[1][0]])\n",
        "            print('Variances',gmm.covariances_)\n",
        "            print('Priors',gmm.weights_)\n",
        "            test_rcis=[-5,-4,-3.5,-3,-2.5,-2,-1.5,-1,-0.5,0,0.5,1,1.5,2,3]\n",
        "            print(test_rcis)\n",
        "            print(self._apply_threshold(np.array(test_rcis).reshape((-1,1))))\n",
        "        else:\n",
        "            print('Discriminator',self.mechanism,self.discriminator)\n",
        "    \n",
        "    def _show_sizes(self,label,values):\n",
        "        a = np.count_nonzero(values==1)\n",
        "        b = np.count_nonzero(values==0)\n",
        "        print('%s 1:0 %d:%d %5.2f%%'%(label,a,b,100*a/(a+b)))\n",
        "        \n",
        "    def save_model(self,filename):\n",
        "        if self.model is not None:\n",
        "            filepath = MODEL_DIR + filename\n",
        "            #self.model.save(filepath)\n",
        "            print('? Saved model to',filepath)\n",
        "        \n",
        "    def load_model(self,filename):\n",
        "        filepath = MODEL_DIR + filename\n",
        "        #self.model = keras.models.load_model(filepath)\n",
        "        print('? Loaded model from',filepath)\n",
        "        \n",
        "    def train_new_model(self,train_ids,train_seq,train_rci,\n",
        "            valid_ids=None,valid_seq=None,valid_rci=None):\n",
        "        print(datetime.now())\n",
        "        X_train,y_rci = self._get_X_y(train_ids,train_seq,train_rci) \n",
        "        self._prepare_threshold(y_rci,True)  \n",
        "        self._explain_threshold()\n",
        "        y_train = self._apply_threshold(y_rci)\n",
        "        self._show_sizes('Train',y_train)\n",
        "        #cw = class_weight.compute_class_weight('balanced', classes=[0,1], y=y_train)\n",
        "        #cwd = {0: cw[0], 1: cw[1]}\n",
        "        #print('Computed class weights:',cwd)\n",
        "        if valid_ids is not None:\n",
        "            X_valid,y_rci = self._get_X_y(valid_ids,valid_seq,valid_rci) \n",
        "            y_valid = self._apply_threshold(y_rci)\n",
        "            self._show_sizes('Valid',y_valid)\n",
        "        y_rci = None\n",
        "\n",
        "        self.model=build_model()\n",
        "        \n",
        "        print(\"FIT\")\n",
        "        print(datetime.now())\n",
        "        self.model.fit(X_train, y_train) # sample weight\n",
        "\n",
        "    def test_without_training(self,test_ids,test_seq,test_rci):\n",
        "        # For final test, do no train.\n",
        "        # Assume set_sequences() set the test set.\n",
        "        print(datetime.now())\n",
        "        X_test,y_rci = self._get_X_y(test_ids,test_seq,test_rci) \n",
        "        y_test = self._apply_threshold(y_rci)\n",
        "        y_rci = None\n",
        "        \n",
        "        print(\"PREDICT\")\n",
        "        print(datetime.now())        \n",
        "        yhat_pairs=self.model.predict_proba(X_test)  # [ prob of 0, prob of 1 ]\n",
        "        yhat_pred=[pair[1] for pair in yhat_pairs]\n",
        "        yhat_classes=self.model.predict(X_test)  # 0 or 1\n",
        "        \n",
        "        print('debug pred',yhat_pred[:3])\n",
        "        print('debug class',yhat_classes[:3])\n",
        "\n",
        "        self._show_sizes('Test',y_test)\n",
        "        self._show_sizes('Predict',yhat_classes)\n",
        "        print('Test sizes',X_test.shape,y_test.shape)\n",
        "        print('Distrib of scores:',np.mean(yhat_pred),'mean',np.std(yhat_pred),'std')\n",
        "        print('Range of scores:',np.min(yhat_pred),'to',np.max(yhat_pred))\n",
        "        print('Score threshold',self.score_threshold)\n",
        "        cm1 = confusion_matrix(y_test,yhat_classes)\n",
        "        print('Confusion matrix\\n',cm1)\n",
        "        cm2 = confusion_matrix(y_test,yhat_classes,normalize='all')\n",
        "        print('Normalized matrix\\n',cm2)\n",
        "\n",
        "        accuracy = accuracy_score(y_test, yhat_classes)*100.\n",
        "        precision = precision_score(y_test, yhat_classes)*100.\n",
        "        recall = recall_score(y_test, yhat_classes)*100.\n",
        "        f1 = f1_score(y_test, yhat_classes)*100.\n",
        "        prc_Y, prc_X, prc_bins = precision_recall_curve(y_test, yhat_pred)\n",
        "        auprc = auc(prc_X,prc_Y)*100.\n",
        "        auroc = roc_auc_score(y_test, yhat_pred)*100.\n",
        "        mcc = matthews_corrcoef(y_test, yhat_classes)\n",
        "\n",
        "        correct_pos = 0\n",
        "        correct_neg = 0\n",
        "        wrong_pos = 0\n",
        "        wrong_neg = 0\n",
        "        for i in range(len(y_test)):\n",
        "            if yhat_pred[i]>=0.65:\n",
        "                if y_test[i]==1:\n",
        "                    correct_pos += 1\n",
        "                else:\n",
        "                    wrong_pos += 1\n",
        "            elif yhat_pred[i]<=0.35:\n",
        "                if y_test[i]==0:\n",
        "                    correct_neg += 1\n",
        "                else:\n",
        "                    wrong_neg += 1\n",
        "        print('Extreme scores correct, pos:neg',correct_pos,correct_neg)  \n",
        "        print('Extreme scores incorrect pos:neg',wrong_pos,wrong_neg)  \n",
        "\n",
        "        #self._pretty_pictures(y_test,yhat_pred,prc_X,prc_Y)\n",
        "        \n",
        "        print('Accuracy: %.2f%% Precision: %.2f%% Recall: %.2f%%' % (accuracy,precision,recall)) \n",
        "        print('F1: %.2f%% MCC: %.4f' % (f1,mcc)) \n",
        "        print('AUPRC: %.2f%% AUROC: %.2f%%' % (auprc,auroc)) \n",
        "\n",
        "        self.cv_accuracy.append(accuracy)\n",
        "        self.cv_precision.append(precision)\n",
        "        self.cv_recall.append(recall)\n",
        "        self.cv_f1.append(f1)\n",
        "        self.cv_mcc.append(mcc)\n",
        "        self.cv_auprc.append(auprc)\n",
        "        self.cv_auroc.append(auroc)\n",
        "\n",
        "    def _pretty_pictures(self,y_valid,yhat_pred,prc_X,prc_Y):\n",
        "        count_ones= len(y_valid[y_valid==1])\n",
        "        count_zeros= len(y_valid[y_valid==0])\n",
        "        guess = max(count_ones,count_zeros) / len(y_valid)\n",
        "        # PRC\n",
        "        plt.plot(prc_X, prc_Y, marker='.')\n",
        "        plt.plot([0, 1], [guess,guess], linestyle='--')\n",
        "        plt.xlabel('Recall')\n",
        "        plt.ylabel('Precision')\n",
        "        plt.show()\n",
        "        # ROC\n",
        "        fpr, tpr, roc_bins = roc_curve(y_valid, yhat_pred)\n",
        "        plt.plot(fpr, tpr, marker='.')\n",
        "        plt.xlabel('False Positive Rate')\n",
        "        plt.ylabel('True Positive Rate')\n",
        "        plt.show()\n",
        "        \n",
        "    def get_statistics(self):\n",
        "        return \\\n",
        "        self.cv_accuracy,\\\n",
        "        self.cv_precision,\\\n",
        "        self.cv_recall,\\\n",
        "        self.cv_f1,\\\n",
        "        self.cv_mcc,\\\n",
        "        self.cv_auprc,\\\n",
        "        self.cv_auroc\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Td97uyyj5qDq"
      },
      "source": [
        "## Train"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 42,
      "metadata": {
        "id": "poe3rwAF4Ir7"
      },
      "outputs": [],
      "source": [
        "class Separator():\n",
        "    def __init__(self):\n",
        "        self.train_ids = []\n",
        "        self.train_seq = []\n",
        "        self.train_rci = dict()\n",
        "        self.val_ids = []\n",
        "        self.val_seq = []\n",
        "        self.val_rci = dict()\n",
        "    def load(self,data_dir,rep,fold):\n",
        "        filename='cv.{}.{}.validation_genes.txt'.format(rep,fold)\n",
        "        filename = data_dir + filename\n",
        "        self.val_genes = set()\n",
        "        print('Opening file',filename)\n",
        "        with open(filename,'r') as fin:\n",
        "            for line in fin:\n",
        "                gene_id = line.strip()\n",
        "                self.val_genes.add(gene_id)\n",
        "    def process(self,allids,allseq,gene_to_rci):\n",
        "        size = len(allids)\n",
        "        for t in range(size):\n",
        "            gene_id,tran_id = allids[t]\n",
        "            oneX            = allseq[t]\n",
        "            oneY            = gene_to_rci[gene_id]\n",
        "            in_middle = gene_to_rci[gene_id] >= -2 and gene_to_rci[gene_id] <= 0\n",
        "            in_tails = gene_to_rci[gene_id] < -2 or gene_to_rci[gene_id] > 0\n",
        "            if gene_id in self.val_genes:\n",
        "                if FILTER_TEST and (\\\n",
        "                    (FILTER_TAILS_TEST and in_tails) or \\\n",
        "                    (FILTER_MIDDLE_TEST and in_middle)):\n",
        "                    pass\n",
        "                else:\n",
        "                    self.val_ids.append(allids[t])\n",
        "                    self.val_seq.append(allseq[t])\n",
        "                    self.val_rci[gene_id]=gene_to_rci[gene_id]\n",
        "            else:\n",
        "                if FILTER_TRAIN and (\\\n",
        "                    (FILTER_TAILS_TRAIN and in_tails) or \\\n",
        "                    (FILTER_MIDDLE_TRAIN and in_middle)):\n",
        "                    pass\n",
        "                else:\n",
        "                    self.train_ids.append(allids[t])\n",
        "                    self.train_seq.append(allseq[t])\n",
        "                    self.train_rci[gene_id]=gene_to_rci[gene_id]\n",
        "    def get_ids(self):\n",
        "        return self.train_ids,self.val_ids\n",
        "    def get_seq(self):\n",
        "        return self.train_seq,self.val_seq\n",
        "    def get_rci(self):\n",
        "        return self.train_rci,self.val_rci"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 43,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "XC9m0W-pFLBH",
        "outputId": "1e65008b-47ea-451a-804a-238495786431",
        "scrolled": false
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2023-04-18 13:34:35.789840\n",
            "Load RCI from /content/drive/My Drive/data/Localization/TrainTest/TrainTest_ver43/train.lncRNA_RCI.csv\n",
            "Number of RCI values loaded 4371\n",
            "Num RCI: 4371\n",
            "Load sequence from /content/drive/My Drive/data/Localization/TrainTest/TrainTest_ver43/train.canon_lncRNA_transcripts.csv\n",
            "Loaded 4371 sequences.\n",
            "\n",
            "Training # 1 1\n",
            "2023-04-18 13:34:42.490256\n",
            "Opening file /content/drive/My Drive/data/Localization/TrainTest/TrainTest_ver43/cv.1.1.validation_genes.txt\n",
            "2023-04-18 13:34:42.498125\n",
            "Discriminator ZERO -1\n",
            "Train 1:0 939:1175 44.42%\n",
            "Valid 1:0 228:311 42.30%\n",
            "FIT\n",
            "2023-04-18 13:34:42.516226\n",
            "\n",
            "Testing # 1 1\n",
            "2023-04-18 13:34:52.805143\n",
            "2023-04-18 13:34:52.805215\n",
            "PREDICT\n",
            "2023-04-18 13:34:52.809039\n",
            "debug pred [0.2942026589849957, 0.22750141114937814, 0.3327914687000699]\n",
            "debug class [0 0 0]\n",
            "Test 1:0 228:311 42.30%\n",
            "Predict 1:0 196:343 36.36%\n",
            "Test sizes (539, 1025) (539,)\n",
            "Distrib of scores: 0.4449385827729592 mean 0.1940849180461532 std\n",
            "Range of scores: 0.06318736304441164 to 0.9464080579653094\n",
            "Score threshold 0.5\n",
            "Confusion matrix\n",
            " [[238  73]\n",
            " [105 123]]\n",
            "Normalized matrix\n",
            " [[0.44155844 0.13543599]\n",
            " [0.19480519 0.22820037]]\n",
            "Extreme scores correct, pos:neg 62 143\n",
            "Extreme scores incorrect pos:neg 35 43\n",
            "Accuracy: 66.98% Precision: 62.76% Recall: 53.95%\n",
            "F1: 58.02% MCC: 0.3130\n",
            "AUPRC: 59.82% AUROC: 69.96%\n",
            " accuracy [66.97588126159555]\n",
            " precision [62.755102040816325]\n",
            " recall [53.94736842105263]\n",
            " F1 [58.018867924528294]\n",
            " MCC [0.3129766630373976]\n",
            " AUPRC [59.82420934941637]\n",
            " AUROC [69.95684548993061]\n",
            "\n",
            "Training # 1 2\n",
            "2023-04-18 13:34:54.360582\n",
            "Opening file /content/drive/My Drive/data/Localization/TrainTest/TrainTest_ver43/cv.1.2.validation_genes.txt\n",
            "2023-04-18 13:34:54.379539\n",
            "Discriminator ZERO -1\n",
            "Train 1:0 926:1181 43.95%\n",
            "Valid 1:0 241:305 44.14%\n",
            "FIT\n",
            "2023-04-18 13:34:54.394839\n",
            "\n",
            "Testing # 1 2\n",
            "2023-04-18 13:35:04.673427\n",
            "2023-04-18 13:35:04.673484\n",
            "PREDICT\n",
            "2023-04-18 13:35:04.678812\n",
            "debug pred [0.2456566742239469, 0.6430217677162241, 0.18312632132042267]\n",
            "debug class [0 1 0]\n",
            "Test 1:0 241:305 44.14%\n",
            "Predict 1:0 203:343 37.18%\n",
            "Test sizes (546, 1025) (546,)\n",
            "Distrib of scores: 0.4391394638367557 mean 0.17949463648207856 std\n",
            "Range of scores: 0.06163751873140673 to 0.8955248157482902\n",
            "Score threshold 0.5\n",
            "Confusion matrix\n",
            " [[231  74]\n",
            " [112 129]]\n",
            "Normalized matrix\n",
            " [[0.42307692 0.13553114]\n",
            " [0.20512821 0.23626374]]\n",
            "Extreme scores correct, pos:neg 54 149\n",
            "Extreme scores incorrect pos:neg 20 47\n",
            "Accuracy: 65.93% Precision: 63.55% Recall: 53.53%\n",
            "F1: 58.11% MCC: 0.3007\n",
            "AUPRC: 65.08% AUROC: 71.56%\n",
            " accuracy [65.93406593406593]\n",
            " precision [63.54679802955665]\n",
            " recall [53.52697095435685]\n",
            " F1 [58.108108108108105]\n",
            " MCC [0.3006817576919483]\n",
            " AUPRC [65.08475215245915]\n",
            " AUROC [71.55975783960274]\n",
            "\n",
            "Training # 1 3\n",
            "2023-04-18 13:35:06.237192\n",
            "Opening file /content/drive/My Drive/data/Localization/TrainTest/TrainTest_ver43/cv.1.3.validation_genes.txt\n",
            "2023-04-18 13:35:06.252918\n",
            "Discriminator ZERO -1\n",
            "Train 1:0 951:1183 44.56%\n",
            "Valid 1:0 216:303 41.62%\n",
            "FIT\n",
            "2023-04-18 13:35:06.269846\n",
            "\n",
            "Testing # 1 3\n",
            "2023-04-18 13:35:16.730794\n",
            "2023-04-18 13:35:16.730877\n",
            "PREDICT\n",
            "2023-04-18 13:35:16.734783\n",
            "debug pred [0.4903305674095354, 0.263427548344665, 0.223099845149682]\n",
            "debug class [0 0 0]\n",
            "Test 1:0 216:303 41.62%\n",
            "Predict 1:0 186:333 35.84%\n",
            "Test sizes (519, 1025) (519,)\n",
            "Distrib of scores: 0.4453347269068167 mean 0.17700854356057003 std\n",
            "Range of scores: 0.055721191707523564 to 0.9190211808655199\n",
            "Score threshold 0.5\n",
            "Confusion matrix\n",
            " [[231  72]\n",
            " [102 114]]\n",
            "Normalized matrix\n",
            " [[0.44508671 0.13872832]\n",
            " [0.19653179 0.21965318]]\n",
            "Extreme scores correct, pos:neg 49 137\n",
            "Extreme scores incorrect pos:neg 26 30\n",
            "Accuracy: 66.47% Precision: 61.29% Recall: 52.78%\n",
            "F1: 56.72% MCC: 0.2983\n",
            "AUPRC: 60.62% AUROC: 72.18%\n",
            " accuracy [66.47398843930635]\n",
            " precision [61.29032258064516]\n",
            " recall [52.77777777777778]\n",
            " F1 [56.71641791044777]\n",
            " MCC [0.2982621788003131]\n",
            " AUPRC [60.615090720773814]\n",
            " AUROC [72.18249602738052]\n",
            "\n",
            "Training # 1 4\n",
            "2023-04-18 13:35:18.466669\n",
            "Opening file /content/drive/My Drive/data/Localization/TrainTest/TrainTest_ver43/cv.1.4.validation_genes.txt\n",
            "2023-04-18 13:35:18.476120\n",
            "Discriminator ZERO -1\n",
            "Train 1:0 936:1202 43.78%\n",
            "Valid 1:0 231:284 44.85%\n",
            "FIT\n",
            "2023-04-18 13:35:18.493777\n",
            "\n",
            "Testing # 1 4\n",
            "2023-04-18 13:35:28.687415\n",
            "2023-04-18 13:35:28.687872\n",
            "PREDICT\n",
            "2023-04-18 13:35:28.693014\n",
            "debug pred [0.39538903669585845, 0.3108723347239996, 0.44630881613620677]\n",
            "debug class [0 0 0]\n",
            "Test 1:0 231:284 44.85%\n",
            "Predict 1:0 166:349 32.23%\n",
            "Test sizes (515, 1025) (515,)\n",
            "Distrib of scores: 0.4304454733840177 mean 0.17483879268727304 std\n",
            "Range of scores: 0.05470609602732204 to 0.9008537930827241\n",
            "Score threshold 0.5\n",
            "Confusion matrix\n",
            " [[232  52]\n",
            " [117 114]]\n",
            "Normalized matrix\n",
            " [[0.45048544 0.10097087]\n",
            " [0.22718447 0.22135922]]\n",
            "Extreme scores correct, pos:neg 45 128\n",
            "Extreme scores incorrect pos:neg 17 46\n",
            "Accuracy: 67.18% Precision: 68.67% Recall: 49.35%\n",
            "F1: 57.43% MCC: 0.3303\n",
            "AUPRC: 64.52% AUROC: 70.79%\n",
            " accuracy [67.18446601941747]\n",
            " precision [68.67469879518072]\n",
            " recall [49.35064935064935]\n",
            " F1 [57.43073047858942]\n",
            " MCC [0.33031718661230836]\n",
            " AUPRC [64.51914449332222]\n",
            " AUROC [70.7868422657155]\n",
            "\n",
            "Training # 1 5\n",
            "2023-04-18 13:35:30.172349\n",
            "Opening file /content/drive/My Drive/data/Localization/TrainTest/TrainTest_ver43/cv.1.5.validation_genes.txt\n",
            "2023-04-18 13:35:30.187663\n",
            "Discriminator ZERO -1\n",
            "Train 1:0 916:1203 43.23%\n",
            "Valid 1:0 251:283 47.00%\n",
            "FIT\n",
            "2023-04-18 13:35:30.205195\n",
            "\n",
            "Testing # 1 5\n",
            "2023-04-18 13:35:39.562690\n",
            "2023-04-18 13:35:39.563096\n",
            "PREDICT\n",
            "2023-04-18 13:35:39.568353\n",
            "debug pred [0.2737959917851595, 0.18322552493888467, 0.35653747180872347]\n",
            "debug class [0 0 0]\n",
            "Test 1:0 251:283 47.00%\n",
            "Predict 1:0 186:348 34.83%\n",
            "Test sizes (534, 1025) (534,)\n",
            "Distrib of scores: 0.43801329691958646 mean 0.16856109386666412 std\n",
            "Range of scores: 0.07287373733793447 to 0.8726466123808251\n",
            "Score threshold 0.5\n",
            "Confusion matrix\n",
            " [[226  57]\n",
            " [122 129]]\n",
            "Normalized matrix\n",
            " [[0.42322097 0.10674157]\n",
            " [0.22846442 0.24157303]]\n",
            "Extreme scores correct, pos:neg 56 129\n",
            "Extreme scores incorrect pos:neg 16 42\n",
            "Accuracy: 66.48% Precision: 69.35% Recall: 51.39%\n",
            "F1: 59.04% MCC: 0.3274\n",
            "AUPRC: 70.71% AUROC: 73.96%\n",
            " accuracy [66.47940074906367]\n",
            " precision [69.35483870967742]\n",
            " recall [51.39442231075697]\n",
            " F1 [59.03890160183066]\n",
            " MCC [0.3273985878528939]\n",
            " AUPRC [70.70943437979052]\n",
            " AUROC [73.96139822336097]\n",
            "\n",
            "Training # 2 1\n",
            "2023-04-18 13:35:41.908295\n",
            "Opening file /content/drive/My Drive/data/Localization/TrainTest/TrainTest_ver43/cv.2.1.validation_genes.txt\n",
            "2023-04-18 13:35:41.924537\n",
            "Discriminator ZERO -1\n",
            "Train 1:0 941:1187 44.22%\n",
            "Valid 1:0 226:299 43.05%\n",
            "FIT\n",
            "2023-04-18 13:35:41.948494\n",
            "\n",
            "Testing # 2 1\n",
            "2023-04-18 13:35:51.480153\n",
            "2023-04-18 13:35:51.480231\n",
            "PREDICT\n",
            "2023-04-18 13:35:51.483943\n",
            "debug pred [0.40945920757480886, 0.23055008206050104, 0.2776503658031012]\n",
            "debug class [0 0 0]\n",
            "Test 1:0 226:299 43.05%\n",
            "Predict 1:0 184:341 35.05%\n",
            "Test sizes (525, 1025) (525,)\n",
            "Distrib of scores: 0.43442793043823935 mean 0.18482705163278607 std\n",
            "Range of scores: 0.05975221417344401 to 0.9133796944565937\n",
            "Score threshold 0.5\n",
            "Confusion matrix\n",
            " [[227  72]\n",
            " [114 112]]\n",
            "Normalized matrix\n",
            " [[0.43238095 0.13714286]\n",
            " [0.21714286 0.21333333]]\n",
            "Extreme scores correct, pos:neg 50 130\n",
            "Extreme scores incorrect pos:neg 22 48\n",
            "Accuracy: 64.57% Precision: 60.87% Recall: 49.56%\n",
            "F1: 54.63% MCC: 0.2644\n",
            "AUPRC: 59.61% AUROC: 68.46%\n",
            " accuracy [64.57142857142857]\n",
            " precision [60.86956521739131]\n",
            " recall [49.557522123893804]\n",
            " F1 [54.63414634146342]\n",
            " MCC [0.2643968880801209]\n",
            " AUPRC [59.607327603772745]\n",
            " AUROC [68.45976263059757]\n",
            "\n",
            "Training # 2 2\n",
            "2023-04-18 13:35:53.088973\n",
            "Opening file /content/drive/My Drive/data/Localization/TrainTest/TrainTest_ver43/cv.2.2.validation_genes.txt\n",
            "2023-04-18 13:35:53.104745\n",
            "Discriminator ZERO -1\n",
            "Train 1:0 929:1189 43.86%\n",
            "Valid 1:0 238:297 44.49%\n",
            "FIT\n",
            "2023-04-18 13:35:53.128049\n",
            "\n",
            "Testing # 2 2\n",
            "2023-04-18 13:36:03.277153\n",
            "2023-04-18 13:36:03.277206\n",
            "PREDICT\n",
            "2023-04-18 13:36:03.282233\n",
            "debug pred [0.16171687882371485, 0.36153338199767326, 0.40003068123014324]\n",
            "debug class [0 0 0]\n",
            "Test 1:0 238:297 44.49%\n",
            "Predict 1:0 205:330 38.32%\n",
            "Test sizes (535, 1025) (535,)\n",
            "Distrib of scores: 0.45001251621342214 mean 0.18226953277774974 std\n",
            "Range of scores: 0.06175686422952525 to 0.8874101066623499\n",
            "Score threshold 0.5\n",
            "Confusion matrix\n",
            " [[230  67]\n",
            " [100 138]]\n",
            "Normalized matrix\n",
            " [[0.42990654 0.12523364]\n",
            " [0.18691589 0.25794393]]\n",
            "Extreme scores correct, pos:neg 63 142\n",
            "Extreme scores incorrect pos:neg 23 34\n",
            "Accuracy: 68.79% Precision: 67.32% Recall: 57.98%\n",
            "F1: 62.30% MCC: 0.3621\n",
            "AUPRC: 69.75% AUROC: 74.78%\n",
            " accuracy [68.78504672897196]\n",
            " precision [67.31707317073172]\n",
            " recall [57.98319327731093]\n",
            " F1 [62.302483069977434]\n",
            " MCC [0.36210433233081957]\n",
            " AUPRC [69.74579358763576]\n",
            " AUROC [74.78142772260419]\n",
            "\n",
            "Training # 2 3\n",
            "2023-04-18 13:36:05.016049\n",
            "Opening file /content/drive/My Drive/data/Localization/TrainTest/TrainTest_ver43/cv.2.3.validation_genes.txt\n",
            "2023-04-18 13:36:05.029114\n",
            "Discriminator ZERO -1\n",
            "Train 1:0 924:1198 43.54%\n",
            "Valid 1:0 243:288 45.76%\n",
            "FIT\n",
            "2023-04-18 13:36:05.044134\n",
            "\n",
            "Testing # 2 3\n",
            "2023-04-18 13:36:15.086860\n",
            "2023-04-18 13:36:15.086917\n",
            "PREDICT\n",
            "2023-04-18 13:36:15.091202\n",
            "debug pred [0.4535462963734112, 0.7576771460672885, 0.3826950207246212]\n",
            "debug class [0 1 0]\n",
            "Test 1:0 243:288 45.76%\n",
            "Predict 1:0 183:348 34.46%\n",
            "Test sizes (531, 1025) (531,)\n",
            "Distrib of scores: 0.44006208310578027 mean 0.18116340532434802 std\n",
            "Range of scores: 0.060537114589942355 to 0.9113364800206011\n",
            "Score threshold 0.5\n",
            "Confusion matrix\n",
            " [[226  62]\n",
            " [122 121]]\n",
            "Normalized matrix\n",
            " [[0.42561205 0.11676083]\n",
            " [0.22975518 0.22787194]]\n",
            "Extreme scores correct, pos:neg 47 136\n",
            "Extreme scores incorrect pos:neg 30 44\n",
            "Accuracy: 65.35% Precision: 66.12% Recall: 49.79%\n",
            "F1: 56.81% MCC: 0.2963\n",
            "AUPRC: 64.02% AUROC: 71.34%\n",
            " accuracy [65.34839924670433]\n",
            " precision [66.12021857923497]\n",
            " recall [49.794238683127574]\n",
            " F1 [56.8075117370892]\n",
            " MCC [0.29631637359222185]\n",
            " AUPRC [64.02067362230977]\n",
            " AUROC [71.33987768632831]\n",
            "\n",
            "Training # 2 4\n",
            "2023-04-18 13:36:16.564043\n",
            "Opening file /content/drive/My Drive/data/Localization/TrainTest/TrainTest_ver43/cv.2.4.validation_genes.txt\n",
            "2023-04-18 13:36:16.578487\n",
            "Discriminator ZERO -1\n",
            "Train 1:0 953:1173 44.83%\n",
            "Valid 1:0 214:313 40.61%\n",
            "FIT\n",
            "2023-04-18 13:36:16.596677\n",
            "\n",
            "Testing # 2 4\n",
            "2023-04-18 13:36:26.856795\n",
            "2023-04-18 13:36:26.856850\n",
            "PREDICT\n",
            "2023-04-18 13:36:26.861656\n",
            "debug pred [0.2743296327886887, 0.4350869160149009, 0.6039340141371543]\n",
            "debug class [0 0 1]\n",
            "Test 1:0 214:313 40.61%\n",
            "Predict 1:0 185:342 35.10%\n",
            "Test sizes (527, 1025) (527,)\n",
            "Distrib of scores: 0.43519288188987476 mean 0.18517665071803457 std\n",
            "Range of scores: 0.06619554800572237 to 0.917531069902076\n",
            "Score threshold 0.5\n",
            "Confusion matrix\n",
            " [[245  68]\n",
            " [ 97 117]]\n",
            "Normalized matrix\n",
            " [[0.46489564 0.12903226]\n",
            " [0.18406072 0.22201139]]\n",
            "Extreme scores correct, pos:neg 50 143\n",
            "Extreme scores incorrect pos:neg 26 46\n",
            "Accuracy: 68.69% Precision: 63.24% Recall: 54.67%\n",
            "F1: 58.65% MCC: 0.3390\n",
            "AUPRC: 59.78% AUROC: 70.47%\n",
            " accuracy [68.69070208728652]\n",
            " precision [63.24324324324324]\n",
            " recall [54.67289719626168]\n",
            " F1 [58.64661654135338]\n",
            " MCC [0.339003890250172]\n",
            " AUPRC [59.78422082540169]\n",
            " AUROC [70.47266429787106]\n",
            "\n",
            "Training # 2 5\n",
            "2023-04-18 13:36:28.561675\n",
            "Opening file /content/drive/My Drive/data/Localization/TrainTest/TrainTest_ver43/cv.2.5.validation_genes.txt\n",
            "2023-04-18 13:36:28.577735\n",
            "Discriminator ZERO -1\n",
            "Train 1:0 921:1197 43.48%\n",
            "Valid 1:0 246:289 45.98%\n",
            "FIT\n",
            "2023-04-18 13:36:28.593768\n",
            "\n",
            "Testing # 2 5\n",
            "2023-04-18 13:36:39.177031\n",
            "2023-04-18 13:36:39.177269\n",
            "PREDICT\n",
            "2023-04-18 13:36:39.180779\n",
            "debug pred [0.5357683605412372, 0.49397084164413957, 0.28069073465958566]\n",
            "debug class [1 0 0]\n",
            "Test 1:0 246:289 45.98%\n",
            "Predict 1:0 169:366 31.59%\n",
            "Test sizes (535, 1025) (535,)\n",
            "Distrib of scores: 0.43713312657550996 mean 0.18798799036453645 std\n",
            "Range of scores: 0.05896074910505862 to 0.941599768882656\n",
            "Score threshold 0.5\n",
            "Confusion matrix\n",
            " [[234  55]\n",
            " [132 114]]\n",
            "Normalized matrix\n",
            " [[0.43738318 0.10280374]\n",
            " [0.24672897 0.21308411]]\n",
            "Extreme scores correct, pos:neg 59 137\n",
            "Extreme scores incorrect pos:neg 26 50\n",
            "Accuracy: 65.05% Precision: 67.46% Recall: 46.34%\n",
            "F1: 54.94% MCC: 0.2928\n",
            "AUPRC: 66.37% AUROC: 70.70%\n",
            " accuracy [65.04672897196262]\n",
            " precision [67.45562130177515]\n",
            " recall [46.34146341463415]\n",
            " F1 [54.93975903614457]\n",
            " MCC [0.29279212616247086]\n",
            " AUPRC [66.37087561680964]\n",
            " AUROC [70.6979491940248]\n",
            "2023-04-18 13:36:40.916296\n"
          ]
        }
      ],
      "source": [
        "print(datetime.now())\n",
        "\n",
        "accuracy=[]\n",
        "precision=[]\n",
        "recall=[]\n",
        "f1=[]\n",
        "mcc=[]\n",
        "auprc=[]\n",
        "auroc=[]\n",
        "\n",
        "loader = DataLoader()\n",
        "filepath = DATA_DIR+RCI_FILE_TRAIN\n",
        "print(\"Load RCI from\",filepath)\n",
        "gene_to_rci = loader.load_gene_rci_values(filepath,EXCLUSIONS)\n",
        "print('Num RCI:', len(gene_to_rci.keys()))\n",
        "filepath = DATA_DIR+SEQ_FILE_TRAIN\n",
        "print('Load sequence from',filepath)\n",
        "allids,allseq = loader.load_spectra(filepath)  # for MLP (load_sequence() for CNN)\n",
        "print('Loaded',len(allseq),'sequences.')\n",
        "test_gene_to_rci = None\n",
        "test_allids = None\n",
        "test_allseq = None\n",
        "if SEQ_FILE_TEST is not None:\n",
        "    # Train on the entire train set (no cross-validation).\n",
        "    # Evaluate with the test files.\n",
        "    test_loader = DataLoader()\n",
        "    filepath = DATA_DIR+RCI_FILE_TEST\n",
        "    print(\"Load RCI from\",filepath)\n",
        "    test_gene_to_rci = test_loader.load_gene_rci_values(filepath,EXCLUSIONS)\n",
        "    print('Num RCI:', len(test_gene_to_rci.keys()))\n",
        "    filepath = DATA_DIR+SEQ_FILE_TEST\n",
        "    print('Load sequence from',filepath)\n",
        "    test_allids,test_allseq = test_loader.load_spectra(filepath)\n",
        "    print('Loaded',len(test_allseq),'sequences.')\n",
        "\n",
        "for repeat in range(REPEATS):\n",
        "    for fold in range(FOLDS):\n",
        "        show_r = repeat+1  # display one-based counting\n",
        "        show_f = fold+1    # display one-based counting\n",
        "\n",
        "        print()\n",
        "        print(\"Training #\",show_r,show_f)\n",
        "        print(datetime.now())\n",
        "        cvdo = CrossValidator(EPOCHS)\n",
        "        cvdo.set_threshold_mechanism(RCI_THRESHOLD_MECHANISM)\n",
        "        if SEQ_FILE_TEST is None:\n",
        "            # Train on 80% and evaluate on 20%.\n",
        "            separator = Separator()\n",
        "            separator.load(DATA_DIR,show_r,show_f)\n",
        "            separator.process(allids,allseq,gene_to_rci)\n",
        "            train_allids,test_allids = separator.get_ids()\n",
        "            train_allseq,test_allseq = separator.get_seq()\n",
        "            train_gene_to_rci,test_gene_to_rci = separator.get_rci()\n",
        "            cvdo.train_new_model(\n",
        "                train_allids,train_allseq,train_gene_to_rci,\n",
        "                test_allids,test_allseq,test_gene_to_rci)\n",
        "            if SAVE_MODEL_FILENAME is not None:\n",
        "                filename = f\"{SAVE_MODEL_FILENAME}.{show_r}.{show_f}.model\"\n",
        "                cvdo.save_model(filename)\n",
        "        else:\n",
        "            # Train on the entire train set (no cross-validation).\n",
        "            # Evaluate with the test files.\n",
        "            train_allids = allids\n",
        "            train_allseq = allseq\n",
        "            train_gene_to_rci = gene_to_rci\n",
        "            BREAK = True\n",
        "            cvdo.train_new_model(\n",
        "                train_allids,train_allseq,train_gene_to_rci,\n",
        "                None,None,None)\n",
        "\n",
        "        print()\n",
        "        print(\"Testing #\",show_r,show_f)\n",
        "        print(datetime.now())\n",
        "        cvdo.reset_statistics()\n",
        "        cvdo.test_without_training(\n",
        "            test_allids,test_allseq,test_gene_to_rci)\n",
        "        cv_accuracy,cv_precision,cv_recall,cv_f1,cv_mcc,cv_auprc,cv_auroc=\\\n",
        "            cvdo.get_statistics()\n",
        "\n",
        "        print(\" accuracy\" ,  cv_accuracy)\n",
        "        print(\" precision\" , cv_precision)\n",
        "        print(\" recall\" ,    cv_recall)\n",
        "        print(\" F1\" ,        cv_f1)\n",
        "        print(\" MCC\" ,       cv_mcc)\n",
        "        print(\" AUPRC\" ,     cv_auprc)\n",
        "        print(\" AUROC\" ,     cv_auroc)\n",
        "\n",
        "        accuracy.append(cv_accuracy)\n",
        "        precision.append(cv_precision)\n",
        "        recall.append(cv_recall)\n",
        "        f1.append(cv_f1)\n",
        "        mcc.append(cv_mcc)\n",
        "        auprc.append(cv_auprc)\n",
        "        auroc.append(cv_auroc)\n",
        "        if BREAK: break\n",
        "    if BREAK: break\n",
        "print(datetime.now())"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 44,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "HkCeDg_HdQ36",
        "outputId": "8994f70f-fa39-434d-dd3e-3c505e2620b7"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            " accuracy mean 66.55% std 1.42\n",
            " precision mean 65.06% std 3.09\n",
            " recall mean 51.93% std 3.33\n",
            " F1 mean 57.66% std 2.19\n",
            " MCC mean 0.3124 std 0.0279\n",
            " AUPRC mean 64.03% std 4.10\n",
            " AUROC mean 71.42% std 1.86\n",
            " accuracy [[66.97588126159555], [65.93406593406593], [66.47398843930635], [67.18446601941747], [66.47940074906367], [64.57142857142857], [68.78504672897196], [65.34839924670433], [68.69070208728652], [65.04672897196262]]\n",
            " precision [[62.755102040816325], [63.54679802955665], [61.29032258064516], [68.67469879518072], [69.35483870967742], [60.86956521739131], [67.31707317073172], [66.12021857923497], [63.24324324324324], [67.45562130177515]]\n",
            " recall [[53.94736842105263], [53.52697095435685], [52.77777777777778], [49.35064935064935], [51.39442231075697], [49.557522123893804], [57.98319327731093], [49.794238683127574], [54.67289719626168], [46.34146341463415]]\n",
            " F1 [[58.018867924528294], [58.108108108108105], [56.71641791044777], [57.43073047858942], [59.03890160183066], [54.63414634146342], [62.302483069977434], [56.8075117370892], [58.64661654135338], [54.93975903614457]]\n",
            " MCC [[0.3129766630373976], [0.3006817576919483], [0.2982621788003131], [0.33031718661230836], [0.3273985878528939], [0.2643968880801209], [0.36210433233081957], [0.29631637359222185], [0.339003890250172], [0.29279212616247086]]\n",
            " AUPRC [[59.82420934941637], [65.08475215245915], [60.615090720773814], [64.51914449332222], [70.70943437979052], [59.607327603772745], [69.74579358763576], [64.02067362230977], [59.78422082540169], [66.37087561680964]]\n",
            " AUROC [[69.95684548993061], [71.55975783960274], [72.18249602738052], [70.7868422657155], [73.96139822336097], [68.45976263059757], [74.78142772260419], [71.33987768632831], [70.47266429787106], [70.6979491940248]]\n"
          ]
        }
      ],
      "source": [
        "def STD (values):\n",
        "    # ddof=1 reduces bias when extrapolating from sample to population\n",
        "    return np.std(values,ddof=1)\n",
        "\n",
        "print(\" accuracy mean %.2f%% std %.2f\" %  (np.mean(accuracy),  STD(accuracy)))\n",
        "print(\" precision mean %.2f%% std %.2f\" % (np.mean(precision), STD(precision)))\n",
        "print(\" recall mean %.2f%% std %.2f\" %    (np.mean(recall),    STD(recall)))\n",
        "print(\" F1 mean %.2f%% std %.2f\" %        (np.mean(f1),        STD(f1)))\n",
        "print(\" MCC mean %.4f std %.4f\" %       (np.mean(mcc),       STD(mcc)))\n",
        "print(\" AUPRC mean %.2f%% std %.2f\" %     (np.mean(auprc),     STD(auprc)))\n",
        "print(\" AUROC mean %.2f%% std %.2f\" %     (np.mean(auroc),     STD(auroc)))\n",
        "\n",
        "print(\" accuracy\"  , accuracy)\n",
        "print(\" precision\" , precision)\n",
        "print(\" recall\"    , recall)\n",
        "print(\" F1\"        , f1)\n",
        "print(\" MCC\"       , mcc)\n",
        "print(\" AUPRC\"     , auprc)\n",
        "print(\" AUROC\"     , auroc)\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 44,
      "metadata": {
        "id": "QjSVa72v4IsA"
      },
      "outputs": [],
      "source": []
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "gpuClass": "standard",
    "kernelspec": {
      "display_name": "Python 3 (ipykernel)",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.10.0"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}